// @vercel/next/dist/adapter/node-handler.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var node_handler_exports = {};
__export(node_handler_exports, {
  getHandlerSource: () => getHandlerSource
});
module.exports = __toCommonJS(node_handler_exports);
const getHandlerSource = (ctx) => `
  process.env.NODE_ENV = 'production';
  process.chdir(__dirname);
  
  require('next/setup-node-env')
  
  const _n_handler = (${ctx.isMiddleware ? () => {
  const path = require("path");
  const relativeDistDir = process.env.__PRIVATE_RELATIVE_DIST_DIR;
  const SYMBOL_FOR_REQ_CONTEXT = Symbol.for("@vercel/request-context");
  function getRequestContext() {
    const fromSymbol = globalThis;
    return fromSymbol[SYMBOL_FOR_REQ_CONTEXT]?.get?.() ?? {};
  }
  return async function handler(request) {
    let middlewareHandler = await require("./" + path.posix.join(relativeDistDir, "server", "middleware.js"));
    middlewareHandler = middlewareHandler.handler || middlewareHandler;
    const context = getRequestContext();
    const response = await middlewareHandler(request, {
      waitUntil: context.waitUntil,
      requestMeta: {
        // we use '.' for relative project dir since we process.chdir
        // to the same directory as the handler file so everything is
        // relative to that/project dir
        relativeProjectDir: "."
      }
    });
    return response;
  };
} : (() => {
  const path = require("path");
  const relativeDistDir = process.env.__PRIVATE_RELATIVE_DIST_DIR;
  const prerenderFallbackFalseMap = process.env.__PRIVATE_PRERENDER_FALLBACK_MAP;
  const {
    dynamicRoutes: dynamicRoutesRaw,
    staticRoutes: staticRoutesRaw,
    i18n
  } = require("./" + path.posix.join(relativeDistDir, "routes-manifest.json"));
  const matchOperatorsRegex = /[|\\{}()[\]^$+*?.-]/g;
  function escapeStringRegexp(str) {
    return str.replace(matchOperatorsRegex, "\\$&");
  }
  const dynamicRoutes = dynamicRoutesRaw.map((item) => {
    return {
      ...item,
      namedRegex: new RegExp(item.namedRegex || item.regex)
    };
  });
  const staticRoutes = staticRoutesRaw.map((route) => {
    return {
      ...route,
      namedRegex: new RegExp(
        "^" + escapeStringRegexp(route.page) + "$"
      )
    };
  });
  let appPathRoutesManifest = {};
  try {
    appPathRoutesManifest = require("./" + path.posix.join(
      relativeDistDir,
      "app-path-routes-manifest.json"
    ));
  } catch (_) {
  }
  const inversedAppRoutesManifest = Object.entries(
    appPathRoutesManifest
  ).reduce(
    (manifest, [originalKey, normalizedKey]) => {
      manifest[normalizedKey] = originalKey;
      return manifest;
    },
    {}
  );
  function normalizeLocalePath(pathname, locales) {
    if (!locales) return { pathname };
    const lowercasedLocales = locales.map(
      (locale) => locale.toLowerCase()
    );
    const segments = pathname.split("/", 2);
    if (!segments[1]) return { pathname };
    const segment = segments[1].toLowerCase();
    const index = lowercasedLocales.indexOf(segment);
    if (index < 0) return { pathname };
    const detectedLocale = locales[index];
    pathname = pathname.slice(detectedLocale.length + 1) || "/";
    return { pathname, locale: detectedLocale };
  }
  function normalizeDataPath(pathname) {
    if (!(pathname || "/").startsWith("/_next/data")) {
      return pathname;
    }
    pathname = pathname.replace(/\/_next\/data\/[^/]{1,}/, "").replace(/\.json$/, "");
    if (pathname === "/index") {
      return "/";
    }
    return pathname;
  }
  function matchUrlToPage(urlPathname) {
    urlPathname = normalizeDataPath(urlPathname);
    for (const suffixRegex of [
      /\.segments(\/.*)\.segment\.rsc$/,
      /\.rsc$/
    ]) {
      urlPathname = urlPathname.replace(suffixRegex, "");
    }
    const urlPathnameWithLocale = urlPathname;
    const normalizeResult = normalizeLocalePath(
      urlPathname,
      i18n?.locales
    );
    urlPathname = normalizeResult.pathname;
    urlPathname = urlPathname.replace(/\/$/, "") || "/";
    const combinedRoutes = [...staticRoutes, ...dynamicRoutes];
    for (const route of combinedRoutes) {
      if (route.page === urlPathname) {
        return {
          matchedPathname: inversedAppRoutesManifest[route.page] || route.page,
          locale: normalizeResult.locale
        };
      }
    }
    for (const route of [...staticRoutes, ...dynamicRoutes]) {
      const matches = urlPathname.match(route.namedRegex);
      if (matches || urlPathname === "/index" && route.namedRegex.test("/")) {
        const fallbackFalseMap = prerenderFallbackFalseMap[route.page];
        if (fallbackFalseMap && !(fallbackFalseMap.includes(urlPathname) || fallbackFalseMap.includes(urlPathnameWithLocale))) {
          continue;
        }
        return {
          matchedPathname: inversedAppRoutesManifest[route.page] || route.page,
          locale: normalizeResult.locale,
          matches
        };
      }
    }
    return {
      matchedPathname: inversedAppRoutesManifest[urlPathname] || urlPathname,
      locale: normalizeResult.locale
    };
  }
  const SYMBOL_FOR_REQ_CONTEXT = Symbol.for("@vercel/request-context");
  function getRequestContext() {
    const fromSymbol = globalThis;
    return fromSymbol[SYMBOL_FOR_REQ_CONTEXT]?.get?.() ?? {};
  }
  const RouterServerContextSymbol = Symbol.for(
    "@next/router-server-methods"
  );
  const routerServerGlobal = globalThis;
  if (!routerServerGlobal[RouterServerContextSymbol]) {
    routerServerGlobal[RouterServerContextSymbol] = {};
  }
  routerServerGlobal[RouterServerContextSymbol]["."] = {
    async render404(req, res) {
      let mod;
      try {
        try {
          mod = await require("./" + path.posix.join(
            relativeDistDir,
            "server",
            "app",
            `_not-found`,
            "page.js"
          ));
        } catch {
        }
        if (!mod) {
          mod = await require("./" + path.posix.join(
            relativeDistDir,
            "server",
            "pages",
            `404.js`
          ));
        }
      } catch (_) {
        mod = await require("./" + path.posix.join(
          relativeDistDir,
          "server",
          "pages",
          `_error.js`
        ));
      }
      res.statusCode = 404;
      if (mod) {
        await mod.handler(req, res, {
          waitUntil: getRequestContext().waitUntil
        });
      } else {
        res.end("This page could not be found");
      }
    }
  };
  function fixMojibake(input) {
    const bytes = new Uint8Array(input.length);
    for (let i = 0; i < input.length; i++) {
      bytes[i] = input.charCodeAt(i);
    }
    const decoder = new TextDecoder("utf-8");
    return decoder.decode(bytes);
  }
  return async function handler(req, res, internalMetadata) {
    try {
      const parsedUrl = new URL(req.url || "/", "http://n");
      const initURL = `https://${req.headers.host || "localhost"}${parsedUrl.pathname}${parsedUrl.search}`;
      let urlPathname = typeof req.headers["x-matched-path"] === "string" ? fixMojibake(req.headers["x-matched-path"]) : void 0;
      if (typeof urlPathname !== "string") {
        urlPathname = parsedUrl.pathname || "/";
      }
      const {
        matchedPathname: page,
        locale,
        matches
      } = matchUrlToPage(urlPathname);
      const isAppDir = page.match(/\/(page|route)$/);
      let addedMatchesToUrl = false;
      for (const matchKey in matches?.groups || {}) {
        const matchValue = matches?.groups?.[matchKey];
        if (!parsedUrl.searchParams.has(matchKey) && matchValue) {
          parsedUrl.searchParams.set(matchKey, matchValue);
          addedMatchesToUrl = true;
        }
      }
      if (addedMatchesToUrl) {
        req.url = `${parsedUrl.pathname}${parsedUrl.searchParams.size > 0 ? "?" : ""}${parsedUrl.searchParams.toString()}`;
      }
      const mod = await require("./" + path.posix.join(
        relativeDistDir,
        "server",
        isAppDir ? "app" : "pages",
        `${page === "/" ? "index" : page}.js`
      ));
      await mod.handler(req, res, {
        waitUntil: getRequestContext().waitUntil,
        requestMeta: {
          ...internalMetadata,
          minimalMode: true,
          // we use '.' for relative project dir since we process.chdir
          // to the same directory as the handler file so everything is
          // relative to that/project dir
          relativeProjectDir: ".",
          locale,
          initURL
        }
      });
    } catch (error) {
      console.error(`Failed to handle ${req.url}`, error);
      throw error;
    }
  };
}).toString()})()
  
  module.exports = _n_handler
  
  ${ctx.isMiddleware ? "" : `
    module.exports.getRequestHandlerWithMetadata = (metadata) => {
      return (req, res) => _n_handler(req, res, metadata)
    }
  `}
  
  `.replaceAll(
  "process.env.__PRIVATE_RELATIVE_DIST_DIR",
  `"${ctx.projectRelativeDistDir}"`
).replaceAll(
  "process.env.__PRIVATE_PRERENDER_FALLBACK_MAP",
  JSON.stringify(ctx.prerenderFallbackFalseMap)
).replaceAll(
  "process.env.__PRIVATE_NEXT_CONFIG",
  JSON.stringify(ctx.nextConfig)
);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getHandlerSource
});
