// @vercel/python/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// ../../node_modules/.pnpm/nice-try@1.0.5/node_modules/nice-try/src/index.js
var require_src = __commonJS({
  "../../node_modules/.pnpm/nice-try@1.0.5/node_modules/nice-try/src/index.js"(exports, module2) {
    "use strict";
    module2.exports = function(fn) {
      try {
        return fn();
      } catch (e) {
      }
    };
  }
});

// ../../node_modules/.pnpm/isexe@2.0.0/node_modules/isexe/windows.js
var require_windows = __commonJS({
  "../../node_modules/.pnpm/isexe@2.0.0/node_modules/isexe/windows.js"(exports, module2) {
    module2.exports = isexe;
    isexe.sync = sync;
    var fs8 = require("fs");
    function checkPathExt(path, options) {
      var pathext = options.pathExt !== void 0 ? options.pathExt : process.env.PATHEXT;
      if (!pathext) {
        return true;
      }
      pathext = pathext.split(";");
      if (pathext.indexOf("") !== -1) {
        return true;
      }
      for (var i = 0; i < pathext.length; i++) {
        var p = pathext[i].toLowerCase();
        if (p && path.substr(-p.length).toLowerCase() === p) {
          return true;
        }
      }
      return false;
    }
    function checkStat(stat, path, options) {
      if (!stat.isSymbolicLink() && !stat.isFile()) {
        return false;
      }
      return checkPathExt(path, options);
    }
    function isexe(path, options, cb) {
      fs8.stat(path, function(er, stat) {
        cb(er, er ? false : checkStat(stat, path, options));
      });
    }
    function sync(path, options) {
      return checkStat(fs8.statSync(path), path, options);
    }
  }
});

// ../../node_modules/.pnpm/isexe@2.0.0/node_modules/isexe/mode.js
var require_mode = __commonJS({
  "../../node_modules/.pnpm/isexe@2.0.0/node_modules/isexe/mode.js"(exports, module2) {
    module2.exports = isexe;
    isexe.sync = sync;
    var fs8 = require("fs");
    function isexe(path, options, cb) {
      fs8.stat(path, function(er, stat) {
        cb(er, er ? false : checkStat(stat, options));
      });
    }
    function sync(path, options) {
      return checkStat(fs8.statSync(path), options);
    }
    function checkStat(stat, options) {
      return stat.isFile() && checkMode(stat, options);
    }
    function checkMode(stat, options) {
      var mod = stat.mode;
      var uid = stat.uid;
      var gid = stat.gid;
      var myUid = options.uid !== void 0 ? options.uid : process.getuid && process.getuid();
      var myGid = options.gid !== void 0 ? options.gid : process.getgid && process.getgid();
      var u = parseInt("100", 8);
      var g = parseInt("010", 8);
      var o = parseInt("001", 8);
      var ug = u | g;
      var ret = mod & o || mod & g && gid === myGid || mod & u && uid === myUid || mod & ug && myUid === 0;
      return ret;
    }
  }
});

// ../../node_modules/.pnpm/isexe@2.0.0/node_modules/isexe/index.js
var require_isexe = __commonJS({
  "../../node_modules/.pnpm/isexe@2.0.0/node_modules/isexe/index.js"(exports, module2) {
    var fs8 = require("fs");
    var core;
    if (process.platform === "win32" || global.TESTING_WINDOWS) {
      core = require_windows();
    } else {
      core = require_mode();
    }
    module2.exports = isexe;
    isexe.sync = sync;
    function isexe(path, options, cb) {
      if (typeof options === "function") {
        cb = options;
        options = {};
      }
      if (!cb) {
        if (typeof Promise !== "function") {
          throw new TypeError("callback not provided");
        }
        return new Promise(function(resolve2, reject) {
          isexe(path, options || {}, function(er, is) {
            if (er) {
              reject(er);
            } else {
              resolve2(is);
            }
          });
        });
      }
      core(path, options || {}, function(er, is) {
        if (er) {
          if (er.code === "EACCES" || options && options.ignoreErrors) {
            er = null;
            is = false;
          }
        }
        cb(er, is);
      });
    }
    function sync(path, options) {
      try {
        return core.sync(path, options || {});
      } catch (er) {
        if (options && options.ignoreErrors || er.code === "EACCES") {
          return false;
        } else {
          throw er;
        }
      }
    }
  }
});

// ../../node_modules/.pnpm/which@1.3.1/node_modules/which/which.js
var require_which = __commonJS({
  "../../node_modules/.pnpm/which@1.3.1/node_modules/which/which.js"(exports, module2) {
    module2.exports = which2;
    which2.sync = whichSync;
    var isWindows = process.platform === "win32" || process.env.OSTYPE === "cygwin" || process.env.OSTYPE === "msys";
    var path = require("path");
    var COLON = isWindows ? ";" : ":";
    var isexe = require_isexe();
    function getNotFoundError(cmd) {
      var er = new Error("not found: " + cmd);
      er.code = "ENOENT";
      return er;
    }
    function getPathInfo(cmd, opt) {
      var colon = opt.colon || COLON;
      var pathEnv = opt.path || process.env.PATH || "";
      var pathExt = [""];
      pathEnv = pathEnv.split(colon);
      var pathExtExe = "";
      if (isWindows) {
        pathEnv.unshift(process.cwd());
        pathExtExe = opt.pathExt || process.env.PATHEXT || ".EXE;.CMD;.BAT;.COM";
        pathExt = pathExtExe.split(colon);
        if (cmd.indexOf(".") !== -1 && pathExt[0] !== "")
          pathExt.unshift("");
      }
      if (cmd.match(/\//) || isWindows && cmd.match(/\\/))
        pathEnv = [""];
      return {
        env: pathEnv,
        ext: pathExt,
        extExe: pathExtExe
      };
    }
    function which2(cmd, opt, cb) {
      if (typeof opt === "function") {
        cb = opt;
        opt = {};
      }
      var info = getPathInfo(cmd, opt);
      var pathEnv = info.env;
      var pathExt = info.ext;
      var pathExtExe = info.extExe;
      var found = [];
      (function F(i, l) {
        if (i === l) {
          if (opt.all && found.length)
            return cb(null, found);
          else
            return cb(getNotFoundError(cmd));
        }
        var pathPart = pathEnv[i];
        if (pathPart.charAt(0) === '"' && pathPart.slice(-1) === '"')
          pathPart = pathPart.slice(1, -1);
        var p = path.join(pathPart, cmd);
        if (!pathPart && /^\.[\\\/]/.test(cmd)) {
          p = cmd.slice(0, 2) + p;
        }
        ;
        (function E(ii, ll) {
          if (ii === ll)
            return F(i + 1, l);
          var ext = pathExt[ii];
          isexe(p + ext, { pathExt: pathExtExe }, function(er, is) {
            if (!er && is) {
              if (opt.all)
                found.push(p + ext);
              else
                return cb(null, p + ext);
            }
            return E(ii + 1, ll);
          });
        })(0, pathExt.length);
      })(0, pathEnv.length);
    }
    function whichSync(cmd, opt) {
      opt = opt || {};
      var info = getPathInfo(cmd, opt);
      var pathEnv = info.env;
      var pathExt = info.ext;
      var pathExtExe = info.extExe;
      var found = [];
      for (var i = 0, l = pathEnv.length; i < l; i++) {
        var pathPart = pathEnv[i];
        if (pathPart.charAt(0) === '"' && pathPart.slice(-1) === '"')
          pathPart = pathPart.slice(1, -1);
        var p = path.join(pathPart, cmd);
        if (!pathPart && /^\.[\\\/]/.test(cmd)) {
          p = cmd.slice(0, 2) + p;
        }
        for (var j = 0, ll = pathExt.length; j < ll; j++) {
          var cur = p + pathExt[j];
          var is;
          try {
            is = isexe.sync(cur, { pathExt: pathExtExe });
            if (is) {
              if (opt.all)
                found.push(cur);
              else
                return cur;
            }
          } catch (ex) {
          }
        }
      }
      if (opt.all && found.length)
        return found;
      if (opt.nothrow)
        return null;
      throw getNotFoundError(cmd);
    }
  }
});

// ../../node_modules/.pnpm/path-key@2.0.1/node_modules/path-key/index.js
var require_path_key = __commonJS({
  "../../node_modules/.pnpm/path-key@2.0.1/node_modules/path-key/index.js"(exports, module2) {
    "use strict";
    module2.exports = (opts) => {
      opts = opts || {};
      const env = opts.env || process.env;
      const platform = opts.platform || process.platform;
      if (platform !== "win32") {
        return "PATH";
      }
      return Object.keys(env).find((x) => x.toUpperCase() === "PATH") || "Path";
    };
  }
});

// ../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/util/resolveCommand.js
var require_resolveCommand = __commonJS({
  "../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/util/resolveCommand.js"(exports, module2) {
    "use strict";
    var path = require("path");
    var which2 = require_which();
    var pathKey = require_path_key()();
    function resolveCommandAttempt(parsed, withoutPathExt) {
      const cwd = process.cwd();
      const hasCustomCwd = parsed.options.cwd != null;
      if (hasCustomCwd) {
        try {
          process.chdir(parsed.options.cwd);
        } catch (err) {
        }
      }
      let resolved;
      try {
        resolved = which2.sync(parsed.command, {
          path: (parsed.options.env || process.env)[pathKey],
          pathExt: withoutPathExt ? path.delimiter : void 0
        });
      } catch (e) {
      } finally {
        process.chdir(cwd);
      }
      if (resolved) {
        resolved = path.resolve(hasCustomCwd ? parsed.options.cwd : "", resolved);
      }
      return resolved;
    }
    function resolveCommand(parsed) {
      return resolveCommandAttempt(parsed) || resolveCommandAttempt(parsed, true);
    }
    module2.exports = resolveCommand;
  }
});

// ../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/util/escape.js
var require_escape = __commonJS({
  "../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/util/escape.js"(exports, module2) {
    "use strict";
    var metaCharsRegExp = /([()\][%!^"`<>&|;, *?])/g;
    function escapeCommand(arg) {
      arg = arg.replace(metaCharsRegExp, "^$1");
      return arg;
    }
    function escapeArgument(arg, doubleEscapeMetaChars) {
      arg = `${arg}`;
      arg = arg.replace(/(\\*)"/g, '$1$1\\"');
      arg = arg.replace(/(\\*)$/, "$1$1");
      arg = `"${arg}"`;
      arg = arg.replace(metaCharsRegExp, "^$1");
      if (doubleEscapeMetaChars) {
        arg = arg.replace(metaCharsRegExp, "^$1");
      }
      return arg;
    }
    module2.exports.command = escapeCommand;
    module2.exports.argument = escapeArgument;
  }
});

// ../../node_modules/.pnpm/shebang-regex@1.0.0/node_modules/shebang-regex/index.js
var require_shebang_regex = __commonJS({
  "../../node_modules/.pnpm/shebang-regex@1.0.0/node_modules/shebang-regex/index.js"(exports, module2) {
    "use strict";
    module2.exports = /^#!.*/;
  }
});

// ../../node_modules/.pnpm/shebang-command@1.2.0/node_modules/shebang-command/index.js
var require_shebang_command = __commonJS({
  "../../node_modules/.pnpm/shebang-command@1.2.0/node_modules/shebang-command/index.js"(exports, module2) {
    "use strict";
    var shebangRegex = require_shebang_regex();
    module2.exports = function(str) {
      var match = str.match(shebangRegex);
      if (!match) {
        return null;
      }
      var arr = match[0].replace(/#! ?/, "").split(" ");
      var bin = arr[0].split("/").pop();
      var arg = arr[1];
      return bin === "env" ? arg : bin + (arg ? " " + arg : "");
    };
  }
});

// ../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/util/readShebang.js
var require_readShebang = __commonJS({
  "../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/util/readShebang.js"(exports, module2) {
    "use strict";
    var fs8 = require("fs");
    var shebangCommand = require_shebang_command();
    function readShebang(command) {
      const size = 150;
      let buffer;
      if (Buffer.alloc) {
        buffer = Buffer.alloc(size);
      } else {
        buffer = new Buffer(size);
        buffer.fill(0);
      }
      let fd;
      try {
        fd = fs8.openSync(command, "r");
        fs8.readSync(fd, buffer, 0, size, 0);
        fs8.closeSync(fd);
      } catch (e) {
      }
      return shebangCommand(buffer.toString());
    }
    module2.exports = readShebang;
  }
});

// ../../node_modules/.pnpm/semver@5.7.2/node_modules/semver/semver.js
var require_semver = __commonJS({
  "../../node_modules/.pnpm/semver@5.7.2/node_modules/semver/semver.js"(exports, module2) {
    exports = module2.exports = SemVer;
    var debug10;
    if (typeof process === "object" && process.env && process.env.NODE_DEBUG && /\bsemver\b/i.test(process.env.NODE_DEBUG)) {
      debug10 = function() {
        var args = Array.prototype.slice.call(arguments, 0);
        args.unshift("SEMVER");
        console.log.apply(console, args);
      };
    } else {
      debug10 = function() {
      };
    }
    exports.SEMVER_SPEC_VERSION = "2.0.0";
    var MAX_LENGTH = 256;
    var MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || /* istanbul ignore next */
    9007199254740991;
    var MAX_SAFE_COMPONENT_LENGTH = 16;
    var MAX_SAFE_BUILD_LENGTH = MAX_LENGTH - 6;
    var re = exports.re = [];
    var safeRe = exports.safeRe = [];
    var src = exports.src = [];
    var R = 0;
    var LETTERDASHNUMBER = "[a-zA-Z0-9-]";
    var safeRegexReplacements = [
      ["\\s", 1],
      ["\\d", MAX_LENGTH],
      [LETTERDASHNUMBER, MAX_SAFE_BUILD_LENGTH]
    ];
    function makeSafeRe(value) {
      for (var i2 = 0; i2 < safeRegexReplacements.length; i2++) {
        var token = safeRegexReplacements[i2][0];
        var max = safeRegexReplacements[i2][1];
        value = value.split(token + "*").join(token + "{0," + max + "}").split(token + "+").join(token + "{1," + max + "}");
      }
      return value;
    }
    var NUMERICIDENTIFIER = R++;
    src[NUMERICIDENTIFIER] = "0|[1-9]\\d*";
    var NUMERICIDENTIFIERLOOSE = R++;
    src[NUMERICIDENTIFIERLOOSE] = "\\d+";
    var NONNUMERICIDENTIFIER = R++;
    src[NONNUMERICIDENTIFIER] = "\\d*[a-zA-Z-]" + LETTERDASHNUMBER + "*";
    var MAINVERSION = R++;
    src[MAINVERSION] = "(" + src[NUMERICIDENTIFIER] + ")\\.(" + src[NUMERICIDENTIFIER] + ")\\.(" + src[NUMERICIDENTIFIER] + ")";
    var MAINVERSIONLOOSE = R++;
    src[MAINVERSIONLOOSE] = "(" + src[NUMERICIDENTIFIERLOOSE] + ")\\.(" + src[NUMERICIDENTIFIERLOOSE] + ")\\.(" + src[NUMERICIDENTIFIERLOOSE] + ")";
    var PRERELEASEIDENTIFIER = R++;
    src[PRERELEASEIDENTIFIER] = "(?:" + src[NUMERICIDENTIFIER] + "|" + src[NONNUMERICIDENTIFIER] + ")";
    var PRERELEASEIDENTIFIERLOOSE = R++;
    src[PRERELEASEIDENTIFIERLOOSE] = "(?:" + src[NUMERICIDENTIFIERLOOSE] + "|" + src[NONNUMERICIDENTIFIER] + ")";
    var PRERELEASE = R++;
    src[PRERELEASE] = "(?:-(" + src[PRERELEASEIDENTIFIER] + "(?:\\." + src[PRERELEASEIDENTIFIER] + ")*))";
    var PRERELEASELOOSE = R++;
    src[PRERELEASELOOSE] = "(?:-?(" + src[PRERELEASEIDENTIFIERLOOSE] + "(?:\\." + src[PRERELEASEIDENTIFIERLOOSE] + ")*))";
    var BUILDIDENTIFIER = R++;
    src[BUILDIDENTIFIER] = LETTERDASHNUMBER + "+";
    var BUILD = R++;
    src[BUILD] = "(?:\\+(" + src[BUILDIDENTIFIER] + "(?:\\." + src[BUILDIDENTIFIER] + ")*))";
    var FULL = R++;
    var FULLPLAIN = "v?" + src[MAINVERSION] + src[PRERELEASE] + "?" + src[BUILD] + "?";
    src[FULL] = "^" + FULLPLAIN + "$";
    var LOOSEPLAIN = "[v=\\s]*" + src[MAINVERSIONLOOSE] + src[PRERELEASELOOSE] + "?" + src[BUILD] + "?";
    var LOOSE = R++;
    src[LOOSE] = "^" + LOOSEPLAIN + "$";
    var GTLT = R++;
    src[GTLT] = "((?:<|>)?=?)";
    var XRANGEIDENTIFIERLOOSE = R++;
    src[XRANGEIDENTIFIERLOOSE] = src[NUMERICIDENTIFIERLOOSE] + "|x|X|\\*";
    var XRANGEIDENTIFIER = R++;
    src[XRANGEIDENTIFIER] = src[NUMERICIDENTIFIER] + "|x|X|\\*";
    var XRANGEPLAIN = R++;
    src[XRANGEPLAIN] = "[v=\\s]*(" + src[XRANGEIDENTIFIER] + ")(?:\\.(" + src[XRANGEIDENTIFIER] + ")(?:\\.(" + src[XRANGEIDENTIFIER] + ")(?:" + src[PRERELEASE] + ")?" + src[BUILD] + "?)?)?";
    var XRANGEPLAINLOOSE = R++;
    src[XRANGEPLAINLOOSE] = "[v=\\s]*(" + src[XRANGEIDENTIFIERLOOSE] + ")(?:\\.(" + src[XRANGEIDENTIFIERLOOSE] + ")(?:\\.(" + src[XRANGEIDENTIFIERLOOSE] + ")(?:" + src[PRERELEASELOOSE] + ")?" + src[BUILD] + "?)?)?";
    var XRANGE = R++;
    src[XRANGE] = "^" + src[GTLT] + "\\s*" + src[XRANGEPLAIN] + "$";
    var XRANGELOOSE = R++;
    src[XRANGELOOSE] = "^" + src[GTLT] + "\\s*" + src[XRANGEPLAINLOOSE] + "$";
    var COERCE = R++;
    src[COERCE] = "(?:^|[^\\d])(\\d{1," + MAX_SAFE_COMPONENT_LENGTH + "})(?:\\.(\\d{1," + MAX_SAFE_COMPONENT_LENGTH + "}))?(?:\\.(\\d{1," + MAX_SAFE_COMPONENT_LENGTH + "}))?(?:$|[^\\d])";
    var LONETILDE = R++;
    src[LONETILDE] = "(?:~>?)";
    var TILDETRIM = R++;
    src[TILDETRIM] = "(\\s*)" + src[LONETILDE] + "\\s+";
    re[TILDETRIM] = new RegExp(src[TILDETRIM], "g");
    safeRe[TILDETRIM] = new RegExp(makeSafeRe(src[TILDETRIM]), "g");
    var tildeTrimReplace = "$1~";
    var TILDE = R++;
    src[TILDE] = "^" + src[LONETILDE] + src[XRANGEPLAIN] + "$";
    var TILDELOOSE = R++;
    src[TILDELOOSE] = "^" + src[LONETILDE] + src[XRANGEPLAINLOOSE] + "$";
    var LONECARET = R++;
    src[LONECARET] = "(?:\\^)";
    var CARETTRIM = R++;
    src[CARETTRIM] = "(\\s*)" + src[LONECARET] + "\\s+";
    re[CARETTRIM] = new RegExp(src[CARETTRIM], "g");
    safeRe[CARETTRIM] = new RegExp(makeSafeRe(src[CARETTRIM]), "g");
    var caretTrimReplace = "$1^";
    var CARET = R++;
    src[CARET] = "^" + src[LONECARET] + src[XRANGEPLAIN] + "$";
    var CARETLOOSE = R++;
    src[CARETLOOSE] = "^" + src[LONECARET] + src[XRANGEPLAINLOOSE] + "$";
    var COMPARATORLOOSE = R++;
    src[COMPARATORLOOSE] = "^" + src[GTLT] + "\\s*(" + LOOSEPLAIN + ")$|^$";
    var COMPARATOR = R++;
    src[COMPARATOR] = "^" + src[GTLT] + "\\s*(" + FULLPLAIN + ")$|^$";
    var COMPARATORTRIM = R++;
    src[COMPARATORTRIM] = "(\\s*)" + src[GTLT] + "\\s*(" + LOOSEPLAIN + "|" + src[XRANGEPLAIN] + ")";
    re[COMPARATORTRIM] = new RegExp(src[COMPARATORTRIM], "g");
    safeRe[COMPARATORTRIM] = new RegExp(makeSafeRe(src[COMPARATORTRIM]), "g");
    var comparatorTrimReplace = "$1$2$3";
    var HYPHENRANGE = R++;
    src[HYPHENRANGE] = "^\\s*(" + src[XRANGEPLAIN] + ")\\s+-\\s+(" + src[XRANGEPLAIN] + ")\\s*$";
    var HYPHENRANGELOOSE = R++;
    src[HYPHENRANGELOOSE] = "^\\s*(" + src[XRANGEPLAINLOOSE] + ")\\s+-\\s+(" + src[XRANGEPLAINLOOSE] + ")\\s*$";
    var STAR = R++;
    src[STAR] = "(<|>)?=?\\s*\\*";
    for (i = 0; i < R; i++) {
      debug10(i, src[i]);
      if (!re[i]) {
        re[i] = new RegExp(src[i]);
        safeRe[i] = new RegExp(makeSafeRe(src[i]));
      }
    }
    var i;
    exports.parse = parse2;
    function parse2(version2, options) {
      if (!options || typeof options !== "object") {
        options = {
          loose: !!options,
          includePrerelease: false
        };
      }
      if (version2 instanceof SemVer) {
        return version2;
      }
      if (typeof version2 !== "string") {
        return null;
      }
      if (version2.length > MAX_LENGTH) {
        return null;
      }
      var r = options.loose ? safeRe[LOOSE] : safeRe[FULL];
      if (!r.test(version2)) {
        return null;
      }
      try {
        return new SemVer(version2, options);
      } catch (er) {
        return null;
      }
    }
    exports.valid = valid;
    function valid(version2, options) {
      var v = parse2(version2, options);
      return v ? v.version : null;
    }
    exports.clean = clean;
    function clean(version2, options) {
      var s = parse2(version2.trim().replace(/^[=v]+/, ""), options);
      return s ? s.version : null;
    }
    exports.SemVer = SemVer;
    function SemVer(version2, options) {
      if (!options || typeof options !== "object") {
        options = {
          loose: !!options,
          includePrerelease: false
        };
      }
      if (version2 instanceof SemVer) {
        if (version2.loose === options.loose) {
          return version2;
        } else {
          version2 = version2.version;
        }
      } else if (typeof version2 !== "string") {
        throw new TypeError("Invalid Version: " + version2);
      }
      if (version2.length > MAX_LENGTH) {
        throw new TypeError("version is longer than " + MAX_LENGTH + " characters");
      }
      if (!(this instanceof SemVer)) {
        return new SemVer(version2, options);
      }
      debug10("SemVer", version2, options);
      this.options = options;
      this.loose = !!options.loose;
      var m = version2.trim().match(options.loose ? safeRe[LOOSE] : safeRe[FULL]);
      if (!m) {
        throw new TypeError("Invalid Version: " + version2);
      }
      this.raw = version2;
      this.major = +m[1];
      this.minor = +m[2];
      this.patch = +m[3];
      if (this.major > MAX_SAFE_INTEGER || this.major < 0) {
        throw new TypeError("Invalid major version");
      }
      if (this.minor > MAX_SAFE_INTEGER || this.minor < 0) {
        throw new TypeError("Invalid minor version");
      }
      if (this.patch > MAX_SAFE_INTEGER || this.patch < 0) {
        throw new TypeError("Invalid patch version");
      }
      if (!m[4]) {
        this.prerelease = [];
      } else {
        this.prerelease = m[4].split(".").map(function(id) {
          if (/^[0-9]+$/.test(id)) {
            var num = +id;
            if (num >= 0 && num < MAX_SAFE_INTEGER) {
              return num;
            }
          }
          return id;
        });
      }
      this.build = m[5] ? m[5].split(".") : [];
      this.format();
    }
    SemVer.prototype.format = function() {
      this.version = this.major + "." + this.minor + "." + this.patch;
      if (this.prerelease.length) {
        this.version += "-" + this.prerelease.join(".");
      }
      return this.version;
    };
    SemVer.prototype.toString = function() {
      return this.version;
    };
    SemVer.prototype.compare = function(other) {
      debug10("SemVer.compare", this.version, this.options, other);
      if (!(other instanceof SemVer)) {
        other = new SemVer(other, this.options);
      }
      return this.compareMain(other) || this.comparePre(other);
    };
    SemVer.prototype.compareMain = function(other) {
      if (!(other instanceof SemVer)) {
        other = new SemVer(other, this.options);
      }
      return compareIdentifiers(this.major, other.major) || compareIdentifiers(this.minor, other.minor) || compareIdentifiers(this.patch, other.patch);
    };
    SemVer.prototype.comparePre = function(other) {
      if (!(other instanceof SemVer)) {
        other = new SemVer(other, this.options);
      }
      if (this.prerelease.length && !other.prerelease.length) {
        return -1;
      } else if (!this.prerelease.length && other.prerelease.length) {
        return 1;
      } else if (!this.prerelease.length && !other.prerelease.length) {
        return 0;
      }
      var i2 = 0;
      do {
        var a = this.prerelease[i2];
        var b = other.prerelease[i2];
        debug10("prerelease compare", i2, a, b);
        if (a === void 0 && b === void 0) {
          return 0;
        } else if (b === void 0) {
          return 1;
        } else if (a === void 0) {
          return -1;
        } else if (a === b) {
          continue;
        } else {
          return compareIdentifiers(a, b);
        }
      } while (++i2);
    };
    SemVer.prototype.inc = function(release, identifier) {
      switch (release) {
        case "premajor":
          this.prerelease.length = 0;
          this.patch = 0;
          this.minor = 0;
          this.major++;
          this.inc("pre", identifier);
          break;
        case "preminor":
          this.prerelease.length = 0;
          this.patch = 0;
          this.minor++;
          this.inc("pre", identifier);
          break;
        case "prepatch":
          this.prerelease.length = 0;
          this.inc("patch", identifier);
          this.inc("pre", identifier);
          break;
        case "prerelease":
          if (this.prerelease.length === 0) {
            this.inc("patch", identifier);
          }
          this.inc("pre", identifier);
          break;
        case "major":
          if (this.minor !== 0 || this.patch !== 0 || this.prerelease.length === 0) {
            this.major++;
          }
          this.minor = 0;
          this.patch = 0;
          this.prerelease = [];
          break;
        case "minor":
          if (this.patch !== 0 || this.prerelease.length === 0) {
            this.minor++;
          }
          this.patch = 0;
          this.prerelease = [];
          break;
        case "patch":
          if (this.prerelease.length === 0) {
            this.patch++;
          }
          this.prerelease = [];
          break;
        case "pre":
          if (this.prerelease.length === 0) {
            this.prerelease = [0];
          } else {
            var i2 = this.prerelease.length;
            while (--i2 >= 0) {
              if (typeof this.prerelease[i2] === "number") {
                this.prerelease[i2]++;
                i2 = -2;
              }
            }
            if (i2 === -1) {
              this.prerelease.push(0);
            }
          }
          if (identifier) {
            if (this.prerelease[0] === identifier) {
              if (isNaN(this.prerelease[1])) {
                this.prerelease = [identifier, 0];
              }
            } else {
              this.prerelease = [identifier, 0];
            }
          }
          break;
        default:
          throw new Error("invalid increment argument: " + release);
      }
      this.format();
      this.raw = this.version;
      return this;
    };
    exports.inc = inc;
    function inc(version2, release, loose, identifier) {
      if (typeof loose === "string") {
        identifier = loose;
        loose = void 0;
      }
      try {
        return new SemVer(version2, loose).inc(release, identifier).version;
      } catch (er) {
        return null;
      }
    }
    exports.diff = diff;
    function diff(version1, version2) {
      if (eq(version1, version2)) {
        return null;
      } else {
        var v1 = parse2(version1);
        var v2 = parse2(version2);
        var prefix = "";
        if (v1.prerelease.length || v2.prerelease.length) {
          prefix = "pre";
          var defaultResult = "prerelease";
        }
        for (var key in v1) {
          if (key === "major" || key === "minor" || key === "patch") {
            if (v1[key] !== v2[key]) {
              return prefix + key;
            }
          }
        }
        return defaultResult;
      }
    }
    exports.compareIdentifiers = compareIdentifiers;
    var numeric = /^[0-9]+$/;
    function compareIdentifiers(a, b) {
      var anum = numeric.test(a);
      var bnum = numeric.test(b);
      if (anum && bnum) {
        a = +a;
        b = +b;
      }
      return a === b ? 0 : anum && !bnum ? -1 : bnum && !anum ? 1 : a < b ? -1 : 1;
    }
    exports.rcompareIdentifiers = rcompareIdentifiers;
    function rcompareIdentifiers(a, b) {
      return compareIdentifiers(b, a);
    }
    exports.major = major;
    function major(a, loose) {
      return new SemVer(a, loose).major;
    }
    exports.minor = minor;
    function minor(a, loose) {
      return new SemVer(a, loose).minor;
    }
    exports.patch = patch;
    function patch(a, loose) {
      return new SemVer(a, loose).patch;
    }
    exports.compare = compare;
    function compare(a, b, loose) {
      return new SemVer(a, loose).compare(new SemVer(b, loose));
    }
    exports.compareLoose = compareLoose;
    function compareLoose(a, b) {
      return compare(a, b, true);
    }
    exports.rcompare = rcompare;
    function rcompare(a, b, loose) {
      return compare(b, a, loose);
    }
    exports.sort = sort;
    function sort(list, loose) {
      return list.sort(function(a, b) {
        return exports.compare(a, b, loose);
      });
    }
    exports.rsort = rsort;
    function rsort(list, loose) {
      return list.sort(function(a, b) {
        return exports.rcompare(a, b, loose);
      });
    }
    exports.gt = gt;
    function gt(a, b, loose) {
      return compare(a, b, loose) > 0;
    }
    exports.lt = lt;
    function lt(a, b, loose) {
      return compare(a, b, loose) < 0;
    }
    exports.eq = eq;
    function eq(a, b, loose) {
      return compare(a, b, loose) === 0;
    }
    exports.neq = neq;
    function neq(a, b, loose) {
      return compare(a, b, loose) !== 0;
    }
    exports.gte = gte;
    function gte(a, b, loose) {
      return compare(a, b, loose) >= 0;
    }
    exports.lte = lte;
    function lte(a, b, loose) {
      return compare(a, b, loose) <= 0;
    }
    exports.cmp = cmp;
    function cmp(a, op, b, loose) {
      switch (op) {
        case "===":
          if (typeof a === "object")
            a = a.version;
          if (typeof b === "object")
            b = b.version;
          return a === b;
        case "!==":
          if (typeof a === "object")
            a = a.version;
          if (typeof b === "object")
            b = b.version;
          return a !== b;
        case "":
        case "=":
        case "==":
          return eq(a, b, loose);
        case "!=":
          return neq(a, b, loose);
        case ">":
          return gt(a, b, loose);
        case ">=":
          return gte(a, b, loose);
        case "<":
          return lt(a, b, loose);
        case "<=":
          return lte(a, b, loose);
        default:
          throw new TypeError("Invalid operator: " + op);
      }
    }
    exports.Comparator = Comparator;
    function Comparator(comp, options) {
      if (!options || typeof options !== "object") {
        options = {
          loose: !!options,
          includePrerelease: false
        };
      }
      if (comp instanceof Comparator) {
        if (comp.loose === !!options.loose) {
          return comp;
        } else {
          comp = comp.value;
        }
      }
      if (!(this instanceof Comparator)) {
        return new Comparator(comp, options);
      }
      comp = comp.trim().split(/\s+/).join(" ");
      debug10("comparator", comp, options);
      this.options = options;
      this.loose = !!options.loose;
      this.parse(comp);
      if (this.semver === ANY) {
        this.value = "";
      } else {
        this.value = this.operator + this.semver.version;
      }
      debug10("comp", this);
    }
    var ANY = {};
    Comparator.prototype.parse = function(comp) {
      var r = this.options.loose ? safeRe[COMPARATORLOOSE] : safeRe[COMPARATOR];
      var m = comp.match(r);
      if (!m) {
        throw new TypeError("Invalid comparator: " + comp);
      }
      this.operator = m[1];
      if (this.operator === "=") {
        this.operator = "";
      }
      if (!m[2]) {
        this.semver = ANY;
      } else {
        this.semver = new SemVer(m[2], this.options.loose);
      }
    };
    Comparator.prototype.toString = function() {
      return this.value;
    };
    Comparator.prototype.test = function(version2) {
      debug10("Comparator.test", version2, this.options.loose);
      if (this.semver === ANY) {
        return true;
      }
      if (typeof version2 === "string") {
        version2 = new SemVer(version2, this.options);
      }
      return cmp(version2, this.operator, this.semver, this.options);
    };
    Comparator.prototype.intersects = function(comp, options) {
      if (!(comp instanceof Comparator)) {
        throw new TypeError("a Comparator is required");
      }
      if (!options || typeof options !== "object") {
        options = {
          loose: !!options,
          includePrerelease: false
        };
      }
      var rangeTmp;
      if (this.operator === "") {
        rangeTmp = new Range(comp.value, options);
        return satisfies(this.value, rangeTmp, options);
      } else if (comp.operator === "") {
        rangeTmp = new Range(this.value, options);
        return satisfies(comp.semver, rangeTmp, options);
      }
      var sameDirectionIncreasing = (this.operator === ">=" || this.operator === ">") && (comp.operator === ">=" || comp.operator === ">");
      var sameDirectionDecreasing = (this.operator === "<=" || this.operator === "<") && (comp.operator === "<=" || comp.operator === "<");
      var sameSemVer = this.semver.version === comp.semver.version;
      var differentDirectionsInclusive = (this.operator === ">=" || this.operator === "<=") && (comp.operator === ">=" || comp.operator === "<=");
      var oppositeDirectionsLessThan = cmp(this.semver, "<", comp.semver, options) && ((this.operator === ">=" || this.operator === ">") && (comp.operator === "<=" || comp.operator === "<"));
      var oppositeDirectionsGreaterThan = cmp(this.semver, ">", comp.semver, options) && ((this.operator === "<=" || this.operator === "<") && (comp.operator === ">=" || comp.operator === ">"));
      return sameDirectionIncreasing || sameDirectionDecreasing || sameSemVer && differentDirectionsInclusive || oppositeDirectionsLessThan || oppositeDirectionsGreaterThan;
    };
    exports.Range = Range;
    function Range(range, options) {
      if (!options || typeof options !== "object") {
        options = {
          loose: !!options,
          includePrerelease: false
        };
      }
      if (range instanceof Range) {
        if (range.loose === !!options.loose && range.includePrerelease === !!options.includePrerelease) {
          return range;
        } else {
          return new Range(range.raw, options);
        }
      }
      if (range instanceof Comparator) {
        return new Range(range.value, options);
      }
      if (!(this instanceof Range)) {
        return new Range(range, options);
      }
      this.options = options;
      this.loose = !!options.loose;
      this.includePrerelease = !!options.includePrerelease;
      this.raw = range.trim().split(/\s+/).join(" ");
      this.set = this.raw.split("||").map(function(range2) {
        return this.parseRange(range2.trim());
      }, this).filter(function(c) {
        return c.length;
      });
      if (!this.set.length) {
        throw new TypeError("Invalid SemVer Range: " + this.raw);
      }
      this.format();
    }
    Range.prototype.format = function() {
      this.range = this.set.map(function(comps) {
        return comps.join(" ").trim();
      }).join("||").trim();
      return this.range;
    };
    Range.prototype.toString = function() {
      return this.range;
    };
    Range.prototype.parseRange = function(range) {
      var loose = this.options.loose;
      var hr = loose ? safeRe[HYPHENRANGELOOSE] : safeRe[HYPHENRANGE];
      range = range.replace(hr, hyphenReplace);
      debug10("hyphen replace", range);
      range = range.replace(safeRe[COMPARATORTRIM], comparatorTrimReplace);
      debug10("comparator trim", range, safeRe[COMPARATORTRIM]);
      range = range.replace(safeRe[TILDETRIM], tildeTrimReplace);
      range = range.replace(safeRe[CARETTRIM], caretTrimReplace);
      var compRe = loose ? safeRe[COMPARATORLOOSE] : safeRe[COMPARATOR];
      var set = range.split(" ").map(function(comp) {
        return parseComparator(comp, this.options);
      }, this).join(" ").split(/\s+/);
      if (this.options.loose) {
        set = set.filter(function(comp) {
          return !!comp.match(compRe);
        });
      }
      set = set.map(function(comp) {
        return new Comparator(comp, this.options);
      }, this);
      return set;
    };
    Range.prototype.intersects = function(range, options) {
      if (!(range instanceof Range)) {
        throw new TypeError("a Range is required");
      }
      return this.set.some(function(thisComparators) {
        return thisComparators.every(function(thisComparator) {
          return range.set.some(function(rangeComparators) {
            return rangeComparators.every(function(rangeComparator) {
              return thisComparator.intersects(rangeComparator, options);
            });
          });
        });
      });
    };
    exports.toComparators = toComparators;
    function toComparators(range, options) {
      return new Range(range, options).set.map(function(comp) {
        return comp.map(function(c) {
          return c.value;
        }).join(" ").trim().split(" ");
      });
    }
    function parseComparator(comp, options) {
      debug10("comp", comp, options);
      comp = replaceCarets(comp, options);
      debug10("caret", comp);
      comp = replaceTildes(comp, options);
      debug10("tildes", comp);
      comp = replaceXRanges(comp, options);
      debug10("xrange", comp);
      comp = replaceStars(comp, options);
      debug10("stars", comp);
      return comp;
    }
    function isX(id) {
      return !id || id.toLowerCase() === "x" || id === "*";
    }
    function replaceTildes(comp, options) {
      return comp.trim().split(/\s+/).map(function(comp2) {
        return replaceTilde(comp2, options);
      }).join(" ");
    }
    function replaceTilde(comp, options) {
      var r = options.loose ? safeRe[TILDELOOSE] : safeRe[TILDE];
      return comp.replace(r, function(_, M, m, p, pr) {
        debug10("tilde", comp, _, M, m, p, pr);
        var ret;
        if (isX(M)) {
          ret = "";
        } else if (isX(m)) {
          ret = ">=" + M + ".0.0 <" + (+M + 1) + ".0.0";
        } else if (isX(p)) {
          ret = ">=" + M + "." + m + ".0 <" + M + "." + (+m + 1) + ".0";
        } else if (pr) {
          debug10("replaceTilde pr", pr);
          ret = ">=" + M + "." + m + "." + p + "-" + pr + " <" + M + "." + (+m + 1) + ".0";
        } else {
          ret = ">=" + M + "." + m + "." + p + " <" + M + "." + (+m + 1) + ".0";
        }
        debug10("tilde return", ret);
        return ret;
      });
    }
    function replaceCarets(comp, options) {
      return comp.trim().split(/\s+/).map(function(comp2) {
        return replaceCaret(comp2, options);
      }).join(" ");
    }
    function replaceCaret(comp, options) {
      debug10("caret", comp, options);
      var r = options.loose ? safeRe[CARETLOOSE] : safeRe[CARET];
      return comp.replace(r, function(_, M, m, p, pr) {
        debug10("caret", comp, _, M, m, p, pr);
        var ret;
        if (isX(M)) {
          ret = "";
        } else if (isX(m)) {
          ret = ">=" + M + ".0.0 <" + (+M + 1) + ".0.0";
        } else if (isX(p)) {
          if (M === "0") {
            ret = ">=" + M + "." + m + ".0 <" + M + "." + (+m + 1) + ".0";
          } else {
            ret = ">=" + M + "." + m + ".0 <" + (+M + 1) + ".0.0";
          }
        } else if (pr) {
          debug10("replaceCaret pr", pr);
          if (M === "0") {
            if (m === "0") {
              ret = ">=" + M + "." + m + "." + p + "-" + pr + " <" + M + "." + m + "." + (+p + 1);
            } else {
              ret = ">=" + M + "." + m + "." + p + "-" + pr + " <" + M + "." + (+m + 1) + ".0";
            }
          } else {
            ret = ">=" + M + "." + m + "." + p + "-" + pr + " <" + (+M + 1) + ".0.0";
          }
        } else {
          debug10("no pr");
          if (M === "0") {
            if (m === "0") {
              ret = ">=" + M + "." + m + "." + p + " <" + M + "." + m + "." + (+p + 1);
            } else {
              ret = ">=" + M + "." + m + "." + p + " <" + M + "." + (+m + 1) + ".0";
            }
          } else {
            ret = ">=" + M + "." + m + "." + p + " <" + (+M + 1) + ".0.0";
          }
        }
        debug10("caret return", ret);
        return ret;
      });
    }
    function replaceXRanges(comp, options) {
      debug10("replaceXRanges", comp, options);
      return comp.split(/\s+/).map(function(comp2) {
        return replaceXRange(comp2, options);
      }).join(" ");
    }
    function replaceXRange(comp, options) {
      comp = comp.trim();
      var r = options.loose ? safeRe[XRANGELOOSE] : safeRe[XRANGE];
      return comp.replace(r, function(ret, gtlt, M, m, p, pr) {
        debug10("xRange", comp, ret, gtlt, M, m, p, pr);
        var xM = isX(M);
        var xm = xM || isX(m);
        var xp = xm || isX(p);
        var anyX = xp;
        if (gtlt === "=" && anyX) {
          gtlt = "";
        }
        if (xM) {
          if (gtlt === ">" || gtlt === "<") {
            ret = "<0.0.0";
          } else {
            ret = "*";
          }
        } else if (gtlt && anyX) {
          if (xm) {
            m = 0;
          }
          p = 0;
          if (gtlt === ">") {
            gtlt = ">=";
            if (xm) {
              M = +M + 1;
              m = 0;
              p = 0;
            } else {
              m = +m + 1;
              p = 0;
            }
          } else if (gtlt === "<=") {
            gtlt = "<";
            if (xm) {
              M = +M + 1;
            } else {
              m = +m + 1;
            }
          }
          ret = gtlt + M + "." + m + "." + p;
        } else if (xm) {
          ret = ">=" + M + ".0.0 <" + (+M + 1) + ".0.0";
        } else if (xp) {
          ret = ">=" + M + "." + m + ".0 <" + M + "." + (+m + 1) + ".0";
        }
        debug10("xRange return", ret);
        return ret;
      });
    }
    function replaceStars(comp, options) {
      debug10("replaceStars", comp, options);
      return comp.trim().replace(safeRe[STAR], "");
    }
    function hyphenReplace($0, from, fM, fm, fp, fpr, fb, to, tM, tm, tp, tpr, tb) {
      if (isX(fM)) {
        from = "";
      } else if (isX(fm)) {
        from = ">=" + fM + ".0.0";
      } else if (isX(fp)) {
        from = ">=" + fM + "." + fm + ".0";
      } else {
        from = ">=" + from;
      }
      if (isX(tM)) {
        to = "";
      } else if (isX(tm)) {
        to = "<" + (+tM + 1) + ".0.0";
      } else if (isX(tp)) {
        to = "<" + tM + "." + (+tm + 1) + ".0";
      } else if (tpr) {
        to = "<=" + tM + "." + tm + "." + tp + "-" + tpr;
      } else {
        to = "<=" + to;
      }
      return (from + " " + to).trim();
    }
    Range.prototype.test = function(version2) {
      if (!version2) {
        return false;
      }
      if (typeof version2 === "string") {
        version2 = new SemVer(version2, this.options);
      }
      for (var i2 = 0; i2 < this.set.length; i2++) {
        if (testSet(this.set[i2], version2, this.options)) {
          return true;
        }
      }
      return false;
    };
    function testSet(set, version2, options) {
      for (var i2 = 0; i2 < set.length; i2++) {
        if (!set[i2].test(version2)) {
          return false;
        }
      }
      if (version2.prerelease.length && !options.includePrerelease) {
        for (i2 = 0; i2 < set.length; i2++) {
          debug10(set[i2].semver);
          if (set[i2].semver === ANY) {
            continue;
          }
          if (set[i2].semver.prerelease.length > 0) {
            var allowed = set[i2].semver;
            if (allowed.major === version2.major && allowed.minor === version2.minor && allowed.patch === version2.patch) {
              return true;
            }
          }
        }
        return false;
      }
      return true;
    }
    exports.satisfies = satisfies;
    function satisfies(version2, range, options) {
      try {
        range = new Range(range, options);
      } catch (er) {
        return false;
      }
      return range.test(version2);
    }
    exports.maxSatisfying = maxSatisfying;
    function maxSatisfying(versions, range, options) {
      var max = null;
      var maxSV = null;
      try {
        var rangeObj = new Range(range, options);
      } catch (er) {
        return null;
      }
      versions.forEach(function(v) {
        if (rangeObj.test(v)) {
          if (!max || maxSV.compare(v) === -1) {
            max = v;
            maxSV = new SemVer(max, options);
          }
        }
      });
      return max;
    }
    exports.minSatisfying = minSatisfying;
    function minSatisfying(versions, range, options) {
      var min = null;
      var minSV = null;
      try {
        var rangeObj = new Range(range, options);
      } catch (er) {
        return null;
      }
      versions.forEach(function(v) {
        if (rangeObj.test(v)) {
          if (!min || minSV.compare(v) === 1) {
            min = v;
            minSV = new SemVer(min, options);
          }
        }
      });
      return min;
    }
    exports.minVersion = minVersion;
    function minVersion(range, loose) {
      range = new Range(range, loose);
      var minver = new SemVer("0.0.0");
      if (range.test(minver)) {
        return minver;
      }
      minver = new SemVer("0.0.0-0");
      if (range.test(minver)) {
        return minver;
      }
      minver = null;
      for (var i2 = 0; i2 < range.set.length; ++i2) {
        var comparators = range.set[i2];
        comparators.forEach(function(comparator) {
          var compver = new SemVer(comparator.semver.version);
          switch (comparator.operator) {
            case ">":
              if (compver.prerelease.length === 0) {
                compver.patch++;
              } else {
                compver.prerelease.push(0);
              }
              compver.raw = compver.format();
            case "":
            case ">=":
              if (!minver || gt(minver, compver)) {
                minver = compver;
              }
              break;
            case "<":
            case "<=":
              break;
            default:
              throw new Error("Unexpected operation: " + comparator.operator);
          }
        });
      }
      if (minver && range.test(minver)) {
        return minver;
      }
      return null;
    }
    exports.validRange = validRange;
    function validRange(range, options) {
      try {
        return new Range(range, options).range || "*";
      } catch (er) {
        return null;
      }
    }
    exports.ltr = ltr;
    function ltr(version2, range, options) {
      return outside(version2, range, "<", options);
    }
    exports.gtr = gtr;
    function gtr(version2, range, options) {
      return outside(version2, range, ">", options);
    }
    exports.outside = outside;
    function outside(version2, range, hilo, options) {
      version2 = new SemVer(version2, options);
      range = new Range(range, options);
      var gtfn, ltefn, ltfn, comp, ecomp;
      switch (hilo) {
        case ">":
          gtfn = gt;
          ltefn = lte;
          ltfn = lt;
          comp = ">";
          ecomp = ">=";
          break;
        case "<":
          gtfn = lt;
          ltefn = gte;
          ltfn = gt;
          comp = "<";
          ecomp = "<=";
          break;
        default:
          throw new TypeError('Must provide a hilo val of "<" or ">"');
      }
      if (satisfies(version2, range, options)) {
        return false;
      }
      for (var i2 = 0; i2 < range.set.length; ++i2) {
        var comparators = range.set[i2];
        var high = null;
        var low = null;
        comparators.forEach(function(comparator) {
          if (comparator.semver === ANY) {
            comparator = new Comparator(">=0.0.0");
          }
          high = high || comparator;
          low = low || comparator;
          if (gtfn(comparator.semver, high.semver, options)) {
            high = comparator;
          } else if (ltfn(comparator.semver, low.semver, options)) {
            low = comparator;
          }
        });
        if (high.operator === comp || high.operator === ecomp) {
          return false;
        }
        if ((!low.operator || low.operator === comp) && ltefn(version2, low.semver)) {
          return false;
        } else if (low.operator === ecomp && ltfn(version2, low.semver)) {
          return false;
        }
      }
      return true;
    }
    exports.prerelease = prerelease;
    function prerelease(version2, options) {
      var parsed = parse2(version2, options);
      return parsed && parsed.prerelease.length ? parsed.prerelease : null;
    }
    exports.intersects = intersects;
    function intersects(r1, r2, options) {
      r1 = new Range(r1, options);
      r2 = new Range(r2, options);
      return r1.intersects(r2);
    }
    exports.coerce = coerce;
    function coerce(version2) {
      if (version2 instanceof SemVer) {
        return version2;
      }
      if (typeof version2 !== "string") {
        return null;
      }
      var match = version2.match(safeRe[COERCE]);
      if (match == null) {
        return null;
      }
      return parse2(match[1] + "." + (match[2] || "0") + "." + (match[3] || "0"));
    }
  }
});

// ../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/parse.js
var require_parse = __commonJS({
  "../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/parse.js"(exports, module2) {
    "use strict";
    var path = require("path");
    var niceTry = require_src();
    var resolveCommand = require_resolveCommand();
    var escape = require_escape();
    var readShebang = require_readShebang();
    var semver = require_semver();
    var isWin3 = process.platform === "win32";
    var isExecutableRegExp = /\.(?:com|exe)$/i;
    var isCmdShimRegExp = /node_modules[\\/].bin[\\/][^\\/]+\.cmd$/i;
    var supportsShellOption = niceTry(() => semver.satisfies(process.version, "^4.8.0 || ^5.7.0 || >= 6.0.0", true)) || false;
    function detectShebang(parsed) {
      parsed.file = resolveCommand(parsed);
      const shebang = parsed.file && readShebang(parsed.file);
      if (shebang) {
        parsed.args.unshift(parsed.file);
        parsed.command = shebang;
        return resolveCommand(parsed);
      }
      return parsed.file;
    }
    function parseNonShell(parsed) {
      if (!isWin3) {
        return parsed;
      }
      const commandFile = detectShebang(parsed);
      const needsShell = !isExecutableRegExp.test(commandFile);
      if (parsed.options.forceShell || needsShell) {
        const needsDoubleEscapeMetaChars = isCmdShimRegExp.test(commandFile);
        parsed.command = path.normalize(parsed.command);
        parsed.command = escape.command(parsed.command);
        parsed.args = parsed.args.map((arg) => escape.argument(arg, needsDoubleEscapeMetaChars));
        const shellCommand = [parsed.command].concat(parsed.args).join(" ");
        parsed.args = ["/d", "/s", "/c", `"${shellCommand}"`];
        parsed.command = process.env.comspec || "cmd.exe";
        parsed.options.windowsVerbatimArguments = true;
      }
      return parsed;
    }
    function parseShell(parsed) {
      if (supportsShellOption) {
        return parsed;
      }
      const shellCommand = [parsed.command].concat(parsed.args).join(" ");
      if (isWin3) {
        parsed.command = typeof parsed.options.shell === "string" ? parsed.options.shell : process.env.comspec || "cmd.exe";
        parsed.args = ["/d", "/s", "/c", `"${shellCommand}"`];
        parsed.options.windowsVerbatimArguments = true;
      } else {
        if (typeof parsed.options.shell === "string") {
          parsed.command = parsed.options.shell;
        } else if (process.platform === "android") {
          parsed.command = "/system/bin/sh";
        } else {
          parsed.command = "/bin/sh";
        }
        parsed.args = ["-c", shellCommand];
      }
      return parsed;
    }
    function parse2(command, args, options) {
      if (args && !Array.isArray(args)) {
        options = args;
        args = null;
      }
      args = args ? args.slice(0) : [];
      options = Object.assign({}, options);
      const parsed = {
        command,
        args,
        options,
        file: void 0,
        original: {
          command,
          args
        }
      };
      return options.shell ? parseShell(parsed) : parseNonShell(parsed);
    }
    module2.exports = parse2;
  }
});

// ../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/enoent.js
var require_enoent = __commonJS({
  "../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/lib/enoent.js"(exports, module2) {
    "use strict";
    var isWin3 = process.platform === "win32";
    function notFoundError(original, syscall) {
      return Object.assign(new Error(`${syscall} ${original.command} ENOENT`), {
        code: "ENOENT",
        errno: "ENOENT",
        syscall: `${syscall} ${original.command}`,
        path: original.command,
        spawnargs: original.args
      });
    }
    function hookChildProcess(cp, parsed) {
      if (!isWin3) {
        return;
      }
      const originalEmit = cp.emit;
      cp.emit = function(name, arg1) {
        if (name === "exit") {
          const err = verifyENOENT(arg1, parsed, "spawn");
          if (err) {
            return originalEmit.call(cp, "error", err);
          }
        }
        return originalEmit.apply(cp, arguments);
      };
    }
    function verifyENOENT(status, parsed) {
      if (isWin3 && status === 1 && !parsed.file) {
        return notFoundError(parsed.original, "spawn");
      }
      return null;
    }
    function verifyENOENTSync(status, parsed) {
      if (isWin3 && status === 1 && !parsed.file) {
        return notFoundError(parsed.original, "spawnSync");
      }
      return null;
    }
    module2.exports = {
      hookChildProcess,
      verifyENOENT,
      verifyENOENTSync,
      notFoundError
    };
  }
});

// ../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/index.js
var require_cross_spawn = __commonJS({
  "../../node_modules/.pnpm/cross-spawn@6.0.5/node_modules/cross-spawn/index.js"(exports, module2) {
    "use strict";
    var cp = require("child_process");
    var parse2 = require_parse();
    var enoent = require_enoent();
    function spawn2(command, args, options) {
      const parsed = parse2(command, args, options);
      const spawned = cp.spawn(parsed.command, parsed.args, parsed.options);
      enoent.hookChildProcess(spawned, parsed);
      return spawned;
    }
    function spawnSync(command, args, options) {
      const parsed = parse2(command, args, options);
      const result = cp.spawnSync(parsed.command, parsed.args, parsed.options);
      result.error = result.error || enoent.verifyENOENTSync(result.status, parsed);
      return result;
    }
    module2.exports = spawn2;
    module2.exports.spawn = spawn2;
    module2.exports.sync = spawnSync;
    module2.exports._parse = parse2;
    module2.exports._enoent = enoent;
  }
});

// ../../node_modules/.pnpm/strip-eof@1.0.0/node_modules/strip-eof/index.js
var require_strip_eof = __commonJS({
  "../../node_modules/.pnpm/strip-eof@1.0.0/node_modules/strip-eof/index.js"(exports, module2) {
    "use strict";
    module2.exports = function(x) {
      var lf = typeof x === "string" ? "\n" : "\n".charCodeAt();
      var cr = typeof x === "string" ? "\r" : "\r".charCodeAt();
      if (x[x.length - 1] === lf) {
        x = x.slice(0, x.length - 1);
      }
      if (x[x.length - 1] === cr) {
        x = x.slice(0, x.length - 1);
      }
      return x;
    };
  }
});

// ../../node_modules/.pnpm/npm-run-path@2.0.2/node_modules/npm-run-path/index.js
var require_npm_run_path = __commonJS({
  "../../node_modules/.pnpm/npm-run-path@2.0.2/node_modules/npm-run-path/index.js"(exports, module2) {
    "use strict";
    var path = require("path");
    var pathKey = require_path_key();
    module2.exports = (opts) => {
      opts = Object.assign({
        cwd: process.cwd(),
        path: process.env[pathKey()]
      }, opts);
      let prev;
      let pth = path.resolve(opts.cwd);
      const ret = [];
      while (prev !== pth) {
        ret.push(path.join(pth, "node_modules/.bin"));
        prev = pth;
        pth = path.resolve(pth, "..");
      }
      ret.push(path.dirname(process.execPath));
      return ret.concat(opts.path).join(path.delimiter);
    };
    module2.exports.env = (opts) => {
      opts = Object.assign({
        env: process.env
      }, opts);
      const env = Object.assign({}, opts.env);
      const path2 = pathKey({ env });
      opts.path = env[path2];
      env[path2] = module2.exports(opts);
      return env;
    };
  }
});

// ../../node_modules/.pnpm/is-stream@1.1.0/node_modules/is-stream/index.js
var require_is_stream = __commonJS({
  "../../node_modules/.pnpm/is-stream@1.1.0/node_modules/is-stream/index.js"(exports, module2) {
    "use strict";
    var isStream = module2.exports = function(stream) {
      return stream !== null && typeof stream === "object" && typeof stream.pipe === "function";
    };
    isStream.writable = function(stream) {
      return isStream(stream) && stream.writable !== false && typeof stream._write === "function" && typeof stream._writableState === "object";
    };
    isStream.readable = function(stream) {
      return isStream(stream) && stream.readable !== false && typeof stream._read === "function" && typeof stream._readableState === "object";
    };
    isStream.duplex = function(stream) {
      return isStream.writable(stream) && isStream.readable(stream);
    };
    isStream.transform = function(stream) {
      return isStream.duplex(stream) && typeof stream._transform === "function" && typeof stream._transformState === "object";
    };
  }
});

// ../../node_modules/.pnpm/wrappy@1.0.2/node_modules/wrappy/wrappy.js
var require_wrappy = __commonJS({
  "../../node_modules/.pnpm/wrappy@1.0.2/node_modules/wrappy/wrappy.js"(exports, module2) {
    module2.exports = wrappy;
    function wrappy(fn, cb) {
      if (fn && cb)
        return wrappy(fn)(cb);
      if (typeof fn !== "function")
        throw new TypeError("need wrapper function");
      Object.keys(fn).forEach(function(k) {
        wrapper[k] = fn[k];
      });
      return wrapper;
      function wrapper() {
        var args = new Array(arguments.length);
        for (var i = 0; i < args.length; i++) {
          args[i] = arguments[i];
        }
        var ret = fn.apply(this, args);
        var cb2 = args[args.length - 1];
        if (typeof ret === "function" && ret !== cb2) {
          Object.keys(cb2).forEach(function(k) {
            ret[k] = cb2[k];
          });
        }
        return ret;
      }
    }
  }
});

// ../../node_modules/.pnpm/once@1.4.0/node_modules/once/once.js
var require_once = __commonJS({
  "../../node_modules/.pnpm/once@1.4.0/node_modules/once/once.js"(exports, module2) {
    var wrappy = require_wrappy();
    module2.exports = wrappy(once);
    module2.exports.strict = wrappy(onceStrict);
    once.proto = once(function() {
      Object.defineProperty(Function.prototype, "once", {
        value: function() {
          return once(this);
        },
        configurable: true
      });
      Object.defineProperty(Function.prototype, "onceStrict", {
        value: function() {
          return onceStrict(this);
        },
        configurable: true
      });
    });
    function once(fn) {
      var f = function() {
        if (f.called)
          return f.value;
        f.called = true;
        return f.value = fn.apply(this, arguments);
      };
      f.called = false;
      return f;
    }
    function onceStrict(fn) {
      var f = function() {
        if (f.called)
          throw new Error(f.onceError);
        f.called = true;
        return f.value = fn.apply(this, arguments);
      };
      var name = fn.name || "Function wrapped with `once`";
      f.onceError = name + " shouldn't be called more than once";
      f.called = false;
      return f;
    }
  }
});

// ../../node_modules/.pnpm/end-of-stream@1.4.1/node_modules/end-of-stream/index.js
var require_end_of_stream = __commonJS({
  "../../node_modules/.pnpm/end-of-stream@1.4.1/node_modules/end-of-stream/index.js"(exports, module2) {
    var once = require_once();
    var noop = function() {
    };
    var isRequest = function(stream) {
      return stream.setHeader && typeof stream.abort === "function";
    };
    var isChildProcess = function(stream) {
      return stream.stdio && Array.isArray(stream.stdio) && stream.stdio.length === 3;
    };
    var eos = function(stream, opts, callback) {
      if (typeof opts === "function")
        return eos(stream, null, opts);
      if (!opts)
        opts = {};
      callback = once(callback || noop);
      var ws = stream._writableState;
      var rs = stream._readableState;
      var readable = opts.readable || opts.readable !== false && stream.readable;
      var writable = opts.writable || opts.writable !== false && stream.writable;
      var onlegacyfinish = function() {
        if (!stream.writable)
          onfinish();
      };
      var onfinish = function() {
        writable = false;
        if (!readable)
          callback.call(stream);
      };
      var onend = function() {
        readable = false;
        if (!writable)
          callback.call(stream);
      };
      var onexit = function(exitCode) {
        callback.call(stream, exitCode ? new Error("exited with error code: " + exitCode) : null);
      };
      var onerror = function(err) {
        callback.call(stream, err);
      };
      var onclose = function() {
        if (readable && !(rs && rs.ended))
          return callback.call(stream, new Error("premature close"));
        if (writable && !(ws && ws.ended))
          return callback.call(stream, new Error("premature close"));
      };
      var onrequest = function() {
        stream.req.on("finish", onfinish);
      };
      if (isRequest(stream)) {
        stream.on("complete", onfinish);
        stream.on("abort", onclose);
        if (stream.req)
          onrequest();
        else
          stream.on("request", onrequest);
      } else if (writable && !ws) {
        stream.on("end", onlegacyfinish);
        stream.on("close", onlegacyfinish);
      }
      if (isChildProcess(stream))
        stream.on("exit", onexit);
      stream.on("end", onend);
      stream.on("finish", onfinish);
      if (opts.error !== false)
        stream.on("error", onerror);
      stream.on("close", onclose);
      return function() {
        stream.removeListener("complete", onfinish);
        stream.removeListener("abort", onclose);
        stream.removeListener("request", onrequest);
        if (stream.req)
          stream.req.removeListener("finish", onfinish);
        stream.removeListener("end", onlegacyfinish);
        stream.removeListener("close", onlegacyfinish);
        stream.removeListener("finish", onfinish);
        stream.removeListener("exit", onexit);
        stream.removeListener("end", onend);
        stream.removeListener("error", onerror);
        stream.removeListener("close", onclose);
      };
    };
    module2.exports = eos;
  }
});

// ../../node_modules/.pnpm/pump@3.0.2/node_modules/pump/index.js
var require_pump = __commonJS({
  "../../node_modules/.pnpm/pump@3.0.2/node_modules/pump/index.js"(exports, module2) {
    var once = require_once();
    var eos = require_end_of_stream();
    var fs8;
    try {
      fs8 = require("fs");
    } catch (e) {
    }
    var noop = function() {
    };
    var ancient = /^v?\.0/.test(process.version);
    var isFn = function(fn) {
      return typeof fn === "function";
    };
    var isFS = function(stream) {
      if (!ancient)
        return false;
      if (!fs8)
        return false;
      return (stream instanceof (fs8.ReadStream || noop) || stream instanceof (fs8.WriteStream || noop)) && isFn(stream.close);
    };
    var isRequest = function(stream) {
      return stream.setHeader && isFn(stream.abort);
    };
    var destroyer = function(stream, reading, writing, callback) {
      callback = once(callback);
      var closed = false;
      stream.on("close", function() {
        closed = true;
      });
      eos(stream, { readable: reading, writable: writing }, function(err) {
        if (err)
          return callback(err);
        closed = true;
        callback();
      });
      var destroyed = false;
      return function(err) {
        if (closed)
          return;
        if (destroyed)
          return;
        destroyed = true;
        if (isFS(stream))
          return stream.close(noop);
        if (isRequest(stream))
          return stream.abort();
        if (isFn(stream.destroy))
          return stream.destroy();
        callback(err || new Error("stream was destroyed"));
      };
    };
    var call = function(fn) {
      fn();
    };
    var pipe = function(from, to) {
      return from.pipe(to);
    };
    var pump = function() {
      var streams = Array.prototype.slice.call(arguments);
      var callback = isFn(streams[streams.length - 1] || noop) && streams.pop() || noop;
      if (Array.isArray(streams[0]))
        streams = streams[0];
      if (streams.length < 2)
        throw new Error("pump requires two streams per minimum");
      var error;
      var destroys = streams.map(function(stream, i) {
        var reading = i < streams.length - 1;
        var writing = i > 0;
        return destroyer(stream, reading, writing, function(err) {
          if (!error)
            error = err;
          if (err)
            destroys.forEach(call);
          if (reading)
            return;
          destroys.forEach(call);
          callback(error);
        });
      });
      return streams.reduce(pipe);
    };
    module2.exports = pump;
  }
});

// ../../node_modules/.pnpm/get-stream@4.1.0/node_modules/get-stream/buffer-stream.js
var require_buffer_stream = __commonJS({
  "../../node_modules/.pnpm/get-stream@4.1.0/node_modules/get-stream/buffer-stream.js"(exports, module2) {
    "use strict";
    var { PassThrough } = require("stream");
    module2.exports = (options) => {
      options = Object.assign({}, options);
      const { array } = options;
      let { encoding } = options;
      const buffer = encoding === "buffer";
      let objectMode = false;
      if (array) {
        objectMode = !(encoding || buffer);
      } else {
        encoding = encoding || "utf8";
      }
      if (buffer) {
        encoding = null;
      }
      let len = 0;
      const ret = [];
      const stream = new PassThrough({ objectMode });
      if (encoding) {
        stream.setEncoding(encoding);
      }
      stream.on("data", (chunk) => {
        ret.push(chunk);
        if (objectMode) {
          len = ret.length;
        } else {
          len += chunk.length;
        }
      });
      stream.getBufferedValue = () => {
        if (array) {
          return ret;
        }
        return buffer ? Buffer.concat(ret, len) : ret.join("");
      };
      stream.getBufferedLength = () => len;
      return stream;
    };
  }
});

// ../../node_modules/.pnpm/get-stream@4.1.0/node_modules/get-stream/index.js
var require_get_stream = __commonJS({
  "../../node_modules/.pnpm/get-stream@4.1.0/node_modules/get-stream/index.js"(exports, module2) {
    "use strict";
    var pump = require_pump();
    var bufferStream = require_buffer_stream();
    var MaxBufferError = class extends Error {
      constructor() {
        super("maxBuffer exceeded");
        this.name = "MaxBufferError";
      }
    };
    function getStream(inputStream, options) {
      if (!inputStream) {
        return Promise.reject(new Error("Expected a stream"));
      }
      options = Object.assign({ maxBuffer: Infinity }, options);
      const { maxBuffer } = options;
      let stream;
      return new Promise((resolve2, reject) => {
        const rejectPromise = (error) => {
          if (error) {
            error.bufferedData = stream.getBufferedValue();
          }
          reject(error);
        };
        stream = pump(inputStream, bufferStream(options), (error) => {
          if (error) {
            rejectPromise(error);
            return;
          }
          resolve2();
        });
        stream.on("data", () => {
          if (stream.getBufferedLength() > maxBuffer) {
            rejectPromise(new MaxBufferError());
          }
        });
      }).then(() => stream.getBufferedValue());
    }
    module2.exports = getStream;
    module2.exports.buffer = (stream, options) => getStream(stream, Object.assign({}, options, { encoding: "buffer" }));
    module2.exports.array = (stream, options) => getStream(stream, Object.assign({}, options, { array: true }));
    module2.exports.MaxBufferError = MaxBufferError;
  }
});

// ../../node_modules/.pnpm/p-finally@1.0.0/node_modules/p-finally/index.js
var require_p_finally = __commonJS({
  "../../node_modules/.pnpm/p-finally@1.0.0/node_modules/p-finally/index.js"(exports, module2) {
    "use strict";
    module2.exports = (promise, onFinally) => {
      onFinally = onFinally || (() => {
      });
      return promise.then(
        (val) => new Promise((resolve2) => {
          resolve2(onFinally());
        }).then(() => val),
        (err) => new Promise((resolve2) => {
          resolve2(onFinally());
        }).then(() => {
          throw err;
        })
      );
    };
  }
});

// ../../node_modules/.pnpm/signal-exit@3.0.7/node_modules/signal-exit/signals.js
var require_signals = __commonJS({
  "../../node_modules/.pnpm/signal-exit@3.0.7/node_modules/signal-exit/signals.js"(exports, module2) {
    module2.exports = [
      "SIGABRT",
      "SIGALRM",
      "SIGHUP",
      "SIGINT",
      "SIGTERM"
    ];
    if (process.platform !== "win32") {
      module2.exports.push(
        "SIGVTALRM",
        "SIGXCPU",
        "SIGXFSZ",
        "SIGUSR2",
        "SIGTRAP",
        "SIGSYS",
        "SIGQUIT",
        "SIGIOT"
        // should detect profiler and enable/disable accordingly.
        // see #21
        // 'SIGPROF'
      );
    }
    if (process.platform === "linux") {
      module2.exports.push(
        "SIGIO",
        "SIGPOLL",
        "SIGPWR",
        "SIGSTKFLT",
        "SIGUNUSED"
      );
    }
  }
});

// ../../node_modules/.pnpm/signal-exit@3.0.7/node_modules/signal-exit/index.js
var require_signal_exit = __commonJS({
  "../../node_modules/.pnpm/signal-exit@3.0.7/node_modules/signal-exit/index.js"(exports, module2) {
    var process2 = global.process;
    var processOk = function(process3) {
      return process3 && typeof process3 === "object" && typeof process3.removeListener === "function" && typeof process3.emit === "function" && typeof process3.reallyExit === "function" && typeof process3.listeners === "function" && typeof process3.kill === "function" && typeof process3.pid === "number" && typeof process3.on === "function";
    };
    if (!processOk(process2)) {
      module2.exports = function() {
        return function() {
        };
      };
    } else {
      assert = require("assert");
      signals = require_signals();
      isWin3 = /^win/i.test(process2.platform);
      EE = require("events");
      if (typeof EE !== "function") {
        EE = EE.EventEmitter;
      }
      if (process2.__signal_exit_emitter__) {
        emitter = process2.__signal_exit_emitter__;
      } else {
        emitter = process2.__signal_exit_emitter__ = new EE();
        emitter.count = 0;
        emitter.emitted = {};
      }
      if (!emitter.infinite) {
        emitter.setMaxListeners(Infinity);
        emitter.infinite = true;
      }
      module2.exports = function(cb, opts) {
        if (!processOk(global.process)) {
          return function() {
          };
        }
        assert.equal(typeof cb, "function", "a callback must be provided for exit handler");
        if (loaded === false) {
          load();
        }
        var ev = "exit";
        if (opts && opts.alwaysLast) {
          ev = "afterexit";
        }
        var remove = function() {
          emitter.removeListener(ev, cb);
          if (emitter.listeners("exit").length === 0 && emitter.listeners("afterexit").length === 0) {
            unload();
          }
        };
        emitter.on(ev, cb);
        return remove;
      };
      unload = function unload2() {
        if (!loaded || !processOk(global.process)) {
          return;
        }
        loaded = false;
        signals.forEach(function(sig) {
          try {
            process2.removeListener(sig, sigListeners[sig]);
          } catch (er) {
          }
        });
        process2.emit = originalProcessEmit;
        process2.reallyExit = originalProcessReallyExit;
        emitter.count -= 1;
      };
      module2.exports.unload = unload;
      emit = function emit2(event, code, signal) {
        if (emitter.emitted[event]) {
          return;
        }
        emitter.emitted[event] = true;
        emitter.emit(event, code, signal);
      };
      sigListeners = {};
      signals.forEach(function(sig) {
        sigListeners[sig] = function listener() {
          if (!processOk(global.process)) {
            return;
          }
          var listeners = process2.listeners(sig);
          if (listeners.length === emitter.count) {
            unload();
            emit("exit", null, sig);
            emit("afterexit", null, sig);
            if (isWin3 && sig === "SIGHUP") {
              sig = "SIGINT";
            }
            process2.kill(process2.pid, sig);
          }
        };
      });
      module2.exports.signals = function() {
        return signals;
      };
      loaded = false;
      load = function load2() {
        if (loaded || !processOk(global.process)) {
          return;
        }
        loaded = true;
        emitter.count += 1;
        signals = signals.filter(function(sig) {
          try {
            process2.on(sig, sigListeners[sig]);
            return true;
          } catch (er) {
            return false;
          }
        });
        process2.emit = processEmit;
        process2.reallyExit = processReallyExit;
      };
      module2.exports.load = load;
      originalProcessReallyExit = process2.reallyExit;
      processReallyExit = function processReallyExit2(code) {
        if (!processOk(global.process)) {
          return;
        }
        process2.exitCode = code || /* istanbul ignore next */
        0;
        emit("exit", process2.exitCode, null);
        emit("afterexit", process2.exitCode, null);
        originalProcessReallyExit.call(process2, process2.exitCode);
      };
      originalProcessEmit = process2.emit;
      processEmit = function processEmit2(ev, arg) {
        if (ev === "exit" && processOk(global.process)) {
          if (arg !== void 0) {
            process2.exitCode = arg;
          }
          var ret = originalProcessEmit.apply(this, arguments);
          emit("exit", process2.exitCode, null);
          emit("afterexit", process2.exitCode, null);
          return ret;
        } else {
          return originalProcessEmit.apply(this, arguments);
        }
      };
    }
    var assert;
    var signals;
    var isWin3;
    var EE;
    var emitter;
    var unload;
    var emit;
    var sigListeners;
    var loaded;
    var load;
    var originalProcessReallyExit;
    var processReallyExit;
    var originalProcessEmit;
    var processEmit;
  }
});

// ../../node_modules/.pnpm/execa@1.0.0/node_modules/execa/lib/errname.js
var require_errname = __commonJS({
  "../../node_modules/.pnpm/execa@1.0.0/node_modules/execa/lib/errname.js"(exports, module2) {
    "use strict";
    var util = require("util");
    var uv;
    if (typeof util.getSystemErrorName === "function") {
      module2.exports = util.getSystemErrorName;
    } else {
      try {
        uv = process.binding("uv");
        if (typeof uv.errname !== "function") {
          throw new TypeError("uv.errname is not a function");
        }
      } catch (err) {
        console.error("execa/lib/errname: unable to establish process.binding('uv')", err);
        uv = null;
      }
      module2.exports = (code) => errname(uv, code);
    }
    module2.exports.__test__ = errname;
    function errname(uv2, code) {
      if (uv2) {
        return uv2.errname(code);
      }
      if (!(code < 0)) {
        throw new Error("err >= 0");
      }
      return `Unknown system error ${code}`;
    }
  }
});

// ../../node_modules/.pnpm/execa@1.0.0/node_modules/execa/lib/stdio.js
var require_stdio = __commonJS({
  "../../node_modules/.pnpm/execa@1.0.0/node_modules/execa/lib/stdio.js"(exports, module2) {
    "use strict";
    var alias = ["stdin", "stdout", "stderr"];
    var hasAlias = (opts) => alias.some((x) => Boolean(opts[x]));
    module2.exports = (opts) => {
      if (!opts) {
        return null;
      }
      if (opts.stdio && hasAlias(opts)) {
        throw new Error(`It's not possible to provide \`stdio\` in combination with one of ${alias.map((x) => `\`${x}\``).join(", ")}`);
      }
      if (typeof opts.stdio === "string") {
        return opts.stdio;
      }
      const stdio = opts.stdio || [];
      if (!Array.isArray(stdio)) {
        throw new TypeError(`Expected \`stdio\` to be of type \`string\` or \`Array\`, got \`${typeof stdio}\``);
      }
      const result = [];
      const len = Math.max(stdio.length, alias.length);
      for (let i = 0; i < len; i++) {
        let value = null;
        if (stdio[i] !== void 0) {
          value = stdio[i];
        } else if (opts[alias[i]] !== void 0) {
          value = opts[alias[i]];
        }
        result[i] = value;
      }
      return result;
    };
  }
});

// ../../node_modules/.pnpm/execa@1.0.0/node_modules/execa/index.js
var require_execa = __commonJS({
  "../../node_modules/.pnpm/execa@1.0.0/node_modules/execa/index.js"(exports, module2) {
    "use strict";
    var path = require("path");
    var childProcess = require("child_process");
    var crossSpawn = require_cross_spawn();
    var stripEof = require_strip_eof();
    var npmRunPath = require_npm_run_path();
    var isStream = require_is_stream();
    var _getStream = require_get_stream();
    var pFinally = require_p_finally();
    var onExit = require_signal_exit();
    var errname = require_errname();
    var stdio = require_stdio();
    var TEN_MEGABYTES = 1e3 * 1e3 * 10;
    function handleArgs(cmd, args, opts) {
      let parsed;
      opts = Object.assign({
        extendEnv: true,
        env: {}
      }, opts);
      if (opts.extendEnv) {
        opts.env = Object.assign({}, process.env, opts.env);
      }
      if (opts.__winShell === true) {
        delete opts.__winShell;
        parsed = {
          command: cmd,
          args,
          options: opts,
          file: cmd,
          original: {
            cmd,
            args
          }
        };
      } else {
        parsed = crossSpawn._parse(cmd, args, opts);
      }
      opts = Object.assign({
        maxBuffer: TEN_MEGABYTES,
        buffer: true,
        stripEof: true,
        preferLocal: true,
        localDir: parsed.options.cwd || process.cwd(),
        encoding: "utf8",
        reject: true,
        cleanup: true
      }, parsed.options);
      opts.stdio = stdio(opts);
      if (opts.preferLocal) {
        opts.env = npmRunPath.env(Object.assign({}, opts, { cwd: opts.localDir }));
      }
      if (opts.detached) {
        opts.cleanup = false;
      }
      if (process.platform === "win32" && path.basename(parsed.command) === "cmd.exe") {
        parsed.args.unshift("/q");
      }
      return {
        cmd: parsed.command,
        args: parsed.args,
        opts,
        parsed
      };
    }
    function handleInput(spawned, input) {
      if (input === null || input === void 0) {
        return;
      }
      if (isStream(input)) {
        input.pipe(spawned.stdin);
      } else {
        spawned.stdin.end(input);
      }
    }
    function handleOutput(opts, val) {
      if (val && opts.stripEof) {
        val = stripEof(val);
      }
      return val;
    }
    function handleShell(fn, cmd, opts) {
      let file = "/bin/sh";
      let args = ["-c", cmd];
      opts = Object.assign({}, opts);
      if (process.platform === "win32") {
        opts.__winShell = true;
        file = process.env.comspec || "cmd.exe";
        args = ["/s", "/c", `"${cmd}"`];
        opts.windowsVerbatimArguments = true;
      }
      if (opts.shell) {
        file = opts.shell;
        delete opts.shell;
      }
      return fn(file, args, opts);
    }
    function getStream(process2, stream, { encoding, buffer, maxBuffer }) {
      if (!process2[stream]) {
        return null;
      }
      let ret;
      if (!buffer) {
        ret = new Promise((resolve2, reject) => {
          process2[stream].once("end", resolve2).once("error", reject);
        });
      } else if (encoding) {
        ret = _getStream(process2[stream], {
          encoding,
          maxBuffer
        });
      } else {
        ret = _getStream.buffer(process2[stream], { maxBuffer });
      }
      return ret.catch((err) => {
        err.stream = stream;
        err.message = `${stream} ${err.message}`;
        throw err;
      });
    }
    function makeError(result, options) {
      const { stdout, stderr } = result;
      let err = result.error;
      const { code, signal } = result;
      const { parsed, joinedCmd } = options;
      const timedOut = options.timedOut || false;
      if (!err) {
        let output = "";
        if (Array.isArray(parsed.opts.stdio)) {
          if (parsed.opts.stdio[2] !== "inherit") {
            output += output.length > 0 ? stderr : `
${stderr}`;
          }
          if (parsed.opts.stdio[1] !== "inherit") {
            output += `
${stdout}`;
          }
        } else if (parsed.opts.stdio !== "inherit") {
          output = `
${stderr}${stdout}`;
        }
        err = new Error(`Command failed: ${joinedCmd}${output}`);
        err.code = code < 0 ? errname(code) : code;
      }
      err.stdout = stdout;
      err.stderr = stderr;
      err.failed = true;
      err.signal = signal || null;
      err.cmd = joinedCmd;
      err.timedOut = timedOut;
      return err;
    }
    function joinCmd(cmd, args) {
      let joinedCmd = cmd;
      if (Array.isArray(args) && args.length > 0) {
        joinedCmd += " " + args.join(" ");
      }
      return joinedCmd;
    }
    module2.exports = (cmd, args, opts) => {
      const parsed = handleArgs(cmd, args, opts);
      const { encoding, buffer, maxBuffer } = parsed.opts;
      const joinedCmd = joinCmd(cmd, args);
      let spawned;
      try {
        spawned = childProcess.spawn(parsed.cmd, parsed.args, parsed.opts);
      } catch (err) {
        return Promise.reject(err);
      }
      let removeExitHandler;
      if (parsed.opts.cleanup) {
        removeExitHandler = onExit(() => {
          spawned.kill();
        });
      }
      let timeoutId = null;
      let timedOut = false;
      const cleanup = () => {
        if (timeoutId) {
          clearTimeout(timeoutId);
          timeoutId = null;
        }
        if (removeExitHandler) {
          removeExitHandler();
        }
      };
      if (parsed.opts.timeout > 0) {
        timeoutId = setTimeout(() => {
          timeoutId = null;
          timedOut = true;
          spawned.kill(parsed.opts.killSignal);
        }, parsed.opts.timeout);
      }
      const processDone = new Promise((resolve2) => {
        spawned.on("exit", (code, signal) => {
          cleanup();
          resolve2({ code, signal });
        });
        spawned.on("error", (err) => {
          cleanup();
          resolve2({ error: err });
        });
        if (spawned.stdin) {
          spawned.stdin.on("error", (err) => {
            cleanup();
            resolve2({ error: err });
          });
        }
      });
      function destroy() {
        if (spawned.stdout) {
          spawned.stdout.destroy();
        }
        if (spawned.stderr) {
          spawned.stderr.destroy();
        }
      }
      const handlePromise = () => pFinally(Promise.all([
        processDone,
        getStream(spawned, "stdout", { encoding, buffer, maxBuffer }),
        getStream(spawned, "stderr", { encoding, buffer, maxBuffer })
      ]).then((arr) => {
        const result = arr[0];
        result.stdout = arr[1];
        result.stderr = arr[2];
        if (result.error || result.code !== 0 || result.signal !== null) {
          const err = makeError(result, {
            joinedCmd,
            parsed,
            timedOut
          });
          err.killed = err.killed || spawned.killed;
          if (!parsed.opts.reject) {
            return err;
          }
          throw err;
        }
        return {
          stdout: handleOutput(parsed.opts, result.stdout),
          stderr: handleOutput(parsed.opts, result.stderr),
          code: 0,
          failed: false,
          killed: false,
          signal: null,
          cmd: joinedCmd,
          timedOut: false
        };
      }), destroy);
      crossSpawn._enoent.hookChildProcess(spawned, parsed.parsed);
      handleInput(spawned, parsed.opts.input);
      spawned.then = (onfulfilled, onrejected) => handlePromise().then(onfulfilled, onrejected);
      spawned.catch = (onrejected) => handlePromise().catch(onrejected);
      return spawned;
    };
    module2.exports.stdout = (...args) => module2.exports(...args).then((x) => x.stdout);
    module2.exports.stderr = (...args) => module2.exports(...args).then((x) => x.stderr);
    module2.exports.shell = (cmd, opts) => handleShell(module2.exports, cmd, opts);
    module2.exports.sync = (cmd, args, opts) => {
      const parsed = handleArgs(cmd, args, opts);
      const joinedCmd = joinCmd(cmd, args);
      if (isStream(parsed.opts.input)) {
        throw new TypeError("The `input` option cannot be a stream in sync mode");
      }
      const result = childProcess.spawnSync(parsed.cmd, parsed.args, parsed.opts);
      result.code = result.status;
      if (result.error || result.status !== 0 || result.signal !== null) {
        const err = makeError(result, {
          joinedCmd,
          parsed
        });
        if (!parsed.opts.reject) {
          return err;
        }
        throw err;
      }
      return {
        stdout: handleOutput(parsed.opts, result.stdout),
        stderr: handleOutput(parsed.opts, result.stderr),
        code: 0,
        failed: false,
        signal: null,
        cmd: joinedCmd,
        timedOut: false
      };
    };
    module2.exports.shellSync = (cmd, opts) => handleShell(module2.exports.sync, cmd, opts);
  }
});

// ../../node_modules/.pnpm/which@3.0.0/node_modules/which/lib/index.js
var require_lib = __commonJS({
  "../../node_modules/.pnpm/which@3.0.0/node_modules/which/lib/index.js"(exports, module2) {
    var isexe = require_isexe();
    var { join: join11, delimiter: delimiter2, sep: sep2, posix } = require("path");
    var isWindows = process.platform === "win32";
    var rSlash = new RegExp(`[${posix.sep}${sep2 === posix.sep ? "" : sep2}]`.replace(/(\\)/g, "\\$1"));
    var rRel = new RegExp(`^\\.${rSlash.source}`);
    var getNotFoundError = (cmd) => Object.assign(new Error(`not found: ${cmd}`), { code: "ENOENT" });
    var getPathInfo = (cmd, {
      path: optPath = process.env.PATH,
      pathExt: optPathExt = process.env.PATHEXT,
      delimiter: optDelimiter = delimiter2
    }) => {
      const pathEnv = cmd.match(rSlash) ? [""] : [
        // windows always checks the cwd first
        ...isWindows ? [process.cwd()] : [],
        ...(optPath || /* istanbul ignore next: very unusual */
        "").split(optDelimiter)
      ];
      if (isWindows) {
        const pathExtExe = optPathExt || [".EXE", ".CMD", ".BAT", ".COM"].join(optDelimiter);
        const pathExt = pathExtExe.split(optDelimiter);
        if (cmd.includes(".") && pathExt[0] !== "") {
          pathExt.unshift("");
        }
        return { pathEnv, pathExt, pathExtExe };
      }
      return { pathEnv, pathExt: [""] };
    };
    var getPathPart = (raw, cmd) => {
      const pathPart = /^".*"$/.test(raw) ? raw.slice(1, -1) : raw;
      const prefix = !pathPart && rRel.test(cmd) ? cmd.slice(0, 2) : "";
      return prefix + join11(pathPart, cmd);
    };
    var which2 = async (cmd, opt = {}) => {
      const { pathEnv, pathExt, pathExtExe } = getPathInfo(cmd, opt);
      const found = [];
      for (const envPart of pathEnv) {
        const p = getPathPart(envPart, cmd);
        for (const ext of pathExt) {
          const withExt = p + ext;
          const is = await isexe(withExt, { pathExt: pathExtExe, ignoreErrors: true });
          if (is) {
            if (!opt.all) {
              return withExt;
            }
            found.push(withExt);
          }
        }
      }
      if (opt.all && found.length) {
        return found;
      }
      if (opt.nothrow) {
        return null;
      }
      throw getNotFoundError(cmd);
    };
    var whichSync = (cmd, opt = {}) => {
      const { pathEnv, pathExt, pathExtExe } = getPathInfo(cmd, opt);
      const found = [];
      for (const pathEnvPart of pathEnv) {
        const p = getPathPart(pathEnvPart, cmd);
        for (const ext of pathExt) {
          const withExt = p + ext;
          const is = isexe.sync(withExt, { pathExt: pathExtExe, ignoreErrors: true });
          if (is) {
            if (!opt.all) {
              return withExt;
            }
            found.push(withExt);
          }
        }
      }
      if (opt.all && found.length) {
        return found;
      }
      if (opt.nothrow) {
        return null;
      }
      throw getNotFoundError(cmd);
    };
    module2.exports = which2;
    which2.sync = whichSync;
  }
});

// ../../node_modules/.pnpm/get-port@5.1.1/node_modules/get-port/index.js
var require_get_port = __commonJS({
  "../../node_modules/.pnpm/get-port@5.1.1/node_modules/get-port/index.js"(exports, module2) {
    "use strict";
    var net = require("net");
    var Locked = class extends Error {
      constructor(port) {
        super(`${port} is locked`);
      }
    };
    var lockedPorts = {
      old: /* @__PURE__ */ new Set(),
      young: /* @__PURE__ */ new Set()
    };
    var releaseOldLockedPortsIntervalMs = 1e3 * 15;
    var interval;
    var getAvailablePort = (options) => new Promise((resolve2, reject) => {
      const server = net.createServer();
      server.unref();
      server.on("error", reject);
      server.listen(options, () => {
        const { port } = server.address();
        server.close(() => {
          resolve2(port);
        });
      });
    });
    var portCheckSequence = function* (ports) {
      if (ports) {
        yield* ports;
      }
      yield 0;
    };
    module2.exports = async (options) => {
      let ports;
      if (options) {
        ports = typeof options.port === "number" ? [options.port] : options.port;
      }
      if (interval === void 0) {
        interval = setInterval(() => {
          lockedPorts.old = lockedPorts.young;
          lockedPorts.young = /* @__PURE__ */ new Set();
        }, releaseOldLockedPortsIntervalMs);
        if (interval.unref) {
          interval.unref();
        }
      }
      for (const port of portCheckSequence(ports)) {
        try {
          let availablePort = await getAvailablePort({ ...options, port });
          while (lockedPorts.old.has(availablePort) || lockedPorts.young.has(availablePort)) {
            if (port !== 0) {
              throw new Locked(port);
            }
            availablePort = await getAvailablePort({ ...options, port });
          }
          lockedPorts.young.add(availablePort);
          return availablePort;
        } catch (error) {
          if (!["EADDRINUSE", "EACCES"].includes(error.code) && !(error instanceof Locked)) {
            throw error;
          }
        }
      }
      throw new Error("No available ports found");
    };
    module2.exports.makeRange = (from, to) => {
      if (!Number.isInteger(from) || !Number.isInteger(to)) {
        throw new TypeError("`from` and `to` must be integer numbers");
      }
      if (from < 1024 || from > 65535) {
        throw new RangeError("`from` must be between 1024 and 65535");
      }
      if (to < 1024 || to > 65536) {
        throw new RangeError("`to` must be between 1024 and 65536");
      }
      if (to < from) {
        throw new RangeError("`to` must be greater than or equal to `from`");
      }
      const generator = function* (from2, to2) {
        for (let port = from2; port <= to2; port++) {
          yield port;
        }
      };
      return generator(from, to);
    };
  }
});

// ../../node_modules/.pnpm/is-port-reachable@3.1.0/node_modules/is-port-reachable/index.js
var require_is_port_reachable = __commonJS({
  "../../node_modules/.pnpm/is-port-reachable@3.1.0/node_modules/is-port-reachable/index.js"(exports, module2) {
    "use strict";
    var net = require("net");
    module2.exports = async (port, { timeout = 1e3, host } = {}) => {
      const promise = new Promise((resolve2, reject) => {
        const socket = new net.Socket();
        const onError = () => {
          socket.destroy();
          reject();
        };
        socket.setTimeout(timeout);
        socket.once("error", onError);
        socket.once("timeout", onError);
        socket.connect(port, host, () => {
          socket.end();
          resolve2();
        });
      });
      try {
        await promise;
        return true;
      } catch (_) {
        return false;
      }
    };
  }
});

// src/index.ts
var src_exports = {};
__export(src_exports, {
  build: () => build,
  defaultShouldServe: () => defaultShouldServe,
  downloadFilesInWorkPath: () => downloadFilesInWorkPath,
  installRequirement: () => installRequirement,
  installRequirementsFile: () => installRequirementsFile,
  shouldServe: () => shouldServe,
  startDevServer: () => startDevServer,
  version: () => version
});
module.exports = __toCommonJS(src_exports);
var import_fs8 = __toESM(require("fs"));
var import_path11 = require("path");

// src/package-versions.ts
var VERCEL_RUNTIME_VERSION = "0.5.6";
var VERCEL_WORKERS_VERSION = "0.0.12";

// src/index.ts
var import_build_utils12 = require("@vercel/build-utils");

// src/install.ts
var import_execa3 = __toESM(require_execa());
var import_fs3 = __toESM(require("fs"));
var import_path5 = require("path");
var import_build_utils4 = require("@vercel/build-utils");
var import_python_analysis2 = require("@vercel/python-analysis");

// src/utils.ts
var import_fs2 = __toESM(require("fs"));
var import_path3 = require("path");
var import_build_utils2 = require("@vercel/build-utils");
var import_execa2 = __toESM(require_execa());

// src/uv.ts
var import_child_process = require("child_process");
var import_path = require("path");
var import_path2 = require("path");
var import_execa = __toESM(require_execa());
var import_fs = __toESM(require("fs"));
var import_os = __toESM(require("os"));
var import_which = __toESM(require_lib());
var import_build_utils = require("@vercel/build-utils");
var UV_VERSION = "0.9.22";
var UV_PYTHON_PATH_PREFIX = "/uv/python/";
var UV_PYTHON_DOWNLOADS_MODE = "automatic";
var isWin = process.platform === "win32";
var uvExec = isWin ? "uv.exe" : "uv";
function findUvInPath() {
  return import_which.default.sync("uv", { nothrow: true });
}
var UvRunner = class {
  constructor(uvPath) {
    this.uvPath = uvPath;
  }
  getPath() {
    return this.uvPath;
  }
  /**
   * List installed Python versions managed by uv.
   * Excludes system Python.
   */
  listInstalledPythons() {
    let output;
    try {
      output = (0, import_child_process.execSync)(
        `${this.uvPath} python list --only-installed --output-format json`,
        { encoding: "utf8", stdio: ["pipe", "pipe", "pipe"] }
      );
    } catch (err) {
      throw new Error(
        `Failed to run 'uv python list': ${err instanceof Error ? err.message : String(err)}`
      );
    }
    if (!output || output.trim() === "" || output.trim() === "[]") {
      return /* @__PURE__ */ new Set();
    }
    let pyList;
    try {
      pyList = JSON.parse(output);
    } catch (err) {
      throw new Error(
        `Failed to parse 'uv python list' output: ${err instanceof Error ? err.message : String(err)}`
      );
    }
    if (process.env.VERCEL_BUILD_IMAGE) {
      pyList = pyList.filter(
        (entry) => entry.path !== null && entry.path.startsWith(UV_PYTHON_PATH_PREFIX) && entry.implementation === "cpython"
      );
    } else {
      pyList = pyList.filter(
        (entry) => entry.path !== null && entry.implementation === "cpython"
      );
    }
    return new Set(
      pyList.map(
        (entry) => `${entry.version_parts.major}.${entry.version_parts.minor}`
      )
    );
  }
  async sync(options) {
    const { venvPath, projectDir, locked, frozen, noBuild, noInstallProject } = options;
    const args = ["sync", "--active", "--no-dev", "--link-mode", "copy"];
    if (frozen) {
      args.push("--frozen");
    } else if (locked) {
      args.push("--locked");
    }
    if (noBuild) {
      args.push("--no-build");
    }
    if (noInstallProject) {
      args.push("--no-install-project");
    }
    args.push("--no-editable");
    await this.runUvCmd(args, projectDir, venvPath);
  }
  async lock(projectDir, options) {
    const args = ["lock"];
    if (options?.noBuild) {
      args.push("--no-build");
    }
    if (options?.upgrade) {
      args.push("--upgrade");
    }
    const pretty = `uv ${args.join(" ")}`;
    (0, import_build_utils.debug)(`Running "${pretty}" in ${projectDir}...`);
    try {
      await (0, import_execa.default)(this.uvPath, args, {
        cwd: projectDir,
        env: getProtectedUvEnv(process.env)
      });
    } catch (err) {
      const error = new Error(
        `Failed to run "${pretty}": ${err instanceof Error ? err.message : String(err)}`
      );
      if (err && typeof err === "object") {
        if ("code" in err) {
          error.code = err.code;
        } else if ("signal" in err) {
          error.code = err.signal;
        }
      }
      throw error;
    }
  }
  async addDependencies(options) {
    const { venvPath, projectDir, dependencies } = options;
    const toAdd = dependencies.filter(Boolean);
    if (!toAdd.length)
      return;
    const args = ["add", "--active", ...toAdd];
    (0, import_build_utils.debug)(`Running "uv ${args.join(" ")}" in ${projectDir}...`);
    await this.runUvCmd(args, projectDir, venvPath);
  }
  async addFromFile(options) {
    const { venvPath, projectDir, requirementsPath } = options;
    const args = ["add", "--active", "-r", requirementsPath];
    (0, import_build_utils.debug)(`Running "uv ${args.join(" ")}" in ${projectDir}...`);
    await this.runUvCmd(args, projectDir, venvPath);
  }
  /**
   * Run a `uv pip` command (e.g., `uv pip install`).
   */
  async pip(options) {
    const { venvPath, projectDir, args } = options;
    const fullArgs = ["pip", ...args];
    await this.runUvCmd(fullArgs, projectDir, venvPath);
  }
  async runUvCmd(args, cwd, venvPath) {
    const pretty = `uv ${args.join(" ")}`;
    (0, import_build_utils.debug)(`Running "${pretty}"...`);
    try {
      await (0, import_execa.default)(this.uvPath, args, {
        cwd,
        env: this.getVenvEnv(venvPath)
      });
    } catch (err) {
      const error = new Error(
        `Failed to run "${pretty}": ${err instanceof Error ? err.message : String(err)}`
      );
      if (err && typeof err === "object") {
        if ("code" in err) {
          error.code = err.code;
        } else if ("signal" in err) {
          error.code = err.signal;
        }
      }
      throw error;
    }
  }
  getVenvEnv(venvPath) {
    const binDir = isWin ? (0, import_path.join)(venvPath, "Scripts") : (0, import_path.join)(venvPath, "bin");
    const existingPath = process.env.PATH || "";
    return {
      ...getProtectedUvEnv(process.env),
      VIRTUAL_ENV: venvPath,
      PATH: existingPath ? `${binDir}${import_path2.delimiter}${existingPath}` : binDir
    };
  }
};
async function getGlobalScriptsDir(pythonPath) {
  const code = `import sysconfig; print(sysconfig.get_path('scripts'))`;
  try {
    const { stdout } = await (0, import_execa.default)(pythonPath, ["-c", code]);
    const out = stdout.trim();
    return out || null;
  } catch (err) {
    (0, import_build_utils.debug)("Failed to resolve Python global scripts directory", err);
    return null;
  }
}
async function getUserScriptsDir(pythonPath) {
  const code = `import sys, sysconfig; print(sysconfig.get_path('scripts', scheme=('nt_user' if sys.platform == 'win32' else 'posix_user')))`.replace(
    /\n/g,
    " "
  );
  try {
    const { stdout } = await (0, import_execa.default)(pythonPath, ["-c", code]);
    const out = stdout.trim();
    return out || null;
  } catch (err) {
    (0, import_build_utils.debug)("Failed to resolve Python user scripts directory", err);
    return null;
  }
}
async function findUvBinary(pythonPath) {
  const found = import_which.default.sync("uv", { nothrow: true });
  if (found)
    return found;
  try {
    const globalScriptsDir = await getGlobalScriptsDir(pythonPath);
    if (globalScriptsDir) {
      const uvPath = (0, import_path.join)(globalScriptsDir, uvExec);
      if (import_fs.default.existsSync(uvPath))
        return uvPath;
    }
  } catch (err) {
    (0, import_build_utils.debug)("Failed to resolve Python global scripts directory", err);
  }
  try {
    const userScriptsDir = await getUserScriptsDir(pythonPath);
    if (userScriptsDir) {
      const uvPath = (0, import_path.join)(userScriptsDir, uvExec);
      if (import_fs.default.existsSync(uvPath))
        return uvPath;
    }
  } catch (err) {
    (0, import_build_utils.debug)("Failed to resolve Python user scripts directory", err);
  }
  try {
    const candidates = [];
    if (!isWin) {
      candidates.push((0, import_path.join)(import_os.default.homedir(), ".local", "bin", "uv"));
      candidates.push("/usr/local/bin/uv");
      candidates.push("/opt/homebrew/bin/uv");
    } else {
      candidates.push("C:\\Users\\Public\\uv\\uv.exe");
    }
    for (const p of candidates) {
      if (import_fs.default.existsSync(p))
        return p;
    }
  } catch (err) {
    (0, import_build_utils.debug)("Failed to resolve uv fallback paths", err);
  }
  return null;
}
async function getUvBinaryOrInstall(pythonPath) {
  const uvBin = await findUvBinary(pythonPath);
  if (uvBin)
    return uvBin;
  try {
    console.log("Installing uv...");
    await (0, import_execa.default)(
      pythonPath,
      [
        "-m",
        "pip",
        "install",
        "--disable-pip-version-check",
        "--no-cache-dir",
        "--user",
        `uv==${UV_VERSION}`
      ],
      { env: { ...process.env, PIP_USER: "1" } }
    );
  } catch (err) {
    throw new Error(
      `Failed to install uv via pip: ${err instanceof Error ? err.message : String(err)}`
    );
  }
  const resolvedUvBin = await findUvBinary(pythonPath);
  if (!resolvedUvBin) {
    throw new Error("Unable to resolve uv binary after pip install");
  }
  console.log(`Installed uv at "${resolvedUvBin}"`);
  return resolvedUvBin;
}
function filterUnsafeUvPipArgs(args) {
  return args.filter((arg) => arg !== "--no-warn-script-location");
}
function getProtectedUvEnv(baseEnv = process.env) {
  return {
    ...baseEnv,
    UV_PYTHON_DOWNLOADS: UV_PYTHON_DOWNLOADS_MODE
  };
}
var UV_BUNDLE_DIR = "_uv";
async function getUvBinaryForBundling(pythonPath) {
  const uvPath = await findUvBinary(pythonPath);
  if (!uvPath) {
    throw new Error(
      "Cannot find uv binary for bundling. Ensure uv is installed and available in PATH."
    );
  }
  const resolvedPath = await import_fs.default.promises.realpath(uvPath);
  return resolvedPath;
}

// src/utils.ts
var isWin2 = process.platform === "win32";
var isInVirtualEnv = () => {
  return process.env.VIRTUAL_ENV;
};
function getVenvBinDir(venvPath) {
  return (0, import_path3.join)(venvPath, isWin2 ? "Scripts" : "bin");
}
function useVirtualEnv(workPath, env, systemPython) {
  const venvDirs = [".venv", "venv"];
  let pythonCmd = systemPython;
  for (const venv of venvDirs) {
    const venvRoot = (0, import_path3.join)(workPath, venv);
    const binDir = process.platform === "win32" ? (0, import_path3.join)(venvRoot, "Scripts") : (0, import_path3.join)(venvRoot, "bin");
    const candidates = process.platform === "win32" ? [(0, import_path3.join)(binDir, "python.exe"), (0, import_path3.join)(binDir, "python")] : [(0, import_path3.join)(binDir, "python3"), (0, import_path3.join)(binDir, "python")];
    const found = candidates.find((p) => import_fs2.default.existsSync(p));
    if (found) {
      pythonCmd = found;
      env.VIRTUAL_ENV = venvRoot;
      env.PATH = `${binDir}${import_path3.delimiter}${env.PATH || ""}`;
      return { pythonCmd, venvRoot };
    }
  }
  return { pythonCmd };
}
function createVenvEnv(venvPath, baseEnv = process.env) {
  const env = {
    ...getProtectedUvEnv(baseEnv),
    VIRTUAL_ENV: venvPath
  };
  const binDir = getVenvBinDir(venvPath);
  const existingPath = env.PATH || process.env.PATH || "";
  env.PATH = existingPath ? `${binDir}${import_path3.delimiter}${existingPath}` : binDir;
  return env;
}
async function ensureVenv({
  pythonPath,
  venvPath,
  uvPath,
  quiet
}) {
  const marker = (0, import_path3.join)(venvPath, "pyvenv.cfg");
  try {
    await import_fs2.default.promises.access(marker);
    return;
  } catch {
  }
  await import_fs2.default.promises.mkdir(venvPath, { recursive: true });
  if (!quiet) {
    console.log(`Creating virtual environment at "${venvPath}"...`);
  }
  if (uvPath) {
    await (0, import_execa2.default)(uvPath, ["venv", venvPath]);
  } else {
    await (0, import_execa2.default)(pythonPath, ["-m", "venv", venvPath]);
  }
}
function getVenvPythonBin(venvPath) {
  return (0, import_path3.join)(getVenvBinDir(venvPath), isWin2 ? "python.exe" : "python");
}
async function runPyprojectScript(workPath, scriptNames, env, useUserVirtualEnv = true) {
  const pyprojectPath = (0, import_path3.join)(workPath, "pyproject.toml");
  if (!import_fs2.default.existsSync(pyprojectPath))
    return false;
  let pyproject = null;
  try {
    pyproject = await (0, import_build_utils2.readConfigFile)(pyprojectPath);
  } catch {
    console.error("Failed to parse pyproject.toml");
    return false;
  }
  const scripts = pyproject?.tool?.vercel?.scripts || {};
  const candidates = typeof scriptNames === "string" ? [scriptNames] : Array.from(scriptNames);
  const scriptToRun = candidates.find((name) => Boolean(scripts[name]));
  if (!scriptToRun)
    return false;
  const systemPython = process.platform === "win32" ? "python" : "python3";
  const finalEnv = { ...process.env, ...env };
  if (useUserVirtualEnv) {
    useVirtualEnv(workPath, finalEnv, systemPython);
  }
  const scriptCommand = scripts[scriptToRun];
  if (typeof scriptCommand === "string" && scriptCommand.trim()) {
    console.log(`Executing: ${scriptCommand}`);
    await (0, import_build_utils2.execCommand)(scriptCommand, {
      cwd: workPath,
      env: finalEnv
    });
    return true;
  }
  return false;
}

// src/version.ts
var import_path4 = require("path");
var import_build_utils3 = require("@vercel/build-utils");
var import_python_analysis = require("@vercel/python-analysis");
function pythonVersionString(pv) {
  return `${pv.major}.${pv.minor}`;
}
var DEFAULT_PYTHON_VERSION = makePythonVersion(3, 12);
var DEFAULT_PYTHON_VERSION_STRING = pythonVersionString(
  DEFAULT_PYTHON_VERSION
);
function makePythonVersion(major, minor, discontinueDate) {
  return {
    major,
    minor,
    pipPath: `pip${major}.${minor}`,
    pythonPath: `python${major}.${minor}`,
    runtime: `python${major}.${minor}`,
    discontinueDate
  };
}
var allOptions = [
  makePythonVersion(3, 14),
  makePythonVersion(3, 13),
  makePythonVersion(3, 12),
  makePythonVersion(3, 11),
  makePythonVersion(3, 10),
  makePythonVersion(3, 9),
  makePythonVersion(3, 6, /* @__PURE__ */ new Date("2022-07-18"))
];
function getDevPythonVersion() {
  return {
    major: 3,
    minor: 0,
    pipPath: "pip3",
    pythonPath: "python3",
    runtime: "python3"
  };
}
function getDefaultPythonVersion({
  isDev
}) {
  if (isDev) {
    return getDevPythonVersion();
  }
  const defaultOption = allOptions.find(
    (opt) => versionsEqual(opt, DEFAULT_PYTHON_VERSION) && isInstalled(opt)
  );
  if (defaultOption) {
    return defaultOption;
  }
  const selection = allOptions.find(isInstalled);
  if (!selection) {
    throw new import_build_utils3.NowBuildError({
      code: "PYTHON_NOT_FOUND",
      link: "https://vercel.link/python-version",
      message: `Unable to find any supported Python versions.`
    });
  }
  return selection;
}
function versionsEqual(a, b) {
  return a.major === b.major && a.minor === b.minor;
}
function versionLessOrEqual(a, b) {
  if (a.major !== b.major)
    return a.major < b.major;
  return a.minor <= b.minor;
}
function toPythonBuild(opt) {
  return {
    version: { major: opt.major, minor: opt.minor },
    implementation: "cpython",
    variant: "default",
    os: "linux",
    architecture: "x86_64",
    libc: "gnu"
  };
}
function getAvailablePythonBuilds() {
  const installed = allOptions.filter(
    (opt) => !isDiscontinued(opt) && isInstalled(opt)
  );
  const defaultOpt = installed.find(
    (opt) => versionsEqual(opt, DEFAULT_PYTHON_VERSION)
  );
  const rest = installed.filter(
    (opt) => !versionsEqual(opt, DEFAULT_PYTHON_VERSION)
  );
  const ordered = defaultOpt ? [defaultOpt, ...rest] : rest;
  return ordered.map(toPythonBuild);
}
function getAllPythonBuilds() {
  return allOptions.map(toPythonBuild);
}
function getPythonVersionForBuild(build2) {
  return allOptions.find(
    (opt) => opt.major === build2.version.major && opt.minor === build2.version.minor
  );
}
function resolvePythonVersion({
  isDev,
  pythonPackage,
  rootDir
}) {
  if (isDev) {
    return {
      pythonVersion: getDevPythonVersion(),
      pythonPackage
    };
  }
  const constraints = pythonPackage.requiresPython;
  const defaultPv = getDefaultPythonVersion({ isDev: false });
  let selection;
  let source;
  let autoUpgraded = false;
  if (!constraints || constraints.length === 0) {
    console.log(
      `No Python version specified in .python-version, pyproject.toml, or Pipfile.lock. Using python version: ${pythonVersionString(defaultPv)}`
    );
    selection = defaultPv;
  } else {
    const defaultBuild = toPythonBuild(defaultPv);
    const availableBuilds = getAvailablePythonBuilds();
    const allBuilds = getAllPythonBuilds();
    const result = (0, import_python_analysis.selectPythonVersion)({
      constraints,
      availableBuilds,
      allBuilds,
      defaultBuild,
      majorMinorOnly: true,
      legacyTildeEquals: true
    });
    source = result.source;
    selection = getPythonVersionForBuild(result.build) ?? defaultPv;
    if (pythonPackage.manifest?.origin && !result.notAvailable && !result.invalidConstraint && !versionLessOrEqual(defaultPv, selection)) {
      const originalVersion = pythonVersionString(selection);
      selection = defaultPv;
      autoUpgraded = true;
      console.log(
        `Python version ${originalVersion} detected in ${source} is below the minimum supported version. Using python version: ${pythonVersionString(selection)}`
      );
    } else if (result.notAvailable) {
      const npv = getPythonVersionForBuild(result.notAvailable.build);
      if (npv && isDiscontinued(npv)) {
        throw new import_build_utils3.NowBuildError({
          code: "BUILD_UTILS_PYTHON_VERSION_DISCONTINUED",
          link: "https://vercel.link/python-version",
          message: `Python version "${pythonVersionString(npv)}" detected in ${source} is discontinued and must be upgraded.`
        });
      }
      if (npv) {
        console.warn(
          `Warning: Python version "${pythonVersionString(npv)}" detected in ${source} is not installed and will be ignored. https://vercel.link/python-version`
        );
      }
      console.log(`Using python version: ${pythonVersionString(selection)}`);
    } else if (result.invalidConstraint) {
      console.warn(
        `Warning: Python version "${result.invalidConstraint.versionString}" detected in ${source} is invalid and will be ignored. https://vercel.link/python-version`
      );
      console.log(`Using python version: ${pythonVersionString(selection)}`);
    } else {
      console.log(
        `Using Python ${pythonVersionString(selection)} from ${source}`
      );
    }
  }
  if (isDiscontinued(selection)) {
    throw new import_build_utils3.NowBuildError({
      code: "BUILD_UTILS_PYTHON_VERSION_DISCONTINUED",
      link: "https://vercel.link/python-version",
      message: `Python version "${pythonVersionString(selection)}" declared in project configuration is discontinued and must be upgraded.`
    });
  }
  if (selection.discontinueDate) {
    const d = selection.discontinueDate.toISOString().split("T")[0];
    const srcSuffix = source ? `detected in ${source}` : "selected by runtime";
    console.warn(
      `Error: Python version "${pythonVersionString(selection)}" ${srcSuffix} has reached End-of-Life. Deployments created on or after ${d} will fail to build. https://vercel.link/python-version`
    );
  }
  let pinVersionFilePath;
  const hasPythonVersionFile = pythonPackage.configs?.some(
    (configSet) => configSet[import_python_analysis.PythonConfigKind.PythonVersion] !== void 0
  ) ?? false;
  if (!hasPythonVersionFile && pythonPackage.manifest && versionLessOrEqual(selection, defaultPv)) {
    if (autoUpgraded || !pythonPackage.manifest.origin && source?.endsWith("pyproject.toml")) {
      const manifestDir = (0, import_path4.join)(rootDir, (0, import_path4.dirname)(pythonPackage.manifest.path));
      pinVersionFilePath = (0, import_path4.join)(manifestDir, ".python-version");
    }
  }
  return {
    pythonVersion: selection,
    pythonPackage,
    versionSource: source,
    pinVersionFilePath
  };
}
function isDiscontinued({ discontinueDate }) {
  const today = Date.now();
  return discontinueDate !== void 0 && discontinueDate.getTime() <= today;
}
var installedPythonsCache = null;
function getInstalledPythons() {
  if (installedPythonsCache !== null) {
    return installedPythonsCache;
  }
  const uvPath = findUvInPath();
  if (!uvPath) {
    throw new import_build_utils3.NowBuildError({
      code: "UV_ERROR",
      link: "https://vercel.link/python-version",
      message: "uv is required but was not found in PATH."
    });
  }
  const uv = new UvRunner(uvPath);
  installedPythonsCache = uv.listInstalledPythons();
  return installedPythonsCache;
}
function isInstalled(pv) {
  try {
    const installed = getInstalledPythons();
    return installed.has(pythonVersionString(pv));
  } catch (err) {
    throw new import_build_utils3.NowBuildError({
      code: "UV_ERROR",
      link: "https://vercel.link/python-version",
      message: err instanceof Error ? err.message : String(err)
    });
  }
}

// src/install.ts
var makeDependencyCheckCode = (dependency) => `
from importlib import util
dep = '${dependency}'.replace('-', '_')
spec = util.find_spec(dep)
print(spec.origin)
`;
async function isInstalled2(pythonPath, dependency, cwd) {
  try {
    const { stdout } = await (0, import_execa3.default)(
      pythonPath,
      ["-c", makeDependencyCheckCode(dependency)],
      {
        stdio: "pipe",
        cwd,
        env: { ...process.env, PYTHONPATH: (0, import_path5.join)(cwd, resolveVendorDir()) }
      }
    );
    return stdout.startsWith(cwd);
  } catch (_err) {
    return false;
  }
}
var makeRequirementsCheckCode = (requirementsPath) => `
import distutils.text_file
import pkg_resources
from pkg_resources import DistributionNotFound, VersionConflict
dependencies = distutils.text_file.TextFile(filename='${requirementsPath}').readlines()
pkg_resources.require(dependencies)
`;
async function areRequirementsInstalled(pythonPath, requirementsPath, cwd) {
  try {
    await (0, import_execa3.default)(
      pythonPath,
      ["-c", makeRequirementsCheckCode(requirementsPath)],
      {
        stdio: "pipe",
        cwd,
        env: { ...process.env, PYTHONPATH: (0, import_path5.join)(cwd, resolveVendorDir()) }
      }
    );
    return true;
  } catch (_err) {
    return false;
  }
}
async function getSitePackagesDirs(pythonBin) {
  const code = `
import json
import sysconfig
paths = []
for key in ("purelib", "platlib"):
    candidate = sysconfig.get_path(key)
    if candidate and candidate not in paths:
        paths.append(candidate)
print(json.dumps(paths))
`.trim();
  const { stdout } = await (0, import_execa3.default)(pythonBin, ["-c", code]);
  try {
    const parsed = JSON.parse(stdout);
    if (Array.isArray(parsed)) {
      return parsed.filter((p) => typeof p === "string");
    }
  } catch (err) {
    (0, import_build_utils4.debug)("Failed to parse site-packages output", err);
  }
  return [];
}
async function getVenvSitePackagesDirs(venvPath) {
  const pythonBin = getVenvPythonBin(venvPath);
  return getSitePackagesDirs(pythonBin);
}
function resolveVendorDir() {
  const vendorDir = process.env.VERCEL_PYTHON_VENDOR_DIR || "_vendor";
  return vendorDir;
}
async function discoverPackage({
  entrypointDir,
  rootDir
}) {
  try {
    return await (0, import_python_analysis2.discoverPythonPackage)({ entrypointDir, rootDir });
  } catch (error) {
    if (error instanceof import_python_analysis2.PythonAnalysisError) {
      if (error.fileContent && (error.code.endsWith("_PARSE_ERROR") || error.code.endsWith("_VALIDATION_ERROR"))) {
        console.log(
          `Failed to parse "${error.path}". File content:
${error.fileContent}`
        );
      }
      throw new import_build_utils4.NowBuildError({
        code: error.code,
        message: error.message,
        link: error.link,
        action: error.action
      });
    }
    throw error;
  }
}
function detectInstallSource(pythonPackage, rootDir) {
  let manifestType = null;
  let manifestPath = null;
  const lockFile = pythonPackage.manifest?.lockFile ?? pythonPackage.workspaceLockFile;
  if (lockFile) {
    if (lockFile.kind === import_python_analysis2.PythonLockFileKind.UvLock) {
      manifestType = "uv.lock";
      manifestPath = (0, import_path5.join)(rootDir, lockFile.path);
    } else if (lockFile.kind === import_python_analysis2.PythonLockFileKind.PylockToml) {
      manifestType = "pylock.toml";
      manifestPath = (0, import_path5.join)(rootDir, lockFile.path);
    }
  } else if (pythonPackage.manifest) {
    manifestType = "pyproject.toml";
    manifestPath = (0, import_path5.join)(rootDir, pythonPackage.manifest.path);
  }
  return { manifestPath, manifestType, pythonPackage };
}
async function ensureUvProject({
  workPath,
  rootDir,
  pythonPackage,
  pythonVersion,
  uv,
  generateLockFile = false,
  requireBinaryWheels = false
}) {
  const { manifestType } = detectInstallSource(pythonPackage, rootDir);
  const manifest = pythonPackage.manifest;
  let projectDir;
  let pyprojectPath;
  let lockPath = null;
  let lockFileProvidedByUser = false;
  if (manifestType === "uv.lock" || manifestType === "pylock.toml") {
    lockFileProvidedByUser = true;
    const lockFile = pythonPackage.manifest?.lockFile ?? pythonPackage.workspaceLockFile;
    if (!lockFile) {
      throw new Error(
        `Expected lock file path to be resolved, but it was null`
      );
    }
    lockPath = (0, import_path5.join)(rootDir, lockFile.path);
    projectDir = (0, import_path5.dirname)(lockPath);
    pyprojectPath = (0, import_path5.join)(projectDir, "pyproject.toml");
    if (!import_fs3.default.existsSync(pyprojectPath)) {
      throw new Error(
        `Expected "pyproject.toml" next to "${lockFile.kind}" in "${projectDir}"`
      );
    }
    console.log(`Installing required dependencies from ${lockFile.kind}...`);
  } else if (manifest) {
    projectDir = (0, import_path5.join)(rootDir, (0, import_path5.dirname)(manifest.path));
    pyprojectPath = (0, import_path5.join)(rootDir, manifest.path);
    const originKind = manifest.origin?.kind;
    if (originKind === import_python_analysis2.PythonManifestConvertedKind.Pipfile) {
      console.log("Installing required dependencies from Pipfile...");
    } else if (originKind === import_python_analysis2.PythonManifestConvertedKind.PipfileLock) {
      console.log("Installing required dependencies from Pipfile.lock...");
    } else if (originKind === import_python_analysis2.PythonManifestConvertedKind.RequirementsTxt || originKind === import_python_analysis2.PythonManifestConvertedKind.RequirementsIn) {
      console.log(
        `Installing required dependencies from ${manifest.origin?.path ?? "requirements.txt"}...`
      );
    } else {
      console.log("Installing required dependencies from pyproject.toml...");
    }
    if (manifest.origin) {
      if (manifest.data.project) {
        manifest.data.project["requires-python"] = `~=${pythonVersion}.0`;
      }
      const content = (0, import_python_analysis2.stringifyManifest)(manifest.data);
      pyprojectPath = (0, import_path5.join)(projectDir, "pyproject.toml");
      await import_fs3.default.promises.writeFile(pyprojectPath, content);
    }
    const workspaceLockFile = pythonPackage.workspaceLockFile;
    if (workspaceLockFile) {
      lockPath = (0, import_path5.join)(rootDir, workspaceLockFile.path);
    } else {
      await uv.lock(
        projectDir,
        requireBinaryWheels ? { noBuild: true, upgrade: true } : void 0
      );
    }
    if (generateLockFile && !lockPath) {
      await uv.lock(
        projectDir,
        requireBinaryWheels ? { noBuild: true, upgrade: true } : void 0
      );
    }
  } else {
    projectDir = workPath;
    pyprojectPath = (0, import_path5.join)(projectDir, "pyproject.toml");
    console.log(
      "No Python manifest found; creating an empty pyproject.toml and uv.lock..."
    );
    const requiresPython = `~=${pythonVersion}.0`;
    const minimalManifest = (0, import_python_analysis2.createMinimalManifest)({
      name: "app",
      requiresPython,
      dependencies: []
    });
    const content = (0, import_python_analysis2.stringifyManifest)(minimalManifest);
    await import_fs3.default.promises.writeFile(pyprojectPath, content);
    await uv.lock(
      projectDir,
      requireBinaryWheels ? { noBuild: true, upgrade: true } : void 0
    );
  }
  const resolvedLockPath = lockPath && import_fs3.default.existsSync(lockPath) ? lockPath : (0, import_path5.join)(projectDir, "uv.lock");
  return {
    projectDir,
    pyprojectPath,
    lockPath: resolvedLockPath,
    lockFileProvidedByUser
  };
}
async function pipInstall(pipPath, uvPath, workPath, args, targetDir) {
  const target = targetDir ? (0, import_path5.join)(targetDir, resolveVendorDir()) : resolveVendorDir();
  process.env.PIP_USER = "0";
  if (uvPath) {
    const uvArgs = [
      "pip",
      "install",
      "--no-compile",
      "--no-cache-dir",
      "--target",
      target,
      ...filterUnsafeUvPipArgs(args)
    ];
    const prettyUv = `${uvPath} ${uvArgs.join(" ")}`;
    (0, import_build_utils4.debug)(`Running "${prettyUv}"...`);
    try {
      await (0, import_execa3.default)(uvPath, uvArgs, {
        cwd: workPath,
        env: getProtectedUvEnv()
      });
      return;
    } catch (err) {
      console.log(`Failed to run "${prettyUv}", falling back to pip`);
      (0, import_build_utils4.debug)(`error: ${err}`);
    }
  }
  const cmdArgs = [
    "install",
    "--disable-pip-version-check",
    "--no-compile",
    "--no-cache-dir",
    "--target",
    target,
    ...args
  ];
  const pretty = `${pipPath} ${cmdArgs.join(" ")}`;
  (0, import_build_utils4.debug)(`Running "${pretty}"...`);
  try {
    await (0, import_execa3.default)(pipPath, cmdArgs, {
      cwd: workPath
    });
  } catch (err) {
    console.log(`Failed to run "${pretty}"`);
    (0, import_build_utils4.debug)(`error: ${err}`);
    throw err;
  }
}
async function installRequirement({
  pythonPath,
  pipPath,
  uvPath,
  dependency,
  version: version2,
  workPath,
  targetDir,
  meta,
  args = []
}) {
  const actualTargetDir = targetDir || workPath;
  if (meta.isDev && await isInstalled2(pythonPath, dependency, actualTargetDir)) {
    (0, import_build_utils4.debug)(
      `Skipping ${dependency} dependency installation, already installed in ${actualTargetDir}`
    );
    return;
  }
  const exact = `${dependency}==${version2}`;
  await pipInstall(pipPath, uvPath, workPath, [exact, ...args], targetDir);
}
async function installRequirementsFile({
  pythonPath,
  pipPath,
  uvPath,
  filePath,
  workPath,
  targetDir,
  meta,
  args = []
}) {
  const actualTargetDir = targetDir || workPath;
  if (meta.isDev && await areRequirementsInstalled(pythonPath, filePath, actualTargetDir)) {
    (0, import_build_utils4.debug)(`Skipping requirements file installation, already installed`);
    return;
  }
  await pipInstall(
    pipPath,
    uvPath,
    workPath,
    ["--upgrade", "-r", filePath, ...args],
    targetDir
  );
}

// src/dependency-externalizer.ts
var import_fs4 = __toESM(require("fs"));
var import_util = require("util");
var import_path6 = require("path");
var import_build_utils5 = require("@vercel/build-utils");
var import_python_analysis3 = require("@vercel/python-analysis");
var readFile = (0, import_util.promisify)(import_fs4.default.readFile);
var LAMBDA_SIZE_THRESHOLD_BYTES = 249 * 1024 * 1024;
var LAMBDA_PACKING_TARGET_BYTES = 245 * 1024 * 1024;
var LAMBDA_EPHEMERAL_STORAGE_BYTES = 500 * 1024 * 1024;
var PythonDependencyExternalizer = class {
  constructor(options) {
    // Populated by analyze()
    this.allVendorFiles = {};
    this.totalBundleSize = 0;
    this.analyzed = false;
    this.venvPath = options.venvPath;
    this.vendorDir = options.vendorDir;
    this.workPath = options.workPath;
    this.uvLockPath = options.uvLockPath;
    this.uvProjectDir = options.uvProjectDir;
    this.projectName = options.projectName;
    this.noBuildCheckFailed = options.noBuildCheckFailed;
    this.pythonPath = options.pythonPath;
    this.hasCustomCommand = options.hasCustomCommand;
    this.alwaysBundlePackages = options.alwaysBundlePackages ?? [];
  }
  shouldEnableRuntimeInstall() {
    if (this.hasCustomCommand) {
      return false;
    }
    const pythonOnHiveEnabled = process.env.VERCEL_PYTHON_ON_HIVE === "1" || process.env.VERCEL_PYTHON_ON_HIVE === "true";
    if (pythonOnHiveEnabled) {
      return false;
    } else if (this.totalBundleSize > LAMBDA_SIZE_THRESHOLD_BYTES && this.uvLockPath !== null) {
      return true;
    }
    return false;
  }
  /**
   * Analyze the bundle: mirror all vendor files, calculate total size,
   * and determine whether runtime installation is needed.
   * Must be called before generateBundle().
   */
  async analyze(files) {
    this.allVendorFiles = await mirrorPackagesIntoVendor({
      venvPath: this.venvPath,
      vendorDirName: this.vendorDir
    });
    const tempFilesForSizing = { ...files };
    for (const [p, f] of Object.entries(this.allVendorFiles)) {
      tempFilesForSizing[p] = f;
    }
    this.totalBundleSize = await calculateBundleSize(tempFilesForSizing);
    this.analyzed = true;
    const totalBundleSizeMB = (this.totalBundleSize / (1024 * 1024)).toFixed(2);
    (0, import_build_utils5.debug)(`Total bundle size: ${totalBundleSizeMB} MB`);
    const runtimeInstallEnabled = this.shouldEnableRuntimeInstall();
    const pythonOnHiveEnabled = process.env.VERCEL_PYTHON_ON_HIVE === "1" || process.env.VERCEL_PYTHON_ON_HIVE === "true";
    if (this.totalBundleSize > LAMBDA_SIZE_THRESHOLD_BYTES && this.hasCustomCommand && !pythonOnHiveEnabled) {
      const limitMB = (LAMBDA_SIZE_THRESHOLD_BYTES / (1024 * 1024)).toFixed(0);
      throw new import_build_utils5.NowBuildError({
        code: "LAMBDA_SIZE_EXCEEDED",
        message: `Total bundle size (${totalBundleSizeMB} MB) exceeds the Lambda size limit (${limitMB} MB).

Runtime dependency installation is not available for projects that use a custom build or install command, because custom commands may install dependencies that are not tracked in uv.lock.

To resolve this, either:
  1. Remove the custom build/install command and let Vercel manage dependencies automatically
  2. Reduce your dependency footprint to fit within the ${limitMB} MB limit`,
        link: "https://vercel.com/docs/functions/runtimes/python#controlling-what-gets-bundled",
        action: "Learn More"
      });
    }
    return {
      runtimeInstallEnabled,
      allVendorFiles: this.allVendorFiles,
      totalBundleSize: this.totalBundleSize
    };
  }
  /**
   * Generate the optimally-packed Lambda bundle.
   * Mutates `files` in place: adds vendor files (private + knapsack-selected
   * public), runtime config, and uv binary.
   * Must be called after analyze().
   */
  async generateBundle(files) {
    if (!this.analyzed) {
      throw new Error(
        "PythonDependencyExternalizer.analyze() must be called before generateBundle()"
      );
    }
    if (!this.uvLockPath || !this.uvProjectDir) {
      throw new import_build_utils5.NowBuildError({
        code: "RUNTIME_DEPENDENCY_INSTALLATION_FAILED",
        message: "Runtime dependency installation requires a uv.lock file and project directory."
      });
    }
    const totalBundleSizeMB = (this.totalBundleSize / (1024 * 1024)).toFixed(2);
    console.log(
      `Bundle size (${totalBundleSizeMB} MB) exceeds limit. Enabling runtime dependency installation.`
    );
    if (this.totalBundleSize > LAMBDA_EPHEMERAL_STORAGE_BYTES) {
      const ephemeralLimitMB = (LAMBDA_EPHEMERAL_STORAGE_BYTES / (1024 * 1024)).toFixed(0);
      throw new import_build_utils5.NowBuildError({
        code: "LAMBDA_SIZE_EXCEEDED",
        message: `Total dependency size (${totalBundleSizeMB} MB) exceeds Lambda ephemeral storage limit (${ephemeralLimitMB} MB). Even with runtime dependency installation, all packages must fit within the ${ephemeralLimitMB} MB ephemeral storage available to Lambda functions. Consider removing unused dependencies or splitting your application into smaller functions.`,
        link: "https://vercel.com/docs/functions/runtimes/python#controlling-what-gets-bundled",
        action: "Learn More"
      });
    }
    if (this.noBuildCheckFailed) {
      throw new import_build_utils5.NowBuildError({
        code: "RUNTIME_DEPENDENCY_INSTALLATION_FAILED",
        message: `Bundle size exceeds the Lambda limit and requires runtime dependency installation, but some packages in your uv.lock file do not have pre-built binary wheels available.
Runtime dependency installation requires all public packages to have binary wheels.

To fix this, either:
 1. Regenerate your lock file with: uv lock --upgrade --no-build, or
 2. Switch the problematic packages to ones that have pre-built wheels available`
      });
    }
    let lockContent;
    try {
      lockContent = await readFile(this.uvLockPath, "utf8");
    } catch (error) {
      if (error instanceof Error) {
        console.log(
          `Failed to read uv.lock file at "${this.uvLockPath}": ${error.message}`
        );
      } else {
        console.log(
          `Failed to read uv.lock file at "${this.uvLockPath}": ${String(error)}`
        );
      }
      throw new import_build_utils5.NowBuildError({
        code: "RUNTIME_DEPENDENCY_INSTALLATION_FAILED",
        message: `Failed to read uv.lock file at "${this.uvLockPath}"`
      });
    }
    let lockFile;
    try {
      lockFile = (0, import_python_analysis3.parseUvLock)(lockContent, this.uvLockPath);
    } catch (error) {
      if (error instanceof import_python_analysis3.PythonAnalysisError) {
        if (error.fileContent) {
          console.log(
            `Failed to parse "${error.path}". File content:
${error.fileContent}`
          );
        }
        throw new import_build_utils5.NowBuildError({
          code: error.code,
          message: error.message
        });
      }
      throw error;
    }
    const excludePackages = [];
    if (this.projectName) {
      excludePackages.push(this.projectName);
      (0, import_build_utils5.debug)(
        `Excluding project package "${this.projectName}" from runtime installation`
      );
    }
    const classification = (0, import_python_analysis3.classifyPackages)({
      lockFile,
      excludePackages
    });
    (0, import_build_utils5.debug)(
      `Package classification: ${classification.privatePackages.length} private, ${classification.publicPackages.length} public`
    );
    if (classification.publicPackages.length === 0) {
      throw new import_build_utils5.NowBuildError({
        code: "RUNTIME_DEPENDENCY_INSTALLATION_FAILED",
        message: "Bundle size exceeds limit but no public packages found for runtime installation."
      });
    }
    const packageSizes = await calculatePerPackageSizes(this.venvPath);
    const alwaysBundled = [
      ...classification.privatePackages,
      "vercel-runtime",
      "vercel_runtime",
      ...this.alwaysBundlePackages
    ];
    const alwaysBundledFiles = await mirrorPackagesIntoVendor({
      venvPath: this.venvPath,
      vendorDirName: this.vendorDir,
      includePackages: alwaysBundled
    });
    const baseFiles = { ...files };
    for (const [p, f] of Object.entries(alwaysBundledFiles)) {
      baseFiles[p] = f;
    }
    const fixedOverhead = await calculateBundleSize(baseFiles);
    let runtimeToolingOverhead = 0;
    if (process.env.VERCEL_BUILD_IMAGE) {
      try {
        const uvBinaryPath = await getUvBinaryForBundling(this.pythonPath);
        const uvStats = await import_fs4.default.promises.stat(uvBinaryPath);
        runtimeToolingOverhead = uvStats.size;
      } catch {
        runtimeToolingOverhead = 50 * 1024 * 1024;
      }
    }
    const remainingCapacity = LAMBDA_PACKING_TARGET_BYTES - fixedOverhead - runtimeToolingOverhead;
    (0, import_build_utils5.debug)(
      `Fixed overhead: ${(fixedOverhead / (1024 * 1024)).toFixed(2)} MB, runtime tooling: ${(runtimeToolingOverhead / (1024 * 1024)).toFixed(2)} MB, remaining capacity for public packages: ${(remainingCapacity / (1024 * 1024)).toFixed(2)} MB`
    );
    const publicPackageSizes = new Map(
      [...packageSizes].filter(
        ([name]) => classification.publicPackages.includes(name)
      )
    );
    const bundledPublic = lambdaKnapsack(publicPackageSizes, remainingCapacity);
    const allBundledPackages = [...alwaysBundled, ...bundledPublic];
    const selectedVendorFiles = await mirrorPackagesIntoVendor({
      venvPath: this.venvPath,
      vendorDirName: this.vendorDir,
      includePackages: allBundledPackages
    });
    for (const [p, f] of Object.entries(selectedVendorFiles)) {
      files[p] = f;
    }
    const bundledPackagesForConfig = [
      ...classification.privatePackages,
      ...bundledPublic,
      ...this.alwaysBundlePackages
    ];
    const projectDirRel = (0, import_path6.relative)(this.workPath, this.uvProjectDir);
    const uvLockRel = (0, import_path6.relative)(this.workPath, this.uvLockPath);
    const isOutsideWorkPath = projectDirRel.startsWith("..") || uvLockRel.startsWith("..");
    if (isOutsideWorkPath) {
      const srcPyproject = (0, import_path6.join)(this.uvProjectDir, "pyproject.toml");
      files[`${UV_BUNDLE_DIR}/pyproject.toml`] = new import_build_utils5.FileFsRef({
        fsPath: srcPyproject
      });
      files[`${UV_BUNDLE_DIR}/uv.lock`] = new import_build_utils5.FileFsRef({
        fsPath: this.uvLockPath
      });
    }
    const runtimeConfigData = JSON.stringify({
      projectDir: isOutsideWorkPath ? UV_BUNDLE_DIR : projectDirRel,
      bundledPackages: bundledPackagesForConfig
    });
    files[`${UV_BUNDLE_DIR}/_runtime_config.json`] = new import_build_utils5.FileBlob({
      data: runtimeConfigData
    });
    if (process.env.VERCEL_BUILD_IMAGE) {
      try {
        const uvBinaryPath = await getUvBinaryForBundling(this.pythonPath);
        const uvBundleDir = (0, import_path6.join)(this.workPath, UV_BUNDLE_DIR);
        const uvLocalPath = (0, import_path6.join)(uvBundleDir, "uv");
        await import_fs4.default.promises.mkdir(uvBundleDir, { recursive: true });
        await import_fs4.default.promises.copyFile(uvBinaryPath, uvLocalPath);
        await import_fs4.default.promises.chmod(uvLocalPath, 493);
        const uvBundlePath = `${UV_BUNDLE_DIR}/uv`;
        files[uvBundlePath] = new import_build_utils5.FileFsRef({
          fsPath: uvLocalPath,
          mode: 33261
          // Regular file + executable
        });
        (0, import_build_utils5.debug)(`Bundled uv binary from ${uvBinaryPath} to ${uvLocalPath}`);
      } catch (err) {
        throw new import_build_utils5.NowBuildError({
          code: "RUNTIME_DEPENDENCY_INSTALLATION_FAILED",
          message: `Failed to bundle uv binary for runtime installation: ${err instanceof Error ? err.message : String(err)}`
        });
      }
    }
    const finalBundleSize = await calculateBundleSize(files);
    if (finalBundleSize > LAMBDA_SIZE_THRESHOLD_BYTES) {
      const finalSizeMB = (finalBundleSize / (1024 * 1024)).toFixed(2);
      const limitMB = (LAMBDA_SIZE_THRESHOLD_BYTES / (1024 * 1024)).toFixed(0);
      throw new import_build_utils5.NowBuildError({
        code: "LAMBDA_SIZE_EXCEEDED",
        message: `Bundle size (${finalSizeMB} MB) exceeds Lambda limit (${limitMB} MB) even after deferring public packages to runtime installation. This usually means your private packages or source code are too large. Consider reducing the size of private dependencies or splitting your application.`,
        link: "https://vercel.com/docs/functions/runtimes/python#controlling-what-gets-bundled",
        action: "Learn More"
      });
    }
  }
};
async function mirrorPackagesIntoVendor({
  venvPath,
  vendorDirName,
  includePackages
}) {
  const vendorFiles = {};
  if (includePackages && includePackages.length === 0) {
    return vendorFiles;
  }
  const includeSet = includePackages ? new Set(includePackages.map(import_python_analysis3.normalizePackageName)) : null;
  const sitePackageDirs = await getVenvSitePackagesDirs(venvPath);
  for (const dir of sitePackageDirs) {
    if (!import_fs4.default.existsSync(dir))
      continue;
    const resolvedDir = (0, import_path6.resolve)(dir);
    const dirPrefix = resolvedDir + import_path6.sep;
    const distributions = await (0, import_python_analysis3.scanDistributions)(dir);
    for (const [name, dist] of distributions) {
      if (includeSet && !includeSet.has(name))
        continue;
      for (const { path: rawPath } of dist.files) {
        const filePath = rawPath.replaceAll("/", import_path6.sep);
        if (!(0, import_path6.resolve)(resolvedDir, filePath).startsWith(dirPrefix)) {
          continue;
        }
        if (filePath.endsWith(".pyc") || filePath.split(import_path6.sep).includes("__pycache__")) {
          continue;
        }
        const srcFsPath = (0, import_path6.join)(dir, filePath);
        if (!import_fs4.default.existsSync(srcFsPath)) {
          continue;
        }
        const bundlePath = (0, import_path6.join)(vendorDirName, filePath).replace(/\\/g, "/");
        vendorFiles[bundlePath] = new import_build_utils5.FileFsRef({ fsPath: srcFsPath });
      }
    }
  }
  (0, import_build_utils5.debug)(
    `Mirrored ${Object.keys(vendorFiles).length} files` + (includePackages ? ` from ${includePackages.length} packages` : "")
  );
  return vendorFiles;
}
async function calculateBundleSize(files) {
  let totalSize = 0;
  for (const filePath of Object.keys(files)) {
    const file = files[filePath];
    if ("fsPath" in file && file.fsPath) {
      try {
        const stats = await import_fs4.default.promises.stat(file.fsPath);
        totalSize += stats.size;
      } catch (err) {
        console.warn(
          `Warning: Failed to stat file ${file.fsPath}, size will not be included in bundle calculation: ${err}`
        );
      }
    } else if ("data" in file) {
      const data = file.data;
      totalSize += typeof data === "string" ? Buffer.byteLength(data) : data.length;
    }
  }
  return totalSize;
}
function lambdaKnapsack(packages, capacity) {
  if (capacity <= 0) {
    return [];
  }
  const sorted = [...packages.entries()].sort(([, a], [, b]) => b - a);
  const bundled = [];
  let remaining = capacity;
  for (const [name, size] of sorted) {
    if (size <= remaining) {
      bundled.push(name);
      remaining -= size;
    }
  }
  return bundled;
}
async function calculatePerPackageSizes(venvPath) {
  const sizes = /* @__PURE__ */ new Map();
  const sitePackageDirs = await getVenvSitePackagesDirs(venvPath);
  for (const dir of sitePackageDirs) {
    if (!import_fs4.default.existsSync(dir))
      continue;
    const resolvedDir = (0, import_path6.resolve)(dir);
    const dirPrefix = resolvedDir + import_path6.sep;
    const distributions = await (0, import_python_analysis3.scanDistributions)(dir);
    for (const [name, dist] of distributions) {
      let totalSize = 0;
      for (const { path: rawPath } of dist.files) {
        const filePath = rawPath.replaceAll("/", import_path6.sep);
        if (!(0, import_path6.resolve)(resolvedDir, filePath).startsWith(dirPrefix)) {
          continue;
        }
        if (filePath.endsWith(".pyc") || filePath.split(import_path6.sep).includes("__pycache__")) {
          continue;
        }
        try {
          const stats = await import_fs4.default.promises.stat((0, import_path6.join)(dir, filePath));
          totalSize += stats.size;
        } catch {
        }
      }
      sizes.set(name, totalSize);
    }
  }
  return sizes;
}

// src/start-dev-server.ts
var import_child_process2 = require("child_process");
var import_fs5 = require("fs");
var import_path8 = require("path");
var import_build_utils8 = require("@vercel/build-utils");
var import_get_port = __toESM(require_get_port());
var import_is_port_reachable = __toESM(require_is_port_reachable());

// src/entrypoint.ts
var import_path7 = require("path");
var import_build_utils6 = require("@vercel/build-utils");
var import_build_utils7 = require("@vercel/build-utils");
var PYTHON_ENTRYPOINT_FILENAMES = [
  "app",
  "index",
  "server",
  "main",
  "wsgi",
  "asgi"
];
var PYTHON_ENTRYPOINT_DIRS = ["", "src", "app", "api"];
var PYTHON_CANDIDATE_ENTRYPOINTS = getCandidateEntrypointsInDirs(
  PYTHON_ENTRYPOINT_DIRS
);
function getCandidateEntrypointsInDirs(dirs) {
  return dirs.flatMap(
    (dir) => PYTHON_ENTRYPOINT_FILENAMES.map(
      (filename) => import_path7.posix.join(dir, `${filename}.py`)
    )
  );
}
async function getPyprojectEntrypoint(workPath) {
  const pyprojectData = await (0, import_build_utils7.readConfigFile)((0, import_path7.join)(workPath, "pyproject.toml"));
  if (!pyprojectData)
    return null;
  const scripts = pyprojectData.project?.scripts;
  const appScript = scripts?.app;
  if (typeof appScript !== "string")
    return null;
  const match = appScript.match(/([A-Za-z_][\w.]*)\s*:\s*([A-Za-z_][\w]*)/);
  if (!match)
    return null;
  const modulePath = match[1];
  const relPath = modulePath.replace(/\./g, "/");
  try {
    const fsFiles = await (0, import_build_utils6.glob)("**", workPath);
    const candidates = [`${relPath}.py`, `${relPath}/__init__.py`];
    for (const candidate of candidates) {
      if (fsFiles[candidate])
        return candidate;
    }
    return null;
  } catch {
    (0, import_build_utils6.debug)("Failed to discover Python entrypoint from pyproject.toml");
    return null;
  }
}
async function findValidEntrypoint(fsFiles, candidates) {
  for (const candidate of candidates) {
    if (fsFiles[candidate]) {
      const isValid = await (0, import_build_utils6.isPythonEntrypoint)(fsFiles[candidate]);
      if (isValid) {
        (0, import_build_utils6.debug)(`Detected Python entrypoint: ${candidate}`);
        return candidate;
      }
    }
  }
  return null;
}
async function detectGenericPythonEntrypoint(workPath, configuredEntrypoint) {
  const entry = configuredEntrypoint.endsWith(".py") ? configuredEntrypoint : `${configuredEntrypoint}.py`;
  try {
    const fsFiles = await (0, import_build_utils6.glob)("**", workPath);
    if (fsFiles[entry]) {
      const isValid = await (0, import_build_utils6.isPythonEntrypoint)(fsFiles[entry]);
      if (isValid) {
        (0, import_build_utils6.debug)(`Using configured Python entrypoint: ${entry}`);
        return entry;
      }
    }
    const candidates = PYTHON_CANDIDATE_ENTRYPOINTS.filter(
      (c) => !!fsFiles[c]
    );
    return findValidEntrypoint(fsFiles, candidates);
  } catch {
    (0, import_build_utils6.debug)("Failed to discover Python entrypoint");
    return null;
  }
}
async function detectDjangoPythonEntrypoint(workPath, configuredEntrypoint) {
  const entry = configuredEntrypoint.endsWith(".py") ? configuredEntrypoint : `${configuredEntrypoint}.py`;
  try {
    const fsFiles = await (0, import_build_utils6.glob)("**", workPath);
    if (fsFiles[entry]) {
      const isValid = await (0, import_build_utils6.isPythonEntrypoint)(fsFiles[entry]);
      if (isValid) {
        (0, import_build_utils6.debug)(`Using configured Python entrypoint: ${entry}`);
        return entry;
      }
    }
    const rootGlobs = await (0, import_build_utils6.glob)("*", {
      cwd: workPath,
      includeDirectories: true
    });
    const rootDirs = [
      "",
      ...Object.keys(rootGlobs).filter(
        (name) => !name.startsWith(".") && rootGlobs[name].mode != null && (0, import_build_utils6.isDirectory)(rootGlobs[name].mode)
      )
    ];
    for (const rootDir of rootDirs) {
      const currPath = (0, import_path7.join)(workPath, rootDir);
      const wsgiEntry = await (0, import_build_utils6.getDjangoEntrypoint)(currPath);
      if (wsgiEntry) {
        const fullWsgiEntry = import_path7.posix.join(rootDir, wsgiEntry);
        if (fsFiles[fullWsgiEntry]) {
          (0, import_build_utils6.debug)(`Using Django WSGI entrypoint: ${fullWsgiEntry}`);
          return fullWsgiEntry;
        }
      }
    }
    const baseCandidates = getCandidateEntrypointsInDirs(rootDirs);
    const candidates = baseCandidates.filter((c) => !!fsFiles[c]);
    return findValidEntrypoint(fsFiles, candidates);
  } catch {
    (0, import_build_utils6.debug)("Failed to discover Django Python entrypoint");
    return null;
  }
}
async function detectPythonEntrypoint(framework, workPath, configuredEntrypoint) {
  const entrypoint = framework === "django" ? await detectDjangoPythonEntrypoint(workPath, configuredEntrypoint) : await detectGenericPythonEntrypoint(workPath, configuredEntrypoint);
  if (entrypoint)
    return entrypoint;
  return await getPyprojectEntrypoint(workPath);
}

// src/start-dev-server.ts
var import_python_analysis4 = require("@vercel/python-analysis");
var DEV_SERVER_STARTUP_TIMEOUT = 1e4;
function silenceNodeWarnings() {
  const original = process.emitWarning.bind(
    process
  );
  let active = true;
  const wrapped = (warning, ...args) => {
    if (!active) {
      return original(
        warning,
        ...args
      );
    }
    return;
  };
  process.emitWarning = wrapped;
  return () => {
    if (!active)
      return;
    active = false;
    if (process.emitWarning === wrapped) {
      process.emitWarning = original;
    }
  };
}
var DEV_SHIM_MODULE = "vc_init_dev";
function createLogListener(callback, stream) {
  return (buf) => {
    if (callback) {
      callback(buf);
    } else {
      const s = buf.toString();
      for (const line of s.split(/\r?\n/)) {
        if (line) {
          stream.write(line.endsWith("\n") ? line : line + "\n");
        }
      }
    }
  };
}
function sleep(ms) {
  return new Promise((resolve2) => setTimeout(resolve2, ms));
}
async function checkForPort(port, timeout) {
  const start = Date.now();
  let host = false;
  while (!(host = await getReachableHost(port))) {
    if (Date.now() - start > timeout) {
      break;
    }
    await sleep(100);
  }
  if (!host) {
    throw new Error(`Detecting port ${port} timed out after ${timeout}ms`);
  }
  return host;
}
async function getReachableHost(port) {
  const results = await Promise.all([
    (0, import_is_port_reachable.default)(port, { host: "::1" }).then((r) => r && `[::1]`),
    (0, import_is_port_reachable.default)(port, { host: "127.0.0.1" }).then((r) => r && "127.0.0.1")
  ]);
  return results.find(Boolean) || false;
}
async function syncDependencies({
  workPath,
  uvPath,
  pythonBin,
  env,
  onStdout,
  onStderr
}) {
  const pythonPackage = await discoverPackage({
    entrypointDir: workPath,
    rootDir: workPath
  });
  const installInfo = detectInstallSource(pythonPackage, workPath);
  const { manifestType } = installInfo;
  let { manifestPath } = installInfo;
  const manifest = pythonPackage.manifest;
  if (!manifestType || !manifestPath) {
    (0, import_build_utils8.debug)("No Python project manifest found, skipping dependency sync");
    return;
  }
  if (manifest?.origin && manifestType === "pyproject.toml") {
    const syncDir = (0, import_path8.join)(workPath, ".vercel", "python", "sync");
    (0, import_fs5.mkdirSync)(syncDir, { recursive: true });
    const tempPyproject = (0, import_path8.join)(syncDir, "pyproject.toml");
    const content = (0, import_python_analysis4.stringifyManifest)(manifest.data);
    (0, import_fs5.writeFileSync)(tempPyproject, content, "utf8");
    manifestPath = tempPyproject;
    (0, import_build_utils8.debug)(
      `Wrote converted ${manifest.origin.kind} manifest to ${tempPyproject}`
    );
  }
  const writeOut = (msg) => {
    if (onStdout) {
      onStdout(Buffer.from(msg));
    } else {
      process.stdout.write(msg);
    }
  };
  const writeErr = (msg) => {
    if (onStderr) {
      onStderr(Buffer.from(msg));
    } else {
      process.stderr.write(msg);
    }
  };
  const captured = [];
  try {
    await runSync({
      manifestType,
      manifestPath,
      uvPath,
      pythonBin,
      env,
      onStdout: (data) => captured.push(["stdout", data]),
      onStderr: (data) => captured.push(["stderr", data])
    });
  } catch (err) {
    for (const [channel, chunk] of captured) {
      (channel === "stdout" ? writeOut : writeErr)(chunk.toString());
    }
    throw new import_build_utils8.NowBuildError({
      code: "PYTHON_DEPENDENCY_SYNC_FAILED",
      message: `Failed to install Python dependencies from ${manifestType}: ${err instanceof Error ? err.message : String(err)}`
    });
  }
}
async function runSync({
  manifestType,
  manifestPath,
  uvPath,
  pythonBin,
  env,
  onStdout,
  onStderr
}) {
  const projectDir = (0, import_path8.dirname)(manifestPath);
  const pip = uvPath ? { cmd: uvPath, prefix: ["pip", "install"] } : { cmd: pythonBin, prefix: ["-m", "pip", "install"] };
  let spawnCmd;
  let spawnArgs;
  switch (manifestType) {
    case "uv.lock": {
      if (!uvPath) {
        throw new import_build_utils8.NowBuildError({
          code: "PYTHON_DEPENDENCY_SYNC_FAILED",
          message: "uv is required to install dependencies from uv.lock.",
          link: "https://docs.astral.sh/uv/getting-started/installation/",
          action: "Install uv"
        });
      }
      spawnCmd = uvPath;
      spawnArgs = ["sync"];
      break;
    }
    case "pylock.toml": {
      spawnCmd = pip.cmd;
      spawnArgs = [...pip.prefix, "-r", manifestPath];
      break;
    }
    case "pyproject.toml": {
      spawnCmd = pip.cmd;
      spawnArgs = [...pip.prefix, projectDir];
      break;
    }
    default:
      (0, import_build_utils8.debug)(`Unknown manifest type: ${manifestType}`);
      return;
  }
  await new Promise((resolve2, reject) => {
    (0, import_build_utils8.debug)(`Running "${spawnCmd} ${spawnArgs.join(" ")}" in ${projectDir}...`);
    const child = (0, import_child_process2.spawn)(spawnCmd, spawnArgs, {
      cwd: projectDir,
      env: getProtectedUvEnv(env),
      stdio: ["inherit", "pipe", "pipe"]
    });
    child.stdout?.on("data", (data) => {
      if (onStdout) {
        onStdout(data);
      } else {
        process.stdout.write(data.toString());
      }
    });
    child.stderr?.on("data", (data) => {
      if (onStderr) {
        onStderr(data);
      } else {
        process.stderr.write(data.toString());
      }
    });
    child.on("error", reject);
    child.on("exit", (code, signal) => {
      if (code === 0) {
        resolve2();
      } else {
        reject(
          new Error(
            `Command "${spawnCmd} ${spawnArgs.join(" ")}" failed with code ${code}, signal ${signal}`
          )
        );
      }
    });
  });
}
var PERSISTENT_SERVERS = /* @__PURE__ */ new Map();
var PENDING_STARTS = /* @__PURE__ */ new Map();
var restoreWarnings = null;
var cleanupHandlersInstalled = false;
function installGlobalCleanupHandlers() {
  if (cleanupHandlersInstalled)
    return;
  cleanupHandlersInstalled = true;
  const killAll = () => {
    for (const [key, info] of PERSISTENT_SERVERS.entries()) {
      try {
        process.kill(info.pid, "SIGTERM");
      } catch (err) {
        (0, import_build_utils8.debug)(`Error sending SIGTERM to ${info.pid}: ${err}`);
      }
      try {
        process.kill(info.pid, "SIGKILL");
      } catch (err) {
        (0, import_build_utils8.debug)(`Error sending SIGKILL to ${info.pid}: ${err}`);
      }
      PERSISTENT_SERVERS.delete(key);
    }
    if (restoreWarnings) {
      try {
        restoreWarnings();
      } catch (err) {
        (0, import_build_utils8.debug)(`Error restoring warnings: ${err}`);
      }
      restoreWarnings = null;
    }
  };
  process.on("SIGINT", () => {
    killAll();
  });
  process.on("SIGTERM", () => {
    killAll();
  });
  process.on("exit", () => {
    killAll();
  });
}
function createDevShim(workPath, entry, modulePath) {
  try {
    const vercelPythonDir = (0, import_path8.join)(workPath, ".vercel", "python");
    (0, import_fs5.mkdirSync)(vercelPythonDir, { recursive: true });
    let qualifiedModule = modulePath;
    let extraPythonPath;
    if ((0, import_fs5.existsSync)((0, import_path8.join)(workPath, "__init__.py"))) {
      const pkgName = (0, import_path8.basename)(workPath);
      qualifiedModule = `${pkgName}.${modulePath}`;
      extraPythonPath = (0, import_path8.dirname)(workPath);
    }
    const entryAbs = (0, import_path8.join)(workPath, entry);
    const shimPath = (0, import_path8.join)(vercelPythonDir, `${DEV_SHIM_MODULE}.py`);
    const templatePath = (0, import_path8.join)(__dirname, "..", `${DEV_SHIM_MODULE}.py`);
    const template = (0, import_fs5.readFileSync)(templatePath, "utf8");
    const shimSource = template.replace(/__VC_DEV_MODULE_NAME__/g, qualifiedModule).replace(/__VC_DEV_ENTRY_ABS__/g, entryAbs);
    (0, import_fs5.writeFileSync)(shimPath, shimSource, "utf8");
    (0, import_build_utils8.debug)(`Prepared Python dev shim at ${shimPath}`);
    return { module: DEV_SHIM_MODULE, extraPythonPath };
  } catch (err) {
    (0, import_build_utils8.debug)(`Failed to prepare dev shim: ${err?.message || err}`);
    return null;
  }
}
async function getMultiServicePythonRunner(workPath, env, systemPython, uvPath) {
  const { pythonCmd, venvRoot } = useVirtualEnv(workPath, env, systemPython);
  if (venvRoot) {
    (0, import_build_utils8.debug)(`Using existing virtualenv at ${venvRoot} for multi-service dev`);
    return { command: pythonCmd, args: [] };
  }
  const venvPath = (0, import_path8.join)(workPath, ".venv");
  await ensureVenv({ pythonPath: systemPython, venvPath, uvPath, quiet: true });
  (0, import_build_utils8.debug)(`Created virtualenv at ${venvPath} for multi-service dev`);
  const pythonBin = getVenvPythonBin(venvPath);
  const binDir = getVenvBinDir(venvPath);
  env.VIRTUAL_ENV = venvPath;
  env.PATH = `${binDir}${import_path8.delimiter}${env.PATH || ""}`;
  return { command: pythonBin, args: [] };
}
var startDevServer = async (opts) => {
  const {
    entrypoint: rawEntrypoint,
    workPath,
    meta = {},
    config,
    onStdout,
    onStderr
  } = opts;
  const framework = config?.framework;
  if (!framework) {
    return null;
  }
  if (!restoreWarnings)
    restoreWarnings = silenceNodeWarnings();
  installGlobalCleanupHandlers();
  const entry = await detectPythonEntrypoint(
    framework,
    workPath,
    rawEntrypoint
  );
  if (!entry) {
    const searched = PYTHON_CANDIDATE_ENTRYPOINTS.join(", ");
    throw new import_build_utils8.NowBuildError({
      code: "PYTHON_ENTRYPOINT_NOT_FOUND",
      message: `No ${framework} entrypoint found. Add an 'app' script in pyproject.toml or define an entrypoint in one of: ${searched}.`,
      link: `https://vercel.com/docs/frameworks/backend/${framework?.toLowerCase()}#exporting-the-${framework?.toLowerCase()}-application`,
      action: "Learn More"
    });
  }
  const modulePath = entry.replace(/\.py$/i, "").replace(/[\\/]/g, ".");
  const env = { ...process.env, ...meta.env || {} };
  const serverKey = `${workPath}::${entry}::${framework}`;
  const existing = PERSISTENT_SERVERS.get(serverKey);
  if (existing) {
    return {
      port: existing.port,
      pid: existing.pid,
      shutdown: async () => {
      }
    };
  }
  {
    const pending = PENDING_STARTS.get(serverKey);
    if (pending) {
      const { port, pid } = await pending;
      return {
        port,
        pid,
        shutdown: async () => {
        }
      };
    }
  }
  let childProcess = null;
  let stdoutLogListener = null;
  let stderrLogListener = null;
  let resolveChildReady;
  let rejectChildReady;
  const childReady = new Promise(
    (resolve2, reject) => {
      resolveChildReady = resolve2;
      rejectChildReady = reject;
    }
  );
  PENDING_STARTS.set(serverKey, childReady);
  try {
    const { pythonPath: systemPython } = getDefaultPythonVersion(meta);
    const uvPath = await findUvBinary(systemPython);
    const venv = isInVirtualEnv();
    const serviceCount = meta.serviceCount ?? 0;
    const pythonServiceCount = meta.pythonServiceCount ?? 1;
    if (venv && pythonServiceCount > 1) {
      const yellow = "\x1B[33m";
      const white = "\x1B[1m";
      const reset = "\x1B[0m";
      throw new import_build_utils8.NowBuildError({
        code: "PYTHON_EXTERNAL_VENV_DETECTED",
        message: `Detected activated venv at ${yellow}${venv}${reset}, ${white}vercel dev${reset} manages virtual environments automatically.
Run ${white}deactivate${reset} and try again.`
      });
    }
    let spawnCommand = systemPython;
    let spawnArgsPrefix = [];
    if (serviceCount > 0) {
      const runner = await getMultiServicePythonRunner(
        workPath,
        env,
        systemPython,
        uvPath
      );
      spawnCommand = runner.command;
      spawnArgsPrefix = runner.args;
      (0, import_build_utils8.debug)(
        `Multi-service Python runner: ${spawnCommand} ${spawnArgsPrefix.join(" ")}`
      );
    } else if (venv) {
      (0, import_build_utils8.debug)(`Running in virtualenv at ${venv}`);
    } else {
      const { pythonCmd: venvPythonCmd, venvRoot } = useVirtualEnv(
        workPath,
        env,
        systemPython
      );
      spawnCommand = venvPythonCmd;
      if (venvRoot) {
        (0, import_build_utils8.debug)(`Using virtualenv at ${venvRoot}`);
      } else {
        (0, import_build_utils8.debug)("No virtualenv found");
        try {
          const yellow = "\x1B[33m";
          const reset = "\x1B[0m";
          const venvCmd = process.platform === "win32" ? "python -m venv .venv && .venv\\Scripts\\activate" : "python -m venv .venv && source .venv/bin/activate";
          process.stderr.write(
            `${yellow}Warning: no virtual environment detected in ${workPath}. Using system Python: ${systemPython}.${reset}
If you are using a virtual environment, activate it before running "vercel dev", or create one: ${venvCmd}
`
          );
        } catch (_) {
        }
      }
    }
    if (meta.syncDependencies) {
      const gray = "\x1B[90m";
      const reset = "\x1B[0m";
      const syncMessage = `${gray}Synchronizing dependencies...${reset}
`;
      if (onStdout) {
        onStdout(Buffer.from(syncMessage));
      } else {
        console.log(syncMessage);
      }
      await syncDependencies({
        workPath,
        uvPath,
        pythonBin: spawnCommand,
        env,
        onStdout,
        onStderr
      });
    }
    const port = typeof meta.port === "number" ? meta.port : await (0, import_get_port.default)();
    env.PORT = `${port}`;
    const devShim = createDevShim(workPath, entry, modulePath);
    if (devShim) {
      const vercelPythonDir = (0, import_path8.join)(workPath, ".vercel", "python");
      const pathParts = [vercelPythonDir];
      if (devShim.extraPythonPath) {
        pathParts.push(devShim.extraPythonPath);
      }
      const existingPythonPath = env.PYTHONPATH || "";
      if (existingPythonPath) {
        pathParts.push(existingPythonPath);
      }
      env.PYTHONPATH = pathParts.join(import_path8.delimiter);
    }
    const moduleToRun = devShim?.module || modulePath;
    const pythonArgs = ["-u", "-m", moduleToRun];
    const argv = [...spawnArgsPrefix, ...pythonArgs];
    (0, import_build_utils8.debug)(
      `Starting Python dev server (${framework}): ${spawnCommand} ${argv.join(" ")} [PORT=${port}]`
    );
    if (process.stdout.columns) {
      env.COLUMNS = `${process.stdout.columns}`;
    }
    const child = (0, import_child_process2.spawn)(spawnCommand, argv, {
      cwd: workPath,
      env,
      stdio: ["ignore", "pipe", "pipe"],
      detached: true
    });
    childProcess = child;
    stdoutLogListener = createLogListener(onStdout, process.stdout);
    stderrLogListener = createLogListener(onStderr, process.stderr);
    child.stdout?.on("data", stdoutLogListener);
    child.stderr?.on("data", stderrLogListener);
    const childExited = new Promise((_resolve, reject) => {
      child.once("error", (err) => {
        reject(err);
      });
      child.once("exit", (code, signal) => {
        reject(
          new Error(
            `Python dev server exited before binding (code=${code}, signal=${signal})`
          )
        );
      });
    });
    await Promise.race([
      checkForPort(port, DEV_SERVER_STARTUP_TIMEOUT),
      childExited
    ]);
    const pid = child.pid;
    resolveChildReady({ port, pid });
    PERSISTENT_SERVERS.set(serverKey, {
      port,
      pid,
      child: childProcess,
      stdoutLogListener,
      stderrLogListener
    });
    const shutdown = async () => {
    };
    return { port, pid, shutdown };
  } catch (err) {
    rejectChildReady(err);
    throw err;
  } finally {
    PENDING_STARTS.delete(serverKey);
  }
};

// src/quirks/index.ts
var import_build_utils11 = require("@vercel/build-utils");
var import_python_analysis6 = require("@vercel/python-analysis");

// src/quirks/matplotlib.ts
var matplotlibQuirk = {
  dependency: "matplotlib",
  async run() {
    return {
      env: { MPLCONFIGDIR: "/tmp" }
    };
  }
};

// src/quirks/litellm.ts
var import_fs6 = __toESM(require("fs"));
var import_path9 = require("path");
var import_build_utils9 = require("@vercel/build-utils");
var LAMBDA_ROOT = "/var/task";
var CONFIG_CANDIDATES = [
  "litellm_config.yaml",
  "litellm_config.yml",
  "litellm.yaml",
  "litellm.yml"
];
async function findConfigFile(workPath) {
  for (const name of CONFIG_CANDIDATES) {
    const candidate = (0, import_path9.join)(workPath, name);
    try {
      await import_fs6.default.promises.access(candidate);
      return name;
    } catch {
    }
  }
  return null;
}
var litellmQuirk = {
  dependency: "litellm",
  runsBefore: ["prisma"],
  async run(ctx) {
    const buildEnv = {};
    const env = {};
    const sitePackagesDirs = await getVenvSitePackagesDirs(ctx.venvPath);
    for (const sitePackages of sitePackagesDirs) {
      const schemaPath = (0, import_path9.join)(
        sitePackages,
        "litellm",
        "proxy",
        "schema.prisma"
      );
      try {
        await import_fs6.default.promises.access(schemaPath);
        (0, import_build_utils9.debug)(`LiteLLM quirk: found schema at ${schemaPath}`);
        buildEnv.PRISMA_SCHEMA_PATH = schemaPath;
        break;
      } catch {
      }
    }
    if (!buildEnv.PRISMA_SCHEMA_PATH) {
      (0, import_build_utils9.debug)("LiteLLM quirk: schema.prisma not found in any site-packages");
    }
    if (!process.env.CONFIG_FILE_PATH) {
      const configName = await findConfigFile(ctx.workPath);
      if (configName) {
        (0, import_build_utils9.debug)(`LiteLLM quirk: found config at ${configName}`);
        buildEnv.CONFIG_FILE_PATH = (0, import_path9.join)(ctx.workPath, configName);
        env.CONFIG_FILE_PATH = (0, import_path9.join)(LAMBDA_ROOT, configName);
      }
    } else {
      (0, import_build_utils9.debug)(
        `LiteLLM quirk: CONFIG_FILE_PATH already set to ${process.env.CONFIG_FILE_PATH}`
      );
    }
    return { buildEnv, env };
  }
};

// src/quirks/prisma.ts
var import_fs7 = __toESM(require("fs"));
var import_path10 = require("path");
var import_execa4 = __toESM(require_execa());
var import_build_utils10 = require("@vercel/build-utils");
var import_python_analysis5 = require("@vercel/python-analysis");
function execErrorMessage(err) {
  if (err != null && typeof err === "object" && "stderr" in err) {
    const stderr = String(err.stderr);
    if (stderr)
      return stderr;
  }
  return err instanceof Error ? err.message : String(err);
}
var DUMMY_SCHEMA_NAME = "_prisma_dummy.prisma";
var LAMBDA_ROOT2 = "/var/task";
var RUNTIME_OPENSSL_VERSION = "3.2";
function getLambdaBinaryTarget() {
  return process.arch === "arm64" ? "linux-arm64-openssl-3.0.x" : "rhel-openssl-3.0.x";
}
function buildDummySchema(generatedDir) {
  const lambdaTarget = getLambdaBinaryTarget();
  return `datasource db {
  provider = "sqlite"
  url      = "file:dev.db"
}

generator client {
  provider      = "prisma-client-py"
  binaryTargets = ["native", "${lambdaTarget}"]
  output        = "${generatedDir}"
}

model DummyModel {
  id Int @id
}
`;
}
async function findUserSchema(workPath) {
  const envPath = process.env.PRISMA_SCHEMA_PATH;
  if (envPath) {
    const resolved = (0, import_path10.isAbsolute)(envPath) ? envPath : (0, import_path10.join)(workPath, envPath);
    try {
      await import_fs7.default.promises.access(resolved);
      return resolved;
    } catch {
      (0, import_build_utils10.debug)(`PRISMA_SCHEMA_PATH=${envPath} not found at ${resolved}`);
      return null;
    }
  }
  const candidates = [
    (0, import_path10.join)(workPath, "schema.prisma"),
    (0, import_path10.join)(workPath, "prisma", "schema.prisma")
  ];
  for (const candidate of candidates) {
    try {
      await import_fs7.default.promises.access(candidate);
      return candidate;
    } catch {
    }
  }
  return null;
}
async function collectFiles(dir, base) {
  const result = [];
  let entries;
  try {
    entries = await import_fs7.default.promises.readdir(dir, { withFileTypes: true });
  } catch {
    return result;
  }
  for (const entry of entries) {
    if (entry.name === "__pycache__")
      continue;
    const full = (0, import_path10.join)(dir, entry.name);
    if (entry.isDirectory()) {
      result.push(...await collectFiles(full, base));
    } else {
      result.push((0, import_path10.relative)(base, full));
    }
  }
  return result;
}
async function cleanCacheArtifacts(cacheDir, extras = []) {
  const paths = [
    (0, import_path10.join)(cacheDir, "node_modules"),
    (0, import_path10.join)(cacheDir, "package.json"),
    (0, import_path10.join)(cacheDir, "package-lock.json"),
    ...extras
  ];
  for (const p of paths) {
    try {
      await import_fs7.default.promises.rm(p, { recursive: true, force: true });
    } catch (err) {
      console.warn(
        `could not clean up ${p}: ${err instanceof Error ? err.message : String(err)}`
      );
    }
  }
}
async function isClientGenerated(pythonPath, env) {
  try {
    await (0, import_execa4.default)(pythonPath, ["-c", "import prisma.client"], {
      env,
      stdio: "pipe"
    });
    return true;
  } catch {
    return false;
  }
}
var prismaQuirk = {
  dependency: "prisma",
  async run(ctx) {
    const { venvPath, pythonEnv, workPath } = ctx;
    const pythonPath = getVenvPythonBin(venvPath);
    const runtimeCacheDir = (0, import_path10.join)(
      LAMBDA_ROOT2,
      resolveVendorDir(),
      "prisma",
      "__bincache__"
    );
    const sitePackagesDirs = await getVenvSitePackagesDirs(venvPath);
    let sitePackages;
    for (const dir of sitePackagesDirs) {
      try {
        await import_fs7.default.promises.access((0, import_path10.join)(dir, "prisma"));
        sitePackages = dir;
        break;
      } catch {
      }
    }
    if (!sitePackages) {
      console.warn(
        "prisma: could not find prisma in any site-packages directory"
      );
      return {};
    }
    const cacheDir = (0, import_path10.join)(sitePackages, "prisma", "__bincache__");
    await import_fs7.default.promises.mkdir(cacheDir, { recursive: true });
    const generateEnv = {
      ...pythonEnv,
      PRISMA_BINARY_CACHE_DIR: cacheDir
    };
    const generatedDir = (0, import_path10.join)(workPath, "_prisma_generated");
    const dummySchemaPath = (0, import_path10.join)(workPath, DUMMY_SCHEMA_NAME);
    await import_fs7.default.promises.writeFile(
      dummySchemaPath,
      buildDummySchema(generatedDir)
    );
    (0, import_build_utils10.debug)(`Running prisma generate (dummy) with cache dir: ${cacheDir}`);
    try {
      const dummyResult = await (0, import_execa4.default)(
        pythonPath,
        ["-m", "prisma", "generate", `--schema=${dummySchemaPath}`],
        {
          cwd: workPath,
          env: generateEnv,
          stdio: "pipe"
        }
      );
      if (dummyResult.stdout)
        (0, import_build_utils10.debug)(`prisma generate (dummy) stdout: ${dummyResult.stdout}`);
      if (dummyResult.stderr)
        (0, import_build_utils10.debug)(`prisma generate (dummy) stderr: ${dummyResult.stderr}`);
    } catch (err) {
      throw new import_build_utils10.NowBuildError({
        code: "PRISMA_GENERATE_FAILED",
        message: `Prisma engine download failed during \`prisma generate\`. Check that your prisma version is compatible with this Python version.
` + execErrorMessage(err)
      });
    }
    const srcBinaryPrefix = `query-engine-${getLambdaBinaryTarget()}`;
    const runtimeName = `prisma-query-engine-rhel-openssl-${RUNTIME_OPENSSL_VERSION}.x`;
    const nodeModulesDir = (0, import_path10.join)(cacheDir, "node_modules", "prisma");
    let engineCopied = false;
    try {
      const entries = await import_fs7.default.promises.readdir(nodeModulesDir);
      for (const entry of entries) {
        if (!entry.startsWith(srcBinaryPrefix))
          continue;
        const srcPath = (0, import_path10.join)(nodeModulesDir, entry);
        const destPath = (0, import_path10.join)(cacheDir, runtimeName);
        try {
          await import_fs7.default.promises.access(destPath);
          (0, import_build_utils10.debug)(`Engine binary: ${runtimeName} already exists, skipping`);
        } catch {
          (0, import_build_utils10.debug)(`Engine binary: copying ${entry} -> ${runtimeName}`);
          await import_fs7.default.promises.copyFile(srcPath, destPath);
        }
        engineCopied = true;
      }
    } catch (err) {
      throw new import_build_utils10.NowBuildError({
        code: "PRISMA_ENGINE_NOT_FOUND",
        message: `could not read Prisma engine directory "${nodeModulesDir}". This may indicate an incompatible prisma version.
` + (err instanceof Error ? err.message : String(err))
      });
    }
    if (!engineCopied) {
      throw new import_build_utils10.NowBuildError({
        code: "PRISMA_ENGINE_NOT_FOUND",
        message: `could not find engine binary matching "${srcBinaryPrefix}*" in "${nodeModulesDir}". This may indicate an incompatible prisma version or an unsupported platform (${process.arch}).`
      });
    }
    const shimPath = (0, import_path10.join)(cacheDir, "openssl");
    await import_fs7.default.promises.writeFile(
      shimPath,
      `#!/bin/sh
echo "OpenSSL ${RUNTIME_OPENSSL_VERSION}.0 1 Jan 2024 (Library: OpenSSL ${RUNTIME_OPENSSL_VERSION}.0)"
`
    );
    await import_fs7.default.promises.chmod(shimPath, 493);
    for (const p of [generatedDir, dummySchemaPath]) {
      await import_fs7.default.promises.rm(p, { recursive: true, force: true });
    }
    await cleanCacheArtifacts(cacheDir);
    const generateMode = (process.env.VERCEL_PRISMA_GENERATE_CLIENT ?? "auto").toLowerCase();
    const userSchema = generateMode !== "never" ? await findUserSchema(workPath) : null;
    if (userSchema) {
      let shouldGenerate = true;
      if (generateMode === "auto") {
        const clientAlreadyGenerated = await isClientGenerated(
          pythonPath,
          pythonEnv
        );
        if (clientAlreadyGenerated) {
          (0, import_build_utils10.debug)(
            "Prisma quirk: client already generated, skipping user schema generate"
          );
          shouldGenerate = false;
        }
      }
      if (shouldGenerate) {
        (0, import_build_utils10.debug)(`Running prisma generate with user schema: ${userSchema}`);
        try {
          const userResult = await (0, import_execa4.default)(
            pythonPath,
            ["-m", "prisma", "generate", `--schema=${userSchema}`],
            {
              cwd: workPath,
              env: generateEnv,
              stdio: "pipe"
            }
          );
          if (userResult.stdout)
            (0, import_build_utils10.debug)(`prisma generate stdout: ${userResult.stdout}`);
          if (userResult.stderr)
            (0, import_build_utils10.debug)(`prisma generate stderr: ${userResult.stderr}`);
        } catch (err) {
          throw new import_build_utils10.NowBuildError({
            code: "PRISMA_GENERATE_FAILED",
            message: `\`prisma generate\` failed for schema "${userSchema}".
` + execErrorMessage(err)
          });
        }
        await cleanCacheArtifacts(cacheDir);
      }
    }
    try {
      const allFiles = await collectFiles(
        (0, import_path10.join)(sitePackages, "prisma"),
        sitePackages
      );
      const count = await (0, import_python_analysis5.extendDistRecord)(sitePackages, "prisma", allFiles);
      if (count > 0) {
        (0, import_build_utils10.debug)(`Appended ${count} entries to prisma RECORD`);
      }
    } catch (err) {
      console.warn(
        `could not patch prisma RECORD: ${err instanceof Error ? err.message : String(err)}`
      );
    }
    return {
      env: {
        PRISMA_BINARY_CACHE_DIR: runtimeCacheDir,
        VERCEL_RUNTIME_ENV_PATH_PREPEND: runtimeCacheDir
      },
      buildEnv: { PRISMA_BINARY_CACHE_DIR: cacheDir },
      alwaysBundlePackages: ["prisma"]
    };
  }
};

// src/quirks/index.ts
var quirks = [litellmQuirk, prismaQuirk, matplotlibQuirk];
function toposortQuirks(activated) {
  const nameToQuirk = /* @__PURE__ */ new Map();
  for (const q of activated) {
    nameToQuirk.set((0, import_python_analysis6.normalizePackageName)(q.dependency), q);
  }
  const adj = /* @__PURE__ */ new Map();
  const inDegree = /* @__PURE__ */ new Map();
  for (const q of activated) {
    adj.set(q, /* @__PURE__ */ new Set());
    inDegree.set(q, 0);
  }
  for (const q of activated) {
    if (q.runsBefore) {
      for (const dep of q.runsBefore) {
        const target = nameToQuirk.get((0, import_python_analysis6.normalizePackageName)(dep));
        if (target) {
          adj.get(q).add(target);
          inDegree.set(target, inDegree.get(target) + 1);
        }
      }
    }
    if (q.runsAfter) {
      for (const dep of q.runsAfter) {
        const source = nameToQuirk.get((0, import_python_analysis6.normalizePackageName)(dep));
        if (source) {
          adj.get(source).add(q);
          inDegree.set(q, inDegree.get(q) + 1);
        }
      }
    }
  }
  const queue = [];
  for (const q of activated) {
    if (inDegree.get(q) === 0) {
      queue.push(q);
    }
  }
  const sorted = [];
  while (queue.length > 0) {
    const q = queue.shift();
    sorted.push(q);
    for (const neighbor of adj.get(q)) {
      const deg = inDegree.get(neighbor) - 1;
      inDegree.set(neighbor, deg);
      if (deg === 0) {
        queue.push(neighbor);
      }
    }
  }
  if (sorted.length !== activated.length) {
    const unsorted = activated.filter((q) => !sorted.includes(q)).map((q) => q.dependency);
    throw new Error(
      `Circular dependency detected among quirks: ${unsorted.join(", ")}`
    );
  }
  return sorted;
}
async function runQuirks(ctx) {
  const mergedEnv = {};
  const mergedBuildEnv = {};
  const mergedAlwaysBundle = [];
  const installedNames = /* @__PURE__ */ new Set();
  const sitePackageDirs = await getVenvSitePackagesDirs(ctx.venvPath);
  for (const dir of sitePackageDirs) {
    const distributions = await (0, import_python_analysis6.scanDistributions)(dir);
    for (const name of distributions.keys()) {
      installedNames.add((0, import_python_analysis6.normalizePackageName)(name));
    }
  }
  const activated = quirks.filter((quirk) => {
    const installed = installedNames.has(
      (0, import_python_analysis6.normalizePackageName)(quirk.dependency)
    );
    if (!installed) {
      (0, import_build_utils11.debug)(`Quirk "${quirk.dependency}": not installed, skipping`);
    }
    return installed;
  });
  const sorted = toposortQuirks(activated);
  for (const quirk of sorted) {
    (0, import_build_utils11.debug)(`Quirk "${quirk.dependency}": detected, running fix-up`);
    const result = await quirk.run(ctx);
    if (result.env) {
      Object.assign(mergedEnv, result.env);
    }
    if (result.buildEnv) {
      Object.assign(mergedBuildEnv, result.buildEnv);
      Object.assign(process.env, result.buildEnv);
    }
    if (result.alwaysBundlePackages) {
      mergedAlwaysBundle.push(...result.alwaysBundlePackages);
    }
  }
  return {
    env: mergedEnv,
    buildEnv: mergedBuildEnv,
    alwaysBundlePackages: mergedAlwaysBundle
  };
}

// src/index.ts
var writeFile = import_fs8.default.promises.writeFile;
var version = 3;
async function downloadFilesInWorkPath({
  entrypoint,
  workPath,
  files,
  meta = {}
}) {
  (0, import_build_utils12.debug)("Downloading user files...");
  let downloadedFiles = await (0, import_build_utils12.download)(files, workPath, meta);
  if (meta.isDev) {
    const { devCacheDir = (0, import_path11.join)(workPath, ".now", "cache") } = meta;
    const destCache = (0, import_path11.join)(devCacheDir, (0, import_path11.basename)(entrypoint, ".py"));
    await (0, import_build_utils12.download)(downloadedFiles, destCache);
    downloadedFiles = await (0, import_build_utils12.glob)("**", destCache);
    workPath = destCache;
  }
  return workPath;
}
var build = async ({
  workPath,
  repoRootPath,
  files: originalFiles,
  entrypoint,
  meta = {},
  config,
  span: parentSpan
}) => {
  const builderSpan = parentSpan ?? new import_build_utils12.Span({ name: "vc.builder" });
  const framework = config?.framework;
  const shouldInstallVercelWorkers = config?.hasWorkerServices === true;
  let spawnEnv;
  let projectInstallCommand;
  let hasCustomCommand = false;
  (0, import_build_utils12.debug)(`workPath: ${workPath}`);
  workPath = await downloadFilesInWorkPath({
    workPath,
    files: originalFiles,
    entrypoint,
    meta
  });
  try {
    if (meta.isDev) {
      const setupCfg = (0, import_path11.join)(workPath, "setup.cfg");
      await writeFile(setupCfg, "[install]\nprefix=\n");
    }
  } catch (err) {
    console.log('Failed to create "setup.cfg" file');
    throw err;
  }
  let fsFiles = await (0, import_build_utils12.glob)("**", workPath);
  if ((0, import_build_utils12.isPythonFramework)(framework) && (!fsFiles[entrypoint] || !entrypoint.endsWith(".py"))) {
    const detected = await detectPythonEntrypoint(
      config.framework,
      workPath,
      entrypoint
    );
    if (detected) {
      (0, import_build_utils12.debug)(
        `Resolved Python entrypoint to "${detected}" (configured "${entrypoint}" not found).`
      );
      entrypoint = detected;
    } else {
      const searchedList = PYTHON_CANDIDATE_ENTRYPOINTS.join(", ");
      throw new import_build_utils12.NowBuildError({
        code: `${framework.toUpperCase()}_ENTRYPOINT_NOT_FOUND`,
        message: `No ${framework} entrypoint found. Add an 'app' script in pyproject.toml or define an entrypoint in one of: ${searchedList}.`,
        link: `https://vercel.com/docs/frameworks/backend/${framework}#exporting-the-${framework}-application`,
        action: "Learn More"
      });
    }
  }
  const entryDirectory = (0, import_path11.dirname)(entrypoint);
  const entrypointAbsDir = (0, import_path11.join)(workPath, entryDirectory);
  const rootDir = repoRootPath ?? workPath;
  const pythonPackage = await builderSpan.child("vc.builder.python.discover").trace(
    () => discoverPackage({
      entrypointDir: entrypointAbsDir,
      rootDir
    })
  );
  const { pythonVersion, pinVersionFilePath } = await builderSpan.child("vc.builder.python.version").trace((versionSpan) => {
    const resolution = resolvePythonVersion({
      isDev: meta.isDev,
      pythonPackage,
      rootDir
    });
    versionSpan.setAttributes({
      "python.version": pythonVersionString(resolution.pythonVersion),
      "python.versionSource": resolution.versionSource
    });
    return resolution;
  });
  if (pinVersionFilePath) {
    console.log(
      `Writing .python-version file with version ${pythonVersionString(pythonVersion)}`
    );
    await writeFile(
      pinVersionFilePath,
      `${pythonVersionString(pythonVersion)}
`
    );
  }
  fsFiles = await (0, import_build_utils12.glob)("**", workPath);
  const venvPath = (0, import_path11.join)(workPath, ".vercel", "python", ".venv");
  await builderSpan.child("vc.builder.python.venv").trace(async () => {
    await ensureVenv({
      pythonPath: pythonVersion.pythonPath,
      venvPath
    });
  });
  if ((0, import_build_utils12.isPythonFramework)(framework)) {
    const {
      cliType,
      lockfileVersion,
      packageJsonPackageManager,
      turboSupportsCorepackHome
    } = await (0, import_build_utils12.scanParentDirs)(workPath, true);
    spawnEnv = (0, import_build_utils12.getEnvForPackageManager)({
      cliType,
      lockfileVersion,
      packageJsonPackageManager,
      env: process.env,
      turboSupportsCorepackHome,
      projectCreatedAt: config?.projectSettings?.createdAt
    });
    const installCommand = config?.projectSettings?.installCommand;
    if (typeof installCommand === "string") {
      const trimmed = installCommand.trim();
      if (trimmed) {
        projectInstallCommand = trimmed;
      } else {
        console.log('Skipping "install" command...');
      }
    }
  }
  const baseEnv = spawnEnv || process.env;
  const pythonEnv = createVenvEnv(venvPath, baseEnv);
  pythonEnv.VERCEL_PYTHON_VENV_PATH = venvPath;
  let assumeDepsInstalled = false;
  let uv;
  try {
    const uvPath = await getUvBinaryOrInstall(pythonVersion.pythonPath);
    console.log(`Using uv at "${uvPath}"`);
    uv = new UvRunner(uvPath);
  } catch (err) {
    console.log("Failed to install or locate uv");
    throw new Error(
      `uv is required for this project but failed to install: ${err instanceof Error ? err.message : String(err)}`
    );
  }
  let uvLockPath = null;
  let uvProjectDir = null;
  let projectName;
  let noBuildCheckFailed = false;
  await builderSpan.child(import_build_utils12.BUILDER_INSTALLER_STEP, {
    installCommand: projectInstallCommand || void 0
  }).trace(async () => {
    if (projectInstallCommand) {
      console.log(
        `Running "install" command: \`${projectInstallCommand}\`...`
      );
      await (0, import_build_utils12.execCommand)(projectInstallCommand, {
        env: pythonEnv,
        cwd: workPath
      });
      assumeDepsInstalled = true;
      hasCustomCommand = true;
    } else {
      assumeDepsInstalled = await runPyprojectScript(
        workPath,
        ["vercel-install", "now-install", "install"],
        pythonEnv,
        /* useUserVirtualEnv */
        false
      );
      if (assumeDepsInstalled) {
        hasCustomCommand = true;
      }
    }
    if (!assumeDepsInstalled) {
      const { projectDir, lockPath, lockFileProvidedByUser } = await ensureUvProject({
        workPath,
        rootDir,
        pythonPackage,
        pythonVersion: pythonVersionString(pythonVersion),
        uv,
        generateLockFile: true,
        requireBinaryWheels: false
      });
      uvLockPath = lockPath;
      uvProjectDir = projectDir;
      projectName = pythonPackage?.manifest?.data?.project?.name;
      if (lockFileProvidedByUser) {
        try {
          await uv.sync({
            venvPath,
            projectDir,
            frozen: true,
            noBuild: true,
            noInstallProject: true
          });
        } catch (err) {
          noBuildCheckFailed = true;
          (0, import_build_utils12.debug)(
            `--no-build check failed: ${err instanceof Error ? err.message : String(err)}`
          );
        }
      }
      await uv.sync({
        venvPath,
        projectDir,
        frozen: lockFileProvidedByUser,
        locked: !lockFileProvidedByUser
      });
    }
  });
  if ((0, import_build_utils12.isPythonFramework)(framework)) {
    const projectBuildCommand = config?.projectSettings?.buildCommand ?? // fallback if provided directly on config (some callers set this)
    config?.buildCommand;
    await builderSpan.child(import_build_utils12.BUILDER_COMPILE_STEP, {
      buildCommand: projectBuildCommand || void 0
    }).trace(async () => {
      if (projectBuildCommand) {
        console.log(`Running "${projectBuildCommand}"`);
        await (0, import_build_utils12.execCommand)(projectBuildCommand, {
          env: pythonEnv,
          cwd: workPath
        });
      } else {
        await runPyprojectScript(
          workPath,
          ["vercel-build", "now-build", "build"],
          pythonEnv
        );
      }
    });
  }
  const runtimeDep = baseEnv.VERCEL_RUNTIME_PYTHON || `vercel-runtime==${VERCEL_RUNTIME_VERSION}`;
  (0, import_build_utils12.debug)(`Installing ${runtimeDep}`);
  await uv.pip({
    venvPath,
    projectDir: (0, import_path11.join)(workPath, entryDirectory),
    args: ["install", runtimeDep]
  });
  if (shouldInstallVercelWorkers) {
    const workersDep = baseEnv.VERCEL_WORKERS_PYTHON || `vercel-workers==${VERCEL_WORKERS_VERSION}`;
    (0, import_build_utils12.debug)(`Installing ${workersDep}`);
    await uv.pip({
      venvPath,
      projectDir: (0, import_path11.join)(workPath, entryDirectory),
      args: ["install", workersDep]
    });
  }
  const quirksResult = await runQuirks({ venvPath, pythonEnv, workPath });
  if (quirksResult.buildEnv) {
    Object.assign(pythonEnv, quirksResult.buildEnv);
  }
  (0, import_build_utils12.debug)("Entrypoint is", entrypoint);
  const moduleName = entrypoint.replace(/\//g, ".").replace(/\.py$/i, "");
  const vendorDir = resolveVendorDir();
  const suffix = meta.isDev && !entrypoint.endsWith(".py") ? ".py" : "";
  const entrypointWithSuffix = `${entrypoint}${suffix}`;
  (0, import_build_utils12.debug)("Entrypoint with suffix is", entrypointWithSuffix);
  const runtimeTrampoline = `
import importlib
import os
import os.path
import site
import sys

_here = os.path.dirname(__file__)

os.environ.update({
  "__VC_HANDLER_MODULE_NAME": "${moduleName}",
  "__VC_HANDLER_ENTRYPOINT": "${entrypointWithSuffix}",
  "__VC_HANDLER_ENTRYPOINT_ABS": os.path.join(_here, "${entrypointWithSuffix}"),
  "__VC_HANDLER_VENDOR_DIR": "${vendorDir}",
})

_vendor_rel = '${vendorDir}'
_vendor = os.path.normpath(os.path.join(_here, _vendor_rel))

if os.path.isdir(_vendor):
    # Process .pth files like a real site-packages dir
    site.addsitedir(_vendor)

    # Move _vendor to the front (after script dir if present)
    try:
        while _vendor in sys.path:
            sys.path.remove(_vendor)
    except ValueError:
        pass

    # Put vendored deps ahead of site-packages but after the script dir
    idx = 1 if (sys.path and sys.path[0] in ('', _here)) else 0
    sys.path.insert(idx, _vendor)

    importlib.invalidate_caches()

from vercel_runtime.vc_init import vc_handler
`;
  const predefinedExcludes = [
    ".git/**",
    ".gitignore",
    ".vercel/**",
    ".pnpm-store/**",
    "**/node_modules/**",
    "**/.next/**",
    "**/.nuxt/**",
    "**/.venv/**",
    "**/venv/**",
    "**/__pycache__/**",
    "**/.mypy_cache/**",
    "**/.ruff_cache/**",
    "**/public/**",
    "**/pnpm-lock.yaml",
    "**/yarn.lock",
    "**/package-lock.json"
  ];
  const lambdaEnv = {};
  lambdaEnv.PYTHONPATH = vendorDir;
  Object.assign(lambdaEnv, quirksResult.env);
  const globOptions = {
    cwd: workPath,
    ignore: config && typeof config.excludeFiles === "string" ? [...predefinedExcludes, config.excludeFiles] : predefinedExcludes
  };
  const files = await (0, import_build_utils12.glob)("**", globOptions);
  const depExternalizer = new PythonDependencyExternalizer({
    venvPath,
    vendorDir,
    workPath,
    uvLockPath,
    uvProjectDir,
    projectName,
    noBuildCheckFailed,
    pythonPath: pythonVersion.pythonPath,
    hasCustomCommand,
    alwaysBundlePackages: [
      ...quirksResult.alwaysBundlePackages ?? [],
      ...shouldInstallVercelWorkers ? ["vercel-workers", "vercel_workers"] : []
    ]
  });
  await builderSpan.child("vc.builder.python.bundle").trace(async (bundleSpan) => {
    const depAnalysis = await depExternalizer.analyze(files);
    bundleSpan.setAttributes({
      "python.bundle.totalSizeBytes": String(depAnalysis.totalBundleSize),
      "python.bundle.runtimeInstallEnabled": String(
        depAnalysis.runtimeInstallEnabled
      )
    });
    if (depAnalysis.runtimeInstallEnabled) {
      await depExternalizer.generateBundle(files);
    } else {
      for (const [p, f] of Object.entries(depAnalysis.allVendorFiles)) {
        files[p] = f;
      }
    }
  });
  const handlerPyFilename = "vc__handler__python";
  files[`${handlerPyFilename}.py`] = new import_build_utils12.FileBlob({ data: runtimeTrampoline });
  if (config.framework === "fasthtml") {
    const { SESSKEY = "" } = process.env;
    files[".sesskey"] = new import_build_utils12.FileBlob({ data: `"${SESSKEY}"` });
  }
  const output = new import_build_utils12.Lambda({
    files,
    handler: `${handlerPyFilename}.vc_handler`,
    runtime: pythonVersion.runtime,
    environment: lambdaEnv,
    supportsResponseStreaming: true
  });
  return { output };
};
var shouldServe = (opts) => {
  const framework = opts.config.framework;
  if ((0, import_build_utils12.isPythonFramework)(framework)) {
    const requestPath = opts.requestPath.replace(/\/$/, "");
    if (requestPath.startsWith("api") && opts.hasMatched) {
      return false;
    }
    return true;
  }
  return defaultShouldServe(opts);
};
var defaultShouldServe = ({
  entrypoint,
  files,
  requestPath
}) => {
  requestPath = requestPath.replace(/\/$/, "");
  entrypoint = entrypoint.replace(/\\/g, "/");
  if (entrypoint === requestPath && hasProp(files, entrypoint)) {
    return true;
  }
  const { dir, name } = (0, import_path11.parse)(entrypoint);
  if (name === "index" && dir === requestPath && hasProp(files, entrypoint)) {
    return true;
  }
  return false;
};
function hasProp(obj, key) {
  return Object.hasOwnProperty.call(obj, key);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  build,
  defaultShouldServe,
  downloadFilesInWorkPath,
  installRequirement,
  installRequirementsFile,
  shouldServe,
  startDevServer,
  version
});
